# Extension to secure shell programs: GWAS with PLINK {#shell_extension}

GWAS can also be performed using programs that are executed using shell commands. This is the case for [PLINK](http://zzz.bwh.harvard.edu/plink/), one of the state-of-the-art programs to run GWAS and other genomic data analyses such gene-enviroment interactions or polygenic risc score analyses that require efficient and scalable pipelines.

Resources also allow the use of secure SSH service to run programs on a remote server accessible through SSH containing data and analysis tools where R is just used for launching the analyses and aggregating results. This feature allows us to create functions to analyze data using specific shell programs.

Here we describe how the [PLINK](http://zzz.bwh.harvard.edu/plink/) program can be used to perform GWAS. In this case, the resource describes that access is given via SSH, the credentials required to connect, and the commands that can be run (of which one is plink).

We use the following code to illustrate how analyses should be performed using the `r CRANpkg("resourcer")` package. This code could be considered as the base code for creating a server-side DataSHIELD package as performed in the `plinkDS()` function implemented in the `r Githubpkg("isglobal-brge", "dsOmics")` package


We access the ssh resource called `brge_plink` (Figure \@ref(fig:testResources2)) using the `r CRANpkg("resourcer")` package as follows:


```{r plink_resource}
library(resourcer)
brge_plink <- resourcer::newResource(url="ssh://plink-demo.obiba.org:2222/home/master/brge?exec=ls,plink,plink1",
                                     identity = "master", secret = "nSATfSy9Y8JAo5zK")
client <- resourcer::newResourceClient(brge_plink)
```

This creates an object of this class:

```{r plink_resource_class}
class(client)
```

These are the actions we can do with an `SshResourceClient` object

```{r plink_resource_actions}
names(client)
```

For this specific resource (e.g. PLINK) we can execute these shell commands

```{r plink_resource_allowed}
client$getAllowedCommands()
```

For instance

```{r plink_resource_ls}
client$exec("ls", "-la")
```

Then, to avoid multiple accesses to the resource, it is recommended to create a temporary directory to save our results

```{r plink_resource_temp}
tempDir <- client$tempDir()
tempDir
client$exec("ls", tempDir)
```

Then, we can use R to launch the shell commands

```{r plink_resource_plink}
client$exec('plink1', c('--bfile', 'brge', '--freq', '--out', paste0(tempDir, '/out'), '--noweb'))
```

The results can be retrieved as an R object

```{r plink_resource_retrieve}
outs <- client$exec('ls', tempDir)$output
outs
client$downloadFile(paste0(tempDir, '/out.frq'))  
ans <- readr::read_table("out.frq")
ans
```

Finally temporary directories are removed and the session closed

```{r plink_resource_close}
client$removeTempDir()
client$close()
```
