<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>5 DataSHIELD | Orchestrating non-disclosive big data analyses of data from different resources with R and DataSHIELD</title>
  <meta name="description" content="5 DataSHIELD | Orchestrating non-disclosive big data analyses of data from different resources with R and DataSHIELD" />
  <meta name="generator" content="bookdown 0.20 and GitBook 2.6.7" />

  <meta property="og:title" content="5 DataSHIELD | Orchestrating non-disclosive big data analyses of data from different resources with R and DataSHIELD" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="5 DataSHIELD | Orchestrating non-disclosive big data analyses of data from different resources with R and DataSHIELD" />
  
  
  



<meta name="date" content="2020-09-07" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="opal.html"/>
<link rel="next" href="resources.html"/>
<script src="libs/header-attrs-2.3/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Preamble</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#what-you-will-learn"><i class="fa fa-check"></i><b>1.1</b> What you will learn</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#preliminaries"><i class="fa fa-check"></i><b>1.2</b> Preliminaries</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#acknowledgments"><i class="fa fa-check"></i><b>1.3</b> Acknowledgments</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="learnR.html"><a href="learnR.html"><i class="fa fa-check"></i><b>2</b> Learning R</a></li>
<li class="chapter" data-level="3" data-path="BioC.html"><a href="BioC.html"><i class="fa fa-check"></i><b>3</b> Bioconductor Data Structures</a>
<ul>
<li class="chapter" data-level="3.1" data-path="BioC.html"><a href="BioC.html#snp-array-data"><i class="fa fa-check"></i><b>3.1</b> SNP Array Data</a></li>
<li class="chapter" data-level="3.2" data-path="BioC.html"><a href="BioC.html#expression-sets"><i class="fa fa-check"></i><b>3.2</b> Expression Sets</a></li>
<li class="chapter" data-level="3.3" data-path="BioC.html"><a href="BioC.html#genomic-ranges"><i class="fa fa-check"></i><b>3.3</b> Genomic Ranges</a></li>
<li class="chapter" data-level="3.4" data-path="BioC.html"><a href="BioC.html#summarized-experiments"><i class="fa fa-check"></i><b>3.4</b> Summarized Experiments</a></li>
<li class="chapter" data-level="3.5" data-path="BioC.html"><a href="BioC.html#ranged-summarized-experiments"><i class="fa fa-check"></i><b>3.5</b> Ranged Summarized Experiments</a></li>
<li class="chapter" data-level="3.6" data-path="BioC.html"><a href="BioC.html#multi-data-set"><i class="fa fa-check"></i><b>3.6</b> Multi Data Set</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="opal.html"><a href="opal.html"><i class="fa fa-check"></i><b>4</b> Opal</a>
<ul>
<li class="chapter" data-level="4.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>4.1</b> Introduction</a></li>
<li class="chapter" data-level="4.2" data-path="opal.html"><a href="opal.html#data-management"><i class="fa fa-check"></i><b>4.2</b> Data Management</a></li>
<li class="chapter" data-level="4.3" data-path="opal.html"><a href="opal.html#security"><i class="fa fa-check"></i><b>4.3</b> Security</a></li>
<li class="chapter" data-level="4.4" data-path="opal.html"><a href="opal.html#r-integration"><i class="fa fa-check"></i><b>4.4</b> R Integration</a></li>
<li class="chapter" data-level="4.5" data-path="opal.html"><a href="opal.html#demo"><i class="fa fa-check"></i><b>4.5</b> Demo</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="datashield.html"><a href="datashield.html"><i class="fa fa-check"></i><b>5</b> DataSHIELD</a>
<ul>
<li class="chapter" data-level="5.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>5.1</b> Introduction</a></li>
<li class="chapter" data-level="5.2" data-path="datashield.html"><a href="datashield.html#datashield-r-interface-dsi"><i class="fa fa-check"></i><b>5.2</b> DataSHIELD R Interface (DSI)</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="datashield.html"><a href="datashield.html#classes-structure"><i class="fa fa-check"></i><b>5.2.1</b> Classes Structure</a></li>
<li class="chapter" data-level="5.2.2" data-path="datashield.html"><a href="datashield.html#higher-level-functions"><i class="fa fa-check"></i><b>5.2.2</b> Higher Level Functions</a></li>
<li class="chapter" data-level="5.2.3" data-path="datashield.html"><a href="datashield.html#options"><i class="fa fa-check"></i><b>5.2.3</b> Options</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="datashield.html"><a href="datashield.html#datashieldopal-implementation"><i class="fa fa-check"></i><b>5.3</b> DataSHIELD/Opal Implementation</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="datashield.html"><a href="datashield.html#client"><i class="fa fa-check"></i><b>5.3.1</b> Client</a></li>
<li class="chapter" data-level="5.3.2" data-path="datashield.html"><a href="datashield.html#server"><i class="fa fa-check"></i><b>5.3.2</b> Server</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="opal.html"><a href="opal.html#demo"><i class="fa fa-check"></i><b>5.4</b> Demo</a></li>
</ul></li>
<li class="part"><span><b>II Focus Topics</b></span></li>
<li class="chapter" data-level="6" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>6</b> The Resources</a>
<ul>
<li class="chapter" data-level="6.1" data-path="resources.html"><a href="resources.html#concept"><i class="fa fa-check"></i><b>6.1</b> Concept</a></li>
<li class="chapter" data-level="6.2" data-path="resources.html"><a href="resources.html#types"><i class="fa fa-check"></i><b>6.2</b> Types</a></li>
<li class="chapter" data-level="6.3" data-path="resources.html"><a href="resources.html#definition"><i class="fa fa-check"></i><b>6.3</b> Definition</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="resourcer.html"><a href="resourcer.html"><i class="fa fa-check"></i><b>7</b> The resourcer Package</a>
<ul>
<li class="chapter" data-level="7.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>7.1</b> Introduction</a></li>
<li class="chapter" data-level="7.2" data-path="resourcer.html"><a href="resourcer.html#file-resources"><i class="fa fa-check"></i><b>7.2</b> File Resources</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="resourcer.html"><a href="resourcer.html#file-getter"><i class="fa fa-check"></i><b>7.2.1</b> File Getter</a></li>
<li class="chapter" data-level="7.2.2" data-path="resourcer.html"><a href="resourcer.html#file-data-format"><i class="fa fa-check"></i><b>7.2.2</b> File Data Format</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="resourcer.html"><a href="resourcer.html#database-resources"><i class="fa fa-check"></i><b>7.3</b> Database Resources</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="resourcer.html"><a href="resourcer.html#dbi-connectors"><i class="fa fa-check"></i><b>7.3.1</b> DBI Connectors</a></li>
<li class="chapter" data-level="7.3.2" data-path="resourcer.html"><a href="resourcer.html#use-dplyr"><i class="fa fa-check"></i><b>7.3.2</b> Use dplyr</a></li>
<li class="chapter" data-level="7.3.3" data-path="resourcer.html"><a href="resourcer.html#document-databases"><i class="fa fa-check"></i><b>7.3.3</b> Document Databases</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="resourcer.html"><a href="resourcer.html#computation-resources"><i class="fa fa-check"></i><b>7.4</b> Computation Resources</a></li>
<li class="chapter" data-level="7.5" data-path="resourcer.html"><a href="resourcer.html#interacting-with-r-resources"><i class="fa fa-check"></i><b>7.5</b> Interacting with R Resources</a></li>
<li class="chapter" data-level="7.6" data-path="resourcer.html"><a href="resourcer.html#extending_resources"><i class="fa fa-check"></i><b>7.6</b> Extending Resources</a></li>
<li class="chapter" data-level="7.7" data-path="resourcer.html"><a href="resourcer.html#resource-forms"><i class="fa fa-check"></i><b>7.7</b> Resource Forms</a></li>
<li class="chapter" data-level="7.8" data-path="resourcer.html"><a href="resourcer.html#examples-of-resources"><i class="fa fa-check"></i><b>7.8</b> Examples of Resources</a>
<ul>
<li class="chapter" data-level="7.8.1" data-path="resourcer.html"><a href="resourcer.html#tsv-file-into-a-tibble-or-data.frame"><i class="fa fa-check"></i><b>7.8.1</b> TSV File into a tibble or data.frame</a></li>
<li class="chapter" data-level="7.8.2" data-path="resourcer.html"><a href="resourcer.html#r-data-file-into-a-r-object"><i class="fa fa-check"></i><b>7.8.2</b> R Data File into a R object</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>III Resources Extensions</b></span></li>
<li class="chapter" data-level="8" data-path="omic-extension.html"><a href="omic-extension.html"><i class="fa fa-check"></i><b>8</b> VCF files to GDS to peform GWAS with Bioconductor</a></li>
<li class="chapter" data-level="9" data-path="shell-extension.html"><a href="shell-extension.html"><i class="fa fa-check"></i><b>9</b> Secure shell programs: GWAS with PLINK</a></li>
<li class="part"><span><b>IV Workflows</b></span></li>
<li class="chapter" data-level="10" data-path="setup.html"><a href="setup.html"><i class="fa fa-check"></i><b>10</b> Setup</a>
<ul>
<li class="chapter" data-level="10.1" data-path="setup.html"><a href="setup.html#providing-datashield-packages-in-the-opal-server"><i class="fa fa-check"></i><b>10.1</b> Providing DataSHIELD packages in the opal server</a></li>
<li class="chapter" data-level="10.2" data-path="setup.html"><a href="setup.html#required-r-packages-in-the-client-site-e.g.-local-machine"><i class="fa fa-check"></i><b>10.2</b> Required R Packages in the client site (e.g. local machine)</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="basic-statistical-analyses.html"><a href="basic-statistical-analyses.html"><i class="fa fa-check"></i><b>11</b> Basic statistical analyses</a>
<ul>
<li class="chapter" data-level="11.1" data-path="basic-statistical-analyses.html"><a href="basic-statistical-analyses.html#analysis-from-a-single-study"><i class="fa fa-check"></i><b>11.1</b> Analysis from a single study</a></li>
<li class="chapter" data-level="11.2" data-path="basic-statistical-analyses.html"><a href="basic-statistical-analyses.html#analysis-from-a-multiple-studies"><i class="fa fa-check"></i><b>11.2</b> Analysis from a multiple studies</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="Omic.html"><a href="Omic.html"><i class="fa fa-check"></i><b>12</b> Omic data analysis</a>
<ul>
<li class="chapter" data-level="12.1" data-path="Omic.html"><a href="Omic.html#types-of-analyses-implemented"><i class="fa fa-check"></i><b>12.1</b> Types of analyses implemented</a></li>
<li class="chapter" data-level="12.2" data-path="Omic.html"><a href="Omic.html#differential-gene-expression-dge-analysis"><i class="fa fa-check"></i><b>12.2</b> Differential gene expression (DGE) analysis</a></li>
<li class="chapter" data-level="12.3" data-path="Omic.html"><a href="Omic.html#epigenome-wide-association-analysis-ewas"><i class="fa fa-check"></i><b>12.3</b> Epigenome-wide association analysis (EWAS)</a>
<ul>
<li class="chapter" data-level="12.3.1" data-path="Omic.html"><a href="Omic.html#single-cpg-analysis"><i class="fa fa-check"></i><b>12.3.1</b> Single CpG analysis</a></li>
<li class="chapter" data-level="12.3.2" data-path="Omic.html"><a href="Omic.html#multiple-cpg-analysis"><i class="fa fa-check"></i><b>12.3.2</b> Multiple CpG analysis</a></li>
<li class="chapter" data-level="12.3.3" data-path="Omic.html"><a href="Omic.html#adjusting-for-surrogate-variables"><i class="fa fa-check"></i><b>12.3.3</b> Adjusting for Surrogate Variables</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="Omic.html"><a href="Omic.html#gwas-with-bioconductor"><i class="fa fa-check"></i><b>12.4</b> GWAS with Bioconductor</a></li>
<li class="chapter" data-level="12.5" data-path="Omic.html"><a href="Omic.html#gwas-with-plink"><i class="fa fa-check"></i><b>12.5</b> GWAS with PLINK</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="GIS.html"><a href="GIS.html"><i class="fa fa-check"></i><b>13</b> Geospatial data analysis</a>
<ul>
<li class="chapter" data-level="13.1" data-path="GIS.html"><a href="GIS.html#introducing-the-analysis"><i class="fa fa-check"></i><b>13.1</b> Introducing the analysis</a></li>
<li class="chapter" data-level="13.2" data-path="GIS.html"><a href="GIS.html#setting-up-the-analysis"><i class="fa fa-check"></i><b>13.2</b> Setting up the analysis</a></li>
<li class="chapter" data-level="13.3" data-path="GIS.html"><a href="GIS.html#data-manipulation"><i class="fa fa-check"></i><b>13.3</b> Data manipulation</a></li>
<li class="chapter" data-level="13.4" data-path="GIS.html"><a href="GIS.html#generating-the-final-result"><i class="fa fa-check"></i><b>13.4</b> Generating the final result</a></li>
</ul></li>
<li class="part"><span><b>V Developers</b></span></li>
<li class="chapter" data-level="14" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html"><i class="fa fa-check"></i><b>14</b> DSLite: DataSHIELD Implementation on Local Datasets</a>
<ul>
<li class="chapter" data-level="14.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#development-environment-setup"><i class="fa fa-check"></i><b>14.1</b> Development Environment Setup</a>
<ul>
<li class="chapter" data-level="14.1.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-packages"><i class="fa fa-check"></i><b>14.1.1</b> DataSHIELD Packages</a></li>
<li class="chapter" data-level="14.1.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#test-datasets"><i class="fa fa-check"></i><b>14.1.2</b> Test Datasets</a></li>
<li class="chapter" data-level="14.1.3" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#r-package-development-tools"><i class="fa fa-check"></i><b>14.1.3</b> R Package Development Tools</a></li>
</ul></li>
<li class="chapter" data-level="14.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-development-flow"><i class="fa fa-check"></i><b>14.2</b> DataSHIELD Development Flow</a>
<ul>
<li class="chapter" data-level="14.2.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#dsliteserver"><i class="fa fa-check"></i><b>14.2.1</b> DSLiteServer</a></li>
<li class="chapter" data-level="14.2.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-configuration"><i class="fa fa-check"></i><b>14.2.2</b> DataSHIELD Configuration</a></li>
</ul></li>
<li class="chapter" data-level="14.3" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-sessions"><i class="fa fa-check"></i><b>14.3</b> DataSHIELD Sessions</a></li>
<li class="chapter" data-level="14.4" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#debugging"><i class="fa fa-check"></i><b>14.4</b> Debugging</a></li>
<li class="chapter" data-level="14.5" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#limitations"><i class="fa fa-check"></i><b>14.5</b> Limitations</a>
<ul>
<li class="chapter" data-level="14.5.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#function-parameters-parser"><i class="fa fa-check"></i><b>14.5.1</b> Function Parameters Parser</a></li>
<li class="chapter" data-level="14.5.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#server-side-environments"><i class="fa fa-check"></i><b>14.5.2</b> Server Side Environments</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="15" data-path="creating-datashield-packages.html"><a href="creating-datashield-packages.html"><i class="fa fa-check"></i><b>15</b> Creating DataSHIELD packages</a></li>
<li class="chapter" data-level="16" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html"><i class="fa fa-check"></i><b>16</b> Tips and tricks</a>
<ul>
<li class="chapter" data-level="16.1" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#how-to-install-r-packages-into-opal-server-from-r"><i class="fa fa-check"></i><b>16.1</b> How to install R packages into OPAL server from R</a></li>
<li class="chapter" data-level="16.2" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#how-to-check-whether-there-are-open-r-sesions-in-the-opal-server"><i class="fa fa-check"></i><b>16.2</b> How to check whether there are open R sesions in the OPAL server</a></li>
<li class="chapter" data-level="16.3" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#how"><i class="fa fa-check"></i><b>16.3</b> How</a></li>
</ul></li>
<li class="part"><span><b>VI Appendix</b></span></li>
<li class="chapter" data-level="17" data-path="session-info.html"><a href="session-info.html"><i class="fa fa-check"></i><b>17</b> Session Info</a></li>
<li class="chapter" data-level="18" data-path="contributors.html"><a href="contributors.html"><i class="fa fa-check"></i><b>18</b> Contributors</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating non-disclosive big data analyses of data from different resources with R and DataSHIELD</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="datashield" class="section level1" number="5">
<h1><span class="header-section-number">5</span> DataSHIELD</h1>
<div id="introduction" class="section level2" number="5.1">
<h2><span class="header-section-number">5.1</span> Introduction</h2>
<p>Some research projects require pooling data from several studies to obtain sample sizes large and diverse enough for detecting interactions. Unfortunately, important ethico-legal constraints often prevent or impede the sharing of individual-level data across multiple studies. DataSHIELD aims to address this issue. DataSHIELD is a method that enables advanced statistical analysis of individual-level data from several sources without actually pooling the data from these sources together. DataSHIELD facilitates important research in settings where:</p>
<ul>
<li>a co-analysis of individual-level data from several studies is scientifically necessary but governance restrictions prevent the release or sharing of some of the required data, and/or render data access unacceptably slow,</li>
<li>equivalent governance concerns prevent or hinder access to a single dataset,</li>
<li>a research group wishes to actively share the information held in its data with others but does not wish to cede control of the governance of those data and/or the intellectual property they represent by physically handing over the data themselves,</li>
<li>a dataset which is to be remotely analysed, or included in a multi-study co-analysis, contains data objects (e.g. images) which are too large to be physically transferred to the site of analysis.</li>
</ul>
<p>A typical DataSHIELD infrastructure (see Figure <a href="datashield.html#fig:dsArchitec">5.1</a>) is composed of one central analysis node (the DataSHIELD client) connected to one or several data analysis nodes (the DataSHIELD servers). In each of these server nodes, there is a R server application which can be accessed only through a DataSHIELD compliant middleware application. This middleware application acts as a broker for managing R server sessions in a multi-user environment, assigning data and launching analysis in the R server. The analysis execution environment is then fully controlled: users must be authenticated, must have the proper permissions to access the data of interest and can only perform some predefined assignment and aggregation operations. Importantly, the operations that are permitted are designed to prevent the user having access to individual data items while still allowing useful work to be done with the data. For example, users can fit a generalised linear model to a dataset and receive information about the model coefficients, but are not given the residuals, as these could be used to reconstruct the original data.The reference implementation of this DataSHIELD infrastructure is based on the <a href="http://opaldoc.obiba.org/">Opal</a> data repository.</p>
<div class="figure" style="text-align: center"><span id="fig:dsArchitec"></span>
<p class="caption">
Figure 5.1: Typical DataSHIELD infrastructure, including one central analysis node (client) and several data nodes (servers).
</p>
<img src="fig/singleSiteDSInfrastructure.jpg" alt="Typical DataSHIELD infrastructure, including one central analysis node (client) and several data nodes (servers). " width="80%" />
</div>
<p>The client node interacts programmatically in R with the server nodes using the DataSHIELD Interface implemented as the <a href="https://cran.r-project.org/package=DSI">DSI</a> R package. The DSI defines prototype functions to authenticate the user and to perform assignment and aggregation operations in each of the R servers sitting in the server nodes. The reference implementation of DSI is the <a href="https://cran.r-project.org/package=DSOpal">DSOpal</a> R package. An alternate implementation of DSI is the <a href="https://cran.r-project.org/package=DSLite">DSLite</a> R package which mainly targets DataSHIELD developers by offering a pure R implementation of the whole DataSHIELD infrastructure.</p>
</div>
<div id="datashield-r-interface-dsi" class="section level2" number="5.2">
<h2><span class="header-section-number">5.2</span> DataSHIELD R Interface (DSI)</h2>
<p>The DataSHIELD Interface (DSI) defines a set of <a href="http://adv-r.had.co.nz/S4.html">S4 classes and generic methods</a> that can be implemented for accessing a data repository supporting the DataSHIELD infrastructure: controlled R commands to be executed on the server side are garanteeing that non disclosive information is returned to client side.</p>
<div id="classes-structure" class="section level3" number="5.2.1">
<h3><span class="header-section-number">5.2.1</span> Classes Structure</h3>
<p>The DSI S4 classes are:</p>
<table>
<thead>
<tr class="header">
<th align="left">Class</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left"><code>DSObject</code></td>
<td align="left">A common base class for all DSI,</td>
</tr>
<tr class="even">
<td align="left"><code>DSDriver</code></td>
<td align="left">A class to drive the creation of a connection object,</td>
</tr>
<tr class="odd">
<td align="left"><code>DSConnection</code></td>
<td align="left">Allows the interaction with the remote server; DataSHIELD operations such as aggregation and assignment return a result object; DataSHIELD setup status check can be performed (dataset access, configuration comparision),</td>
</tr>
<tr class="even">
<td align="left"><code>DSResult</code></td>
<td align="left">Wraps access to the result, which can be fetched either synchronously or asynchronously depending on the capabilities of the data repository server.</td>
</tr>
</tbody>
</table>
<p>All classes are <em>virtual</em>: they cannot be instantiated directly and instead must be subclassed. See <a href="https://github.com/datashield/DSOpal">DSOpal</a> for a reference implementation of DSI based on the <a href="http://opaldoc.obiba.org/">Opal</a> data repository.</p>
<p>These S4 classes and generic methods are meant <strong>to be used for implementing connection to a DataSHIELD-aware data repository.</strong></p>
</div>
<div id="higher-level-functions" class="section level3" number="5.2.2">
<h3><span class="header-section-number">5.2.2</span> Higher Level Functions</h3>
<p>In addition to these S4 classes, DSI provides functions to handle a list of remote data repository servers:</p>
<table>
<thead>
<tr class="header">
<th align="left">Functions</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left"><code>datashield.login</code></td>
<td align="left">Create <code>DSConnection</code> objects to the data repositories, using the <code>DSDriver</code> specification.</td>
</tr>
<tr class="even">
<td align="left"><code>datashield.logout</code></td>
<td align="left">Destroy the <code>DSConnection</code>s objects.</td>
</tr>
<tr class="odd">
<td align="left"><code>datashield.aggregate</code>, <code>datashield.assign</code></td>
<td align="left">Typical DataSHIELD operations on <code>DSConnection</code> objects, which result will be fetched through <code>DSResult</code> objects.</td>
</tr>
<tr class="even">
<td align="left"><code>datashield.connections</code>, <code>datashield.connections_default</code>, <code>datashield.connections_find</code></td>
<td align="left">Management of the list of <code>DSConnection</code> objects that will be discovered and used by the client-side analytic functions.</td>
</tr>
<tr class="odd">
<td align="left"><code>datashield.workspaces</code>, <code>datashield.workspace_save</code>, <code>datashield.workspace_rm</code></td>
<td align="left">Manage R images of the remote DataSHIELD sessions (to speed up data analysis sessions).</td>
</tr>
<tr class="even">
<td align="left"><code>datashield.symbols</code>, <code>datashield.symbol_rm</code></td>
<td align="left">Minimalistic management of the R symbols living in the remote DataSHIELD sessions.</td>
</tr>
<tr class="odd">
<td align="left"><code>datashield.tables</code>, <code>datashield.table_status</code></td>
<td align="left">List the tables and their accessibility across a set of data repositories.</td>
</tr>
<tr class="even">
<td align="left"><code>datashield.resources</code>, <code>datashield.resource_status</code></td>
<td align="left">List the resources and their accessibility across a set of data repositories.</td>
</tr>
<tr class="odd">
<td align="left"><code>datashield.pkg_status</code>, <code>datashield.method_status</code>, <code>datashield.methods</code></td>
<td align="left">Utility functions to explore the DataSHIELD setup across a set of data repositories.</td>
</tr>
</tbody>
</table>
<p>These <code>datashield.*</code> functions are meant <strong>to be used by DataSHIELD packages developers and users.</strong></p>
</div>
<div id="options" class="section level3" number="5.2.3">
<h3><span class="header-section-number">5.2.3</span> Options</h3>
<p>Some options can be set to modify the behavior of the DSI:</p>
<table>
<thead>
<tr class="header">
<th align="left">Option</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left"><code>datashield.env</code></td>
<td align="left">The R environment in which the <code>DSConnection</code> object list is to be looked for. Default value is the Global Environment: <code>globalenv()</code>.</td>
</tr>
<tr class="even">
<td align="left"><code>datashield.progress</code></td>
<td align="left">A logical to enable the visibility of the progress bars. Default value is <code>TRUE</code>.</td>
</tr>
<tr class="odd">
<td align="left"><code>datashield.progress.clear</code></td>
<td align="left">A logical to make the progress bar disappear after it has been completed. Default value is <code>FALSE</code>.</td>
</tr>
<tr class="even">
<td align="left"><code>datashiel.error.stop</code></td>
<td align="left">A logical to alter error handling behavior: if <code>TRUE</code> an error is raised when at least one server has failed, otherwise a warning message is issued. Default value is <code>TRUE</code>.</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="datashieldopal-implementation" class="section level2" number="5.3">
<h2><span class="header-section-number">5.3</span> DataSHIELD/Opal Implementation</h2>
<p><a href="opal.html#opal">Opal</a> is a web application that is accessible through web services. It is implementing the DataSHIELD method thanks to the following built-in features:</p>
<ul>
<li>integration with a R server, where the DataSHIELD operations will take place,</li>
<li>secure data management, with fine-grain permissions (to restrict access to individual level data),</li>
<li>web services API, that allows to run Opal operations from a R script.</li>
</ul>
<p>In addition to these features, Opal manages the DataSHIELD configuration which consists of declaring the set of the allowed aggregation/assignment R functions and some R options.</p>
<div id="client" class="section level3" number="5.3.1">
<h3><span class="header-section-number">5.3.1</span> Client</h3>
<p>The <a href="https://cran.r-project.org/package=opalr">opalr</a> R package is a general purpose Opal connection R library (authentication is required) that is used to perform various operations (authorization may be required). The <a href="https://cran.r-project.org/package=DSOpal">DSOpal</a> R package is an implementation of the <em>DSI</em>, built on top of <em>opalr</em>. All the DataSHIELD operations are transparently applied to one or more Opal server using the <a href="datashield.html#higher-level-functions">DSI higher-level functions</a>. Opal also supports asynchronous function calls (submission of a R operation, then later retrieval of the result) which allows to operate on several DataSHIELD analysis nodes in parallel.</p>
</div>
<div id="server" class="section level3" number="5.3.2">
<h3><span class="header-section-number">5.3.2</span> Server</h3>
<p>On the Opal’s R server side, some DataSHIELD-compliant R packages can be managed using the Opal’s web interface: installation, removal of DataSHIELD-compliant R packages and automatic DataSHIELD configuration discovery. Opal guarantees that only the allowed functions can be called. The DataSHIELD-compliant R package guarantees that only aggregated results are returned to the client. The term ‘aggregated’ here means that the data in the R server will go through a function that summarizes individual-level data into a non-disclosive form. For example, obtaining the length of a vector, or obtaining the summary statistics of a vector (min, max, mean, etc.).</p>
<p>These DataSHIELD functions that are customisable. That is, administrators of the Opal server can add, remove, modify or create completely custom aggregating methods that are proposed to DataSHIELD clients.</p>
<div class="figure" style="text-align: center"><span id="fig:opalDSAdmin"></span>
<p class="caption">
Figure 5.2: DataSHIELD configuration in Opal
</p>
<img src="fig/opal_datashield_admin.png" alt="DataSHIELD configuration in Opal" width="90%" />
</div>
<p>When performing a DataSHIELD analysis session, a typical workflow on a single Opal analysis node is the following:</p>
<ul>
<li>authentication of the user (autorization to use DataSHIELD service is required),</li>
<li>creation and initialization of a R server session,</li>
<li>assignment of Opal-managed data into the R server session (data access authorization is required),</li>
<li>processing of the incoming R operation requests (aggregation and assignment function calls authorization is required) that are forwarded to the R server session; non-disclosive aggregated result is then returned to the R client.</li>
</ul>
</div>
</div>
<div id="demo" class="section level2" number="5.4">
<h2><span class="header-section-number">5.4</span> Demo</h2>
<p>Readers can read <a href="http://opaldoc.obiba.org/en/latest/r-user-guide/datashield.html">the DataSHIELD page in Opal documentation</a> to have a global overview about how to use DataSHIELD functions. It describes how to perform basic statistical analyses, linear and generalized linear models and some data visualization. A complete description of how DataSHIELD works, with lots of materials, examples, courses and real data analyses can be obtained in the <a href="https://data2knowledge.atlassian.net/wiki/spaces/DSDEV/overview">DataSHIELD Wiki</a>.</p>
<p>The following is a simple illustration of how to analyze some data available in the <a href="http://opal-demo.obiba.org/">Opal demo site</a>. The <a href="https://opal-demo.obiba.org/ui/index.html#!projects">Projects page</a> gives access to the different projects avaialble in this Opal server. If we select the <code>SURVIVAL</code> project we see that there are three tables:</p>
<div class="figure" style="text-align: center"><span id="fig:survivalTables"></span>
<p class="caption">
Figure 5.3: Tables available in the SURVIVAL project from our Opal example (<a href="https://opal-demo.obiba.org/" class="uri">https://opal-demo.obiba.org/</a>)
</p>
<img src="fig/survival_tables.png" alt="Tables available in the SURVIVAL project from our Opal example (https://opal-demo.obiba.org/)" width="940" />
</div>
<p>First we build a connection object with user credentials and the location of the server for each of the study:</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="datashield.html#cb55-1"></a><span class="kw">library</span>(DSOpal)</span>
<span id="cb55-2"><a href="datashield.html#cb55-2"></a>builder &lt;-<span class="st"> </span><span class="kw">newDSLoginBuilder</span>()</span>
<span id="cb55-3"><a href="datashield.html#cb55-3"></a>builder<span class="op">$</span><span class="kw">append</span>(<span class="dt">server =</span> <span class="st">&quot;study1&quot;</span>, <span class="dt">url =</span> <span class="st">&quot;https://opal-demo.obiba.org&quot;</span>, </span>
<span id="cb55-4"><a href="datashield.html#cb55-4"></a>               <span class="dt">user =</span> <span class="st">&quot;administrator&quot;</span>, <span class="dt">password =</span> <span class="st">&quot;password&quot;</span>)</span>
<span id="cb55-5"><a href="datashield.html#cb55-5"></a>builder<span class="op">$</span><span class="kw">append</span>(<span class="dt">server =</span> <span class="st">&quot;study2&quot;</span>, <span class="dt">url =</span> <span class="st">&quot;https://opal-demo.obiba.org&quot;</span>, </span>
<span id="cb55-6"><a href="datashield.html#cb55-6"></a>               <span class="dt">user =</span> <span class="st">&quot;administrator&quot;</span>, <span class="dt">password =</span> <span class="st">&quot;password&quot;</span>)</span>
<span id="cb55-7"><a href="datashield.html#cb55-7"></a>builder<span class="op">$</span><span class="kw">append</span>(<span class="dt">server =</span> <span class="st">&quot;study3&quot;</span>, <span class="dt">url =</span> <span class="st">&quot;https://opal-demo.obiba.org&quot;</span>, </span>
<span id="cb55-8"><a href="datashield.html#cb55-8"></a>               <span class="dt">user =</span> <span class="st">&quot;administrator&quot;</span>, <span class="dt">password =</span> <span class="st">&quot;password&quot;</span>)</span>
<span id="cb55-9"><a href="datashield.html#cb55-9"></a>logindata &lt;-<span class="st"> </span>builder<span class="op">$</span><span class="kw">build</span>()</span>
<span id="cb55-10"><a href="datashield.html#cb55-10"></a>logindata</span></code></pre></div>
<pre><code>  server                         url table resource     driver          user password token options
1 study1 https://opal-demo.obiba.org                OpalDriver administrator password              
2 study2 https://opal-demo.obiba.org                OpalDriver administrator password              
3 study3 https://opal-demo.obiba.org                OpalDriver administrator password              </code></pre>
<p>Then we perform login into each of the analysis servers and assign for each of them a different table with the same R symbol name:</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="datashield.html#cb57-1"></a>conns &lt;-<span class="st"> </span><span class="kw">datashield.login</span>(logindata)</span>
<span id="cb57-2"><a href="datashield.html#cb57-2"></a><span class="kw">datashield.assign.table</span>(conns, <span class="dt">symbol =</span> <span class="st">&quot;D&quot;</span>, </span>
<span id="cb57-3"><a href="datashield.html#cb57-3"></a>                        <span class="dt">table =</span> <span class="kw">list</span>(<span class="dt">study1 =</span> <span class="st">&quot;SURVIVAL.EXPAND_WITH_MISSING1&quot;</span>,</span>
<span id="cb57-4"><a href="datashield.html#cb57-4"></a>                                     <span class="dt">study2 =</span> <span class="st">&quot;SURVIVAL.EXPAND_WITH_MISSING2&quot;</span>,</span>
<span id="cb57-5"><a href="datashield.html#cb57-5"></a>                                     <span class="dt">study3 =</span> <span class="st">&quot;SURVIVAL.EXPAND_WITH_MISSING3&quot;</span>))</span>
<span id="cb57-6"><a href="datashield.html#cb57-6"></a><span class="kw">datashield.symbols</span>(conns)</span></code></pre></div>
<pre><code>$study1
[1] &quot;D&quot;

$study2
[1] &quot;D&quot;

$study3
[1] &quot;D&quot;</code></pre>
<p>Then we use the DataSHIELD functions from the <code>dsBaseClient</code> package to analyse the data and get non-disclosive summary statistics. The package should be first loaded</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="datashield.html#cb59-1"></a><span class="kw">library</span>(dsBaseClient)</span></code></pre></div>
<p>Then, for example, we can get the column names and the dimension of the dataset from each study:</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="datashield.html#cb60-1"></a><span class="kw">ds.colnames</span>(<span class="st">&quot;D&quot;</span>)</span></code></pre></div>
<pre><code>$study1
 [1] &quot;id&quot;        &quot;study.id&quot;  &quot;time.id&quot;   &quot;starttime&quot; &quot;endtime&quot;   &quot;survtime&quot;  &quot;cens&quot;      &quot;age.60&quot;    &quot;female&quot;    &quot;noise.56&quot;  &quot;pm10.16&quot;  
[12] &quot;bmi.26&quot;   

$study2
 [1] &quot;id&quot;        &quot;study.id&quot;  &quot;time.id&quot;   &quot;starttime&quot; &quot;endtime&quot;   &quot;survtime&quot;  &quot;cens&quot;      &quot;age.60&quot;    &quot;female&quot;    &quot;noise.56&quot;  &quot;pm10.16&quot;  
[12] &quot;bmi.26&quot;   

$study3
 [1] &quot;id&quot;        &quot;study.id&quot;  &quot;time.id&quot;   &quot;starttime&quot; &quot;endtime&quot;   &quot;survtime&quot;  &quot;cens&quot;      &quot;age.60&quot;    &quot;female&quot;    &quot;noise.56&quot;  &quot;pm10.16&quot;  
[12] &quot;bmi.26&quot;   </code></pre>
<p>we can get the dimension of the datasetes:</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="datashield.html#cb62-1"></a><span class="kw">ds.dim</span>(<span class="st">&quot;D&quot;</span>)</span></code></pre></div>
<pre><code>$`dimensions of D in study1`
[1] 2060   12

$`dimensions of D in study2`
[1] 1640   12

$`dimensions of D in study3`
[1] 2688   12

$`dimensions of D in combined studies`
[1] 6388   12</code></pre>
<p>the summary of a variable which includes its class, length, quantiles and mean:</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="datashield.html#cb64-1"></a><span class="kw">ds.summary</span>(<span class="st">&quot;D$age.60&quot;</span>)</span></code></pre></div>
<pre><code>$study1
$study1$class
[1] &quot;numeric&quot;

$study1$length
[1] 2060

$study1$`quantiles &amp; mean`
        5%        10%        25%        50%        75%        90%        95%       Mean 
-27.000000 -21.000000 -13.000000  -3.000000   7.000000  15.000000  20.000000  -3.167883 


$study2
$study2$class
[1] &quot;numeric&quot;

$study2$length
[1] 1640

$study2$`quantiles &amp; mean`
        5%        10%        25%        50%        75%        90%        95%       Mean 
-27.000000 -22.000000 -15.000000  -4.000000   7.000000  14.000000  20.000000  -4.008637 


$study3
$study3$class
[1] &quot;numeric&quot;

$study3$length
[1] 2688

$study3$`quantiles &amp; mean`
        5%        10%        25%        50%        75%        90%        95%       Mean 
-25.700000 -21.000000 -13.000000  -3.000000   7.000000  14.000000  20.000000  -3.089989 </code></pre>
<p>we can get contigency table between categorical variables:</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="datashield.html#cb66-1"></a><span class="kw">ds.table</span>(<span class="st">&quot;D$female&quot;</span>, <span class="st">&quot;D$cens&quot;</span>)</span></code></pre></div>
<pre><code>
 Data in all studies were valid 

Study 1 :  No errors reported from this study
Study 2 :  No errors reported from this study
Study 3 :  No errors reported from this study</code></pre>
<pre><code>$output.list
$output.list$TABLE.STUDY.1_row.props
        D$cens
D$female     0     1      NA
      0  0.726 0.261 0.01280
      1  0.833 0.158 0.00901
      NA 0.750 0.250 0.00000

$output.list$TABLE.STUDY.1_col.props
        D$cens
D$female       0       1    NA
      0  0.42000 0.57700 0.545
      1  0.57300 0.41400 0.455
      NA 0.00743 0.00946 0.000

$output.list$TABLE.STUDY.2_row.props
        D$cens
D$female     0     1      NA
      0  0.759 0.239 0.00286
      1  0.860 0.139 0.00107
      NA 0.800 0.200 0.00000

$output.list$TABLE.STUDY.2_col.props
        D$cens
D$female       0       1    NA
      0  0.39700 0.56000 0.667
      1  0.60000 0.43600 0.333
      NA 0.00299 0.00336 0.000

$output.list$TABLE.STUDY.3_row.props
        D$cens
D$female     0     1      NA
      0  0.714 0.274 0.01190
      1  0.828 0.164 0.00802
      NA 0.706 0.294 0.00000

$output.list$TABLE.STUDY.3_col.props
        D$cens
D$female       0       1    NA
      0  0.40100 0.56300 0.538
      1  0.59300 0.42800 0.462
      NA 0.00574 0.00874 0.000

$output.list$TABLES.COMBINED_all.sources_row.props
        D$cens
D$female     0     1      NA
      0  0.729 0.261 0.00997
      1  0.838 0.155 0.00650
      NA 0.737 0.263 0.00000

$output.list$TABLES.COMBINED_all.sources_col.props
        D$cens
D$female       0       1    NA
      0  0.40600 0.56700 0.549
      1  0.58800 0.42500 0.451
      NA 0.00555 0.00773 0.000

$output.list$TABLE_STUDY.1_counts
        D$cens
D$female   0   1 NA
      0  678 244 12
      1  925 175 10
      NA  12   4  0

$output.list$TABLE_STUDY.2_counts
        D$cens
D$female   0   1 NA
      0  531 167  2
      1  804 130  1
      NA   4   1  0

$output.list$TABLE_STUDY.3_counts
        D$cens
D$female    0   1 NA
      0   839 322 14
      1  1239 245 12
      NA   12   5  0

$output.list$TABLES.COMBINED_all.sources_counts
        D$cens
D$female    0   1 NA
      0  2048 733 28
      1  2968 550 23
      NA   28  10  0


$validity.message
[1] &quot;Data in all studies were valid&quot;</code></pre>
<p>or we can fit a generalized linear model among many other analyses:</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="datashield.html#cb69-1"></a><span class="kw">ds.glm</span>(<span class="dt">formula=</span><span class="st">&quot;cens~female+bmi.26+pm10.16&quot;</span>, <span class="dt">data=</span><span class="st">&quot;D&quot;</span>, <span class="dt">family=</span><span class="st">&quot;binomial&quot;</span>)</span></code></pre></div>
<pre><code>$Nvalid
[1] 6169

$Nmissing
[1] 219

$Ntotal
[1] 6388

$disclosure.risk
       RISK OF DISCLOSURE
study1                  0
study2                  0
study3                  0

$errorMessage
       ERROR MESSAGES
study1 &quot;No errors&quot;   
study2 &quot;No errors&quot;   
study3 &quot;No errors&quot;   

$nsubs
[1] 6169

$iter
[1] 6

$family

Family: binomial 
Link function: logit 


$formula
[1] &quot;cens ~ female + bmi.26 + pm10.16&quot;

$coefficients
              Estimate  Std. Error   z-value      p-value low0.95CI.LP high0.95CI.LP      P_OR low0.95CI.P_OR high0.95CI.P_OR
(Intercept) -0.9465551 0.046427292 -20.38790 2.141278e-92   -1.0375510    -0.8555593 0.2795781      0.2616228       0.2982680
female1     -0.7240154 0.067603806 -10.70968 9.167462e-27   -0.8565164    -0.5915143 0.4848017      0.4246388       0.5534885
bmi.26       0.1239566 0.007308207  16.96129 1.588245e-64    0.1096328     0.1382804 1.1319667      1.1158682       1.1482975
pm10.16      0.6620624 0.037059614  17.86479 2.217477e-71    0.5894269     0.7346979 1.9387867      1.8029548       2.0848521

$dev
[1] 5544.853

$df
[1] 6165

$output.information
[1] &quot;SEE TOP OF OUTPUT FOR INFORMATION ON MISSING DATA AND ERROR MESSAGES&quot;</code></pre>
<p>For a full list of the dsBaseClient functions you can use the following command:</p>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="datashield.html#cb71-1"></a><span class="kw">ds.listClientsideFunctions</span>()</span></code></pre></div>
<pre><code>
### Full search path 
 [1] &quot;.GlobalEnv&quot;                                &quot;package:dsBaseClient&quot;                      &quot;package:DSOpal&quot;                           
 [4] &quot;package:DSI&quot;                               &quot;package:R6&quot;                                &quot;package:progress&quot;                         
 [7] &quot;package:opalr&quot;                             &quot;package:httr&quot;                              &quot;package:airway&quot;                           
[10] &quot;package:SummarizedExperiment&quot;              &quot;package:DelayedArray&quot;                      &quot;package:matrixStats&quot;                      
[13] &quot;package:Homo.sapiens&quot;                      &quot;package:TxDb.Hsapiens.UCSC.hg19.knownGene&quot; &quot;package:org.Hs.eg.db&quot;                     
[16] &quot;package:GO.db&quot;                             &quot;package:OrganismDbi&quot;                       &quot;package:GenomicFeatures&quot;                  
[19] &quot;package:GenomicRanges&quot;                     &quot;package:GenomeInfoDb&quot;                      &quot;package:AnnotationDbi&quot;                    
[22] &quot;package:IRanges&quot;                           &quot;package:S4Vectors&quot;                         &quot;package:stats4&quot;                           
[25] &quot;package:tweeDEseqCountData&quot;                &quot;package:Biobase&quot;                           &quot;package:BiocGenerics&quot;                     
[28] &quot;package:parallel&quot;                          &quot;package:snpStats&quot;                          &quot;package:Matrix&quot;                           
[31] &quot;package:survival&quot;                          &quot;package:forcats&quot;                           &quot;package:stringr&quot;                          
[34] &quot;package:dplyr&quot;                             &quot;package:purrr&quot;                             &quot;package:readr&quot;                            
[37] &quot;package:tidyr&quot;                             &quot;package:tibble&quot;                            &quot;package:ggplot2&quot;                          
[40] &quot;package:tidyverse&quot;                         &quot;package:kableExtra&quot;                        &quot;package:BiocStyle&quot;                        
[43] &quot;tools:rstudio&quot;                             &quot;package:stats&quot;                             &quot;package:graphics&quot;                         
[46] &quot;package:grDevices&quot;                         &quot;package:utils&quot;                             &quot;package:datasets&quot;                         
[49] &quot;package:methods&quot;                           &quot;Autoloads&quot;                                 &quot;package:base&quot;                             

### userDefinedClient functions 
[1] &quot;No clientside functions in this repository&quot;

### dsBaseClient functions 
 [1] &quot;ds.asCharacter&quot;             &quot;ds.asDataMatrix&quot;            &quot;ds.asFactor&quot;                &quot;ds.asInteger&quot;              
 [5] &quot;ds.asList&quot;                  &quot;ds.asLogical&quot;               &quot;ds.asMatrix&quot;                &quot;ds.asNumeric&quot;              
 [9] &quot;ds.assign&quot;                  &quot;ds.Boole&quot;                   &quot;ds.c&quot;                       &quot;ds.cbind&quot;                  
[13] &quot;ds.changeRefGroup&quot;          &quot;ds.class&quot;                   &quot;ds.colnames&quot;                &quot;ds.completeCases&quot;          
[17] &quot;ds.contourPlot&quot;             &quot;ds.cor&quot;                     &quot;ds.corTest&quot;                 &quot;ds.cov&quot;                    
[21] &quot;ds.dataFrame&quot;               &quot;ds.dataFrameFill&quot;           &quot;ds.dataFrameSort&quot;           &quot;ds.dataFrameSubset&quot;        
[25] &quot;ds.densityGrid&quot;             &quot;ds.dim&quot;                     &quot;ds.exists&quot;                  &quot;ds.exp&quot;                    
[29] &quot;ds.glm&quot;                     &quot;ds.glmerSLMA&quot;               &quot;ds.glmSLMA&quot;                 &quot;ds.heatmapPlot&quot;            
[33] &quot;ds.histogram&quot;               &quot;ds.isNA&quot;                    &quot;ds.isValid&quot;                 &quot;ds.length&quot;                 
[37] &quot;ds.levels&quot;                  &quot;ds.lexis&quot;                   &quot;ds.list&quot;                    &quot;ds.listClientsideFunctions&quot;
[41] &quot;ds.listDisclosureSettings&quot;  &quot;ds.listOpals&quot;               &quot;ds.listServersideFunctions&quot; &quot;ds.lmerSLMA&quot;               
[45] &quot;ds.log&quot;                     &quot;ds.look&quot;                    &quot;ds.ls&quot;                      &quot;ds.make&quot;                   
[49] &quot;ds.matrix&quot;                  &quot;ds.matrixDet&quot;               &quot;ds.matrixDet.report&quot;        &quot;ds.matrixDiag&quot;             
[53] &quot;ds.matrixDimnames&quot;          &quot;ds.matrixInvert&quot;            &quot;ds.matrixMult&quot;              &quot;ds.matrixTranspose&quot;        
[57] &quot;ds.mean&quot;                    &quot;ds.meanByClass&quot;             &quot;ds.meanSdGp&quot;                &quot;ds.merge&quot;                  
[61] &quot;ds.message&quot;                 &quot;ds.names&quot;                   &quot;ds.numNA&quot;                   &quot;ds.quantileMean&quot;           
[65] &quot;ds.rbind&quot;                   &quot;ds.rBinom&quot;                  &quot;ds.recodeLevels&quot;            &quot;ds.recodeValues&quot;           
[69] &quot;ds.rep&quot;                     &quot;ds.replaceNA&quot;               &quot;ds.reShape&quot;                 &quot;ds.rm&quot;                     
[73] &quot;ds.rNorm&quot;                   &quot;ds.rowColCalc&quot;              &quot;ds.rPois&quot;                   &quot;ds.rUnif&quot;                  
[77] &quot;ds.sample&quot;                  &quot;ds.scatterPlot&quot;             &quot;ds.seq&quot;                     &quot;ds.setDefaultOpals&quot;        
[81] &quot;ds.setSeed&quot;                 &quot;ds.subset&quot;                  &quot;ds.subsetByClass&quot;           &quot;ds.summary&quot;                
[85] &quot;ds.table&quot;                   &quot;ds.table1D&quot;                 &quot;ds.table2D&quot;                 &quot;ds.tapply&quot;                 
[89] &quot;ds.tapply.assign&quot;           &quot;ds.testObjExists&quot;           &quot;ds.unList&quot;                  &quot;ds.var&quot;                    
[93] &quot;ds.vectorCalc&quot;             

If you cannot see one or more of the clientside functions you expected to find please see above for the full search path. If one of the paths is a possible clientside repository issue the R command ls(pos=&#39;package:dsPackageName&#39;) where &#39;package:dsPackageName&#39; is the full name stated in the search path</code></pre>
<pre><code> [1] &quot;No clientside functions in this repository&quot; &quot;ds.asCharacter&quot;                            
 [3] &quot;ds.asDataMatrix&quot;                            &quot;ds.asFactor&quot;                               
 [5] &quot;ds.asInteger&quot;                               &quot;ds.asList&quot;                                 
 [7] &quot;ds.asLogical&quot;                               &quot;ds.asMatrix&quot;                               
 [9] &quot;ds.asNumeric&quot;                               &quot;ds.assign&quot;                                 
[11] &quot;ds.Boole&quot;                                   &quot;ds.c&quot;                                      
[13] &quot;ds.cbind&quot;                                   &quot;ds.changeRefGroup&quot;                         
[15] &quot;ds.class&quot;                                   &quot;ds.colnames&quot;                               
[17] &quot;ds.completeCases&quot;                           &quot;ds.contourPlot&quot;                            
[19] &quot;ds.cor&quot;                                     &quot;ds.corTest&quot;                                
[21] &quot;ds.cov&quot;                                     &quot;ds.dataFrame&quot;                              
[23] &quot;ds.dataFrameFill&quot;                           &quot;ds.dataFrameSort&quot;                          
[25] &quot;ds.dataFrameSubset&quot;                         &quot;ds.densityGrid&quot;                            
[27] &quot;ds.dim&quot;                                     &quot;ds.exists&quot;                                 
[29] &quot;ds.exp&quot;                                     &quot;ds.glm&quot;                                    
[31] &quot;ds.glmerSLMA&quot;                               &quot;ds.glmSLMA&quot;                                
[33] &quot;ds.heatmapPlot&quot;                             &quot;ds.histogram&quot;                              
[35] &quot;ds.isNA&quot;                                    &quot;ds.isValid&quot;                                
[37] &quot;ds.length&quot;                                  &quot;ds.levels&quot;                                 
[39] &quot;ds.lexis&quot;                                   &quot;ds.list&quot;                                   
[41] &quot;ds.listClientsideFunctions&quot;                 &quot;ds.listDisclosureSettings&quot;                 
[43] &quot;ds.listOpals&quot;                               &quot;ds.listServersideFunctions&quot;                
[45] &quot;ds.lmerSLMA&quot;                                &quot;ds.log&quot;                                    
[47] &quot;ds.look&quot;                                    &quot;ds.ls&quot;                                     
[49] &quot;ds.make&quot;                                    &quot;ds.matrix&quot;                                 
[51] &quot;ds.matrixDet&quot;                               &quot;ds.matrixDet.report&quot;                       
[53] &quot;ds.matrixDiag&quot;                              &quot;ds.matrixDimnames&quot;                         
[55] &quot;ds.matrixInvert&quot;                            &quot;ds.matrixMult&quot;                             
[57] &quot;ds.matrixTranspose&quot;                         &quot;ds.mean&quot;                                   
[59] &quot;ds.meanByClass&quot;                             &quot;ds.meanSdGp&quot;                               
[61] &quot;ds.merge&quot;                                   &quot;ds.message&quot;                                
[63] &quot;ds.names&quot;                                   &quot;ds.numNA&quot;                                  
[65] &quot;ds.quantileMean&quot;                            &quot;ds.rbind&quot;                                  
[67] &quot;ds.rBinom&quot;                                  &quot;ds.recodeLevels&quot;                           
[69] &quot;ds.recodeValues&quot;                            &quot;ds.rep&quot;                                    
[71] &quot;ds.replaceNA&quot;                               &quot;ds.reShape&quot;                                
[73] &quot;ds.rm&quot;                                      &quot;ds.rNorm&quot;                                  
[75] &quot;ds.rowColCalc&quot;                              &quot;ds.rPois&quot;                                  
[77] &quot;ds.rUnif&quot;                                   &quot;ds.sample&quot;                                 
[79] &quot;ds.scatterPlot&quot;                             &quot;ds.seq&quot;                                    
[81] &quot;ds.setDefaultOpals&quot;                         &quot;ds.setSeed&quot;                                
[83] &quot;ds.subset&quot;                                  &quot;ds.subsetByClass&quot;                          
[85] &quot;ds.summary&quot;                                 &quot;ds.table&quot;                                  
[87] &quot;ds.table1D&quot;                                 &quot;ds.table2D&quot;                                
[89] &quot;ds.tapply&quot;                                  &quot;ds.tapply.assign&quot;                          
[91] &quot;ds.testObjExists&quot;                           &quot;ds.unList&quot;                                 
[93] &quot;ds.var&quot;                                     &quot;ds.vectorCalc&quot;                             </code></pre>
<p>Finaly we clean up the R server sessions:</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="datashield.html#cb74-1"></a><span class="kw">datashield.logout</span>(conns)</span></code></pre></div>

</div>
</div>



            </section>

          </div>
        </div>
      </div>
<a href="opal.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="resources.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
