<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>14 Differential gene expression (DGE) analysis | Orchestrating privacy-protected non-disclosive big data analyses of data from different resources with R and DataSHIELD</title>
  <meta name="description" content="14 Differential gene expression (DGE) analysis | Orchestrating privacy-protected non-disclosive big data analyses of data from different resources with R and DataSHIELD" />
  <meta name="generator" content="bookdown 0.20 and GitBook 2.6.7" />

  <meta property="og:title" content="14 Differential gene expression (DGE) analysis | Orchestrating privacy-protected non-disclosive big data analyses of data from different resources with R and DataSHIELD" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="14 Differential gene expression (DGE) analysis | Orchestrating privacy-protected non-disclosive big data analyses of data from different resources with R and DataSHIELD" />
  
  
  



<meta name="date" content="2020-11-16" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="Omic.html"/>
<link rel="next" href="epigenome-wide-association-analysis-ewas.html"/>
<script src="libs/header-attrs-2.3/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="chapter" data-level="" data-path="citation.html"><a href="citation.html"><i class="fa fa-check"></i>Citation</a></li>
<li class="part"><span><b>I Preamble</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#what-you-will-learn"><i class="fa fa-check"></i><b>1.1</b> What you will learn</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#preliminaries"><i class="fa fa-check"></i><b>1.2</b> Preliminaries</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#acknowledgments"><i class="fa fa-check"></i><b>1.3</b> Acknowledgments</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="learnR.html"><a href="learnR.html"><i class="fa fa-check"></i><b>2</b> Learning R</a></li>
<li class="chapter" data-level="3" data-path="BioC.html"><a href="BioC.html"><i class="fa fa-check"></i><b>3</b> Bioconductor Data Structures</a>
<ul>
<li class="chapter" data-level="3.1" data-path="BioC.html"><a href="BioC.html#snp-array-data"><i class="fa fa-check"></i><b>3.1</b> SNP Array Data</a></li>
<li class="chapter" data-level="3.2" data-path="BioC.html"><a href="BioC.html#expression-sets"><i class="fa fa-check"></i><b>3.2</b> Expression Sets</a></li>
<li class="chapter" data-level="3.3" data-path="BioC.html"><a href="BioC.html#genomic-ranges"><i class="fa fa-check"></i><b>3.3</b> Genomic Ranges</a></li>
<li class="chapter" data-level="3.4" data-path="BioC.html"><a href="BioC.html#summarized-experiments"><i class="fa fa-check"></i><b>3.4</b> Summarized Experiments</a></li>
<li class="chapter" data-level="3.5" data-path="BioC.html"><a href="BioC.html#ranged-summarized-experiments"><i class="fa fa-check"></i><b>3.5</b> Ranged Summarized Experiments</a></li>
<li class="chapter" data-level="3.6" data-path="BioC.html"><a href="BioC.html#multi-data-set"><i class="fa fa-check"></i><b>3.6</b> Multi Data Set</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="opal.html"><a href="opal.html"><i class="fa fa-check"></i><b>4</b> Opal</a>
<ul>
<li class="chapter" data-level="4.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>4.1</b> Introduction</a></li>
<li class="chapter" data-level="4.2" data-path="opal.html"><a href="opal.html#data-management"><i class="fa fa-check"></i><b>4.2</b> Data Management</a></li>
<li class="chapter" data-level="4.3" data-path="opal.html"><a href="opal.html#security"><i class="fa fa-check"></i><b>4.3</b> Security</a></li>
<li class="chapter" data-level="4.4" data-path="opal.html"><a href="opal.html#r-integration"><i class="fa fa-check"></i><b>4.4</b> R Integration</a></li>
<li class="chapter" data-level="4.5" data-path="opal.html"><a href="opal.html#opal-demo-site"><i class="fa fa-check"></i><b>4.5</b> Opal demo site</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="datashield.html"><a href="datashield.html"><i class="fa fa-check"></i><b>5</b> DataSHIELD</a>
<ul>
<li class="chapter" data-level="5.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>5.1</b> Introduction</a></li>
<li class="chapter" data-level="5.2" data-path="datashield.html"><a href="datashield.html#datashield-r-interface-dsi"><i class="fa fa-check"></i><b>5.2</b> DataSHIELD R Interface (DSI)</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="datashield.html"><a href="datashield.html#class-structures"><i class="fa fa-check"></i><b>5.2.1</b> Class Structures</a></li>
<li class="chapter" data-level="5.2.2" data-path="datashield.html"><a href="datashield.html#higher-level-functions"><i class="fa fa-check"></i><b>5.2.2</b> Higher Level Functions</a></li>
<li class="chapter" data-level="5.2.3" data-path="datashield.html"><a href="datashield.html#options"><i class="fa fa-check"></i><b>5.2.3</b> Options</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="datashield.html"><a href="datashield.html#datashieldopal-implementation"><i class="fa fa-check"></i><b>5.3</b> DataSHIELD/Opal Implementation</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="datashield.html"><a href="datashield.html#client"><i class="fa fa-check"></i><b>5.3.1</b> Client</a></li>
<li class="chapter" data-level="5.3.2" data-path="datashield.html"><a href="datashield.html#server"><i class="fa fa-check"></i><b>5.3.2</b> Server</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="datashield.html"><a href="datashield.html#demo"><i class="fa fa-check"></i><b>5.4</b> Demo</a></li>
</ul></li>
<li class="part"><span><b>II Focus Topics</b></span></li>
<li class="chapter" data-level="6" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>6</b> The Resources</a>
<ul>
<li class="chapter" data-level="6.1" data-path="resources.html"><a href="resources.html#concept"><i class="fa fa-check"></i><b>6.1</b> Concept</a></li>
<li class="chapter" data-level="6.2" data-path="resources.html"><a href="resources.html#types"><i class="fa fa-check"></i><b>6.2</b> Types</a></li>
<li class="chapter" data-level="6.3" data-path="resources.html"><a href="resources.html#definition"><i class="fa fa-check"></i><b>6.3</b> Definition</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="resourcer.html"><a href="resourcer.html"><i class="fa fa-check"></i><b>7</b> The resourcer Package</a>
<ul>
<li class="chapter" data-level="7.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>7.1</b> Introduction</a></li>
<li class="chapter" data-level="7.2" data-path="resourcer.html"><a href="resourcer.html#file-resources"><i class="fa fa-check"></i><b>7.2</b> File Resources</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="resourcer.html"><a href="resourcer.html#file-getter"><i class="fa fa-check"></i><b>7.2.1</b> File Getter</a></li>
<li class="chapter" data-level="7.2.2" data-path="resourcer.html"><a href="resourcer.html#file-data-format"><i class="fa fa-check"></i><b>7.2.2</b> File Data Format</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="resourcer.html"><a href="resourcer.html#database-resources"><i class="fa fa-check"></i><b>7.3</b> Database Resources</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="resourcer.html"><a href="resourcer.html#dbi-connectors"><i class="fa fa-check"></i><b>7.3.1</b> DBI Connectors</a></li>
<li class="chapter" data-level="7.3.2" data-path="resourcer.html"><a href="resourcer.html#use-dplyr"><i class="fa fa-check"></i><b>7.3.2</b> Use dplyr</a></li>
<li class="chapter" data-level="7.3.3" data-path="resourcer.html"><a href="resourcer.html#document-databases"><i class="fa fa-check"></i><b>7.3.3</b> Document Databases</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="resourcer.html"><a href="resourcer.html#computation-resources"><i class="fa fa-check"></i><b>7.4</b> Computation Resources</a></li>
<li class="chapter" data-level="7.5" data-path="resourcer.html"><a href="resourcer.html#interacting-with-r-resources"><i class="fa fa-check"></i><b>7.5</b> Interacting with R Resources</a></li>
<li class="chapter" data-level="7.6" data-path="resourcer.html"><a href="resourcer.html#extending_resources"><i class="fa fa-check"></i><b>7.6</b> Extending Resources</a></li>
<li class="chapter" data-level="7.7" data-path="resourcer.html"><a href="resourcer.html#resource-forms"><i class="fa fa-check"></i><b>7.7</b> Resource Forms</a></li>
<li class="chapter" data-level="7.8" data-path="resourcer.html"><a href="resourcer.html#examples-of-resources"><i class="fa fa-check"></i><b>7.8</b> Examples of Resources</a></li>
<li class="chapter" data-level="7.9" data-path="resourcer.html"><a href="resourcer.html#using-resources-with-datashield"><i class="fa fa-check"></i><b>7.9</b> Using Resources with DataSHIELD</a>
<ul>
<li class="chapter" data-level="7.9.1" data-path="resourcer.html"><a href="resourcer.html#tsv-file-into-a-tibble-or-data.frame"><i class="fa fa-check"></i><b>7.9.1</b> TSV File into a tibble or data.frame</a></li>
<li class="chapter" data-level="7.9.2" data-path="resourcer.html"><a href="resourcer.html#r-data-file-into-a-r-object"><i class="fa fa-check"></i><b>7.9.2</b> R Data File into a R object</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>III Resources Extensions</b></span></li>
<li class="chapter" data-level="8" data-path="omic-extension.html"><a href="omic-extension.html"><i class="fa fa-check"></i><b>8</b> Extension to VCF files to peform GWAS with Bioconductor</a></li>
<li class="chapter" data-level="9" data-path="shell-extension.html"><a href="shell-extension.html"><i class="fa fa-check"></i><b>9</b> Extension to secure shell programs: GWAS with PLINK</a></li>
<li class="chapter" data-level="10" data-path="extension-to-ga4gh-and-ega-bamvcf-files.html"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html"><i class="fa fa-check"></i><b>10</b> Extension to GA4GH and EGA (BAM/VCF files)</a>
<ul>
<li class="chapter" data-level="10.1" data-path="extension-to-ga4gh-and-ega-bamvcf-files.html"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#upload-ga4gh-and-ega-resources-into-opal"><i class="fa fa-check"></i><b>10.1</b> Upload GA4GH and EGA resources into OPAL</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="extension-to-ga4gh-and-ega-bamvcf-files.html"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#manually"><i class="fa fa-check"></i><b>10.1.1</b> Manually</a></li>
<li class="chapter" data-level="10.1.2" data-path="extension-to-ga4gh-and-ega-bamvcf-files.html"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#using-r"><i class="fa fa-check"></i><b>10.1.2</b> Using R</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="extension-to-ga4gh-and-ega-bamvcf-files.html"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#adding-a-genomic-resource"><i class="fa fa-check"></i><b>10.2</b> Adding a genomic resource</a></li>
<li class="chapter" data-level="10.3" data-path="extension-to-ga4gh-and-ega-bamvcf-files.html"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#descriptive-omic-data-analysis"><i class="fa fa-check"></i><b>10.3</b> Descriptive omic data analysis</a></li>
</ul></li>
<li class="part"><span><b>IV Workflows</b></span></li>
<li class="chapter" data-level="11" data-path="setup.html"><a href="setup.html"><i class="fa fa-check"></i><b>11</b> Setup</a>
<ul>
<li class="chapter" data-level="11.1" data-path="setup.html"><a href="setup.html#providing-datashield-packages-in-the-opal-server"><i class="fa fa-check"></i><b>11.1</b> Providing DataSHIELD packages in the opal server</a></li>
<li class="chapter" data-level="11.2" data-path="setup.html"><a href="setup.html#required-r-packages-in-the-client-site-e.g.-local-machine"><i class="fa fa-check"></i><b>11.2</b> Required R Packages in the client site (e.g.Â local machine)</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="basic-statistical-analyses.html"><a href="basic-statistical-analyses.html"><i class="fa fa-check"></i><b>12</b> Basic statistical analyses</a>
<ul>
<li class="chapter" data-level="12.1" data-path="basic-statistical-analyses.html"><a href="basic-statistical-analyses.html#analysis-from-a-single-study"><i class="fa fa-check"></i><b>12.1</b> Analysis from a single study</a></li>
<li class="chapter" data-level="12.2" data-path="basic-statistical-analyses.html"><a href="basic-statistical-analyses.html#analysis-from-a-multiple-studies"><i class="fa fa-check"></i><b>12.2</b> Analysis from a multiple studies</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="Omic.html"><a href="Omic.html"><i class="fa fa-check"></i><b>13</b> Omic data analysis: types of implemented analyses</a></li>
<li class="chapter" data-level="14" data-path="differential-gene-expression-dge-analysis.html"><a href="differential-gene-expression-dge-analysis.html"><i class="fa fa-check"></i><b>14</b> Differential gene expression (DGE) analysis</a></li>
<li class="chapter" data-level="15" data-path="epigenome-wide-association-analysis-ewas.html"><a href="epigenome-wide-association-analysis-ewas.html"><i class="fa fa-check"></i><b>15</b> Epigenome-wide association analysis (EWAS)</a>
<ul>
<li class="chapter" data-level="15.1" data-path="epigenome-wide-association-analysis-ewas.html"><a href="epigenome-wide-association-analysis-ewas.html#single-cpg-analysis"><i class="fa fa-check"></i><b>15.1</b> Single CpG analysis</a></li>
<li class="chapter" data-level="15.2" data-path="epigenome-wide-association-analysis-ewas.html"><a href="epigenome-wide-association-analysis-ewas.html#multiple-cpg-analysis"><i class="fa fa-check"></i><b>15.2</b> Multiple CpG analysis</a></li>
<li class="chapter" data-level="15.3" data-path="epigenome-wide-association-analysis-ewas.html"><a href="epigenome-wide-association-analysis-ewas.html#adjusting-for-surrogate-variables"><i class="fa fa-check"></i><b>15.3</b> Adjusting for Surrogate Variables</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="genome-wide-association-analyses-gwas.html"><a href="genome-wide-association-analyses-gwas.html"><i class="fa fa-check"></i><b>16</b> Genome-wide association analyses (GWAS)</a>
<ul>
<li class="chapter" data-level="16.1" data-path="genome-wide-association-analyses-gwas.html"><a href="genome-wide-association-analyses-gwas.html#gwasBioc"><i class="fa fa-check"></i><b>16.1</b> GWAS with Bioconductor</a></li>
<li class="chapter" data-level="16.2" data-path="genome-wide-association-analyses-gwas.html"><a href="genome-wide-association-analyses-gwas.html#gwas-with-plink"><i class="fa fa-check"></i><b>16.2</b> GWAS with PLINK</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="GIS.html"><a href="GIS.html"><i class="fa fa-check"></i><b>17</b> Geospatial data analysis</a>
<ul>
<li class="chapter" data-level="17.1" data-path="GIS.html"><a href="GIS.html#introducing-the-analysis"><i class="fa fa-check"></i><b>17.1</b> Introducing the analysis</a></li>
<li class="chapter" data-level="17.2" data-path="GIS.html"><a href="GIS.html#setting-up-the-analysis"><i class="fa fa-check"></i><b>17.2</b> Setting up the analysis</a></li>
<li class="chapter" data-level="17.3" data-path="GIS.html"><a href="GIS.html#data-manipulation"><i class="fa fa-check"></i><b>17.3</b> Data manipulation</a></li>
<li class="chapter" data-level="17.4" data-path="GIS.html"><a href="GIS.html#generating-the-final-result"><i class="fa fa-check"></i><b>17.4</b> Generating the final result</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="exposome.html"><a href="exposome.html"><i class="fa fa-check"></i><b>18</b> Exposome data analysis</a>
<ul>
<li class="chapter" data-level="18.1" data-path="exposome.html"><a href="exposome.html#loading-the-exposome-set"><i class="fa fa-check"></i><b>18.1</b> Loading the Exposome Set</a>
<ul>
<li class="chapter" data-level="18.1.1" data-path="exposome.html"><a href="exposome.html#from-tables-to-exposomeset"><i class="fa fa-check"></i><b>18.1.1</b> From tables to <code>ExposomeSet</code></a></li>
<li class="chapter" data-level="18.1.2" data-path="exposome.html"><a href="exposome.html#from-resource-to-exposomeset"><i class="fa fa-check"></i><b>18.1.2</b> From resource to <code>ExposomeSet</code></a></li>
</ul></li>
<li class="chapter" data-level="18.2" data-path="exposome.html"><a href="exposome.html#exploring-the-loaded-dataset"><i class="fa fa-check"></i><b>18.2</b> Exploring the loaded dataset</a>
<ul>
<li class="chapter" data-level="18.2.1" data-path="exposome.html"><a href="exposome.html#family-names"><i class="fa fa-check"></i><b>18.2.1</b> Family names</a></li>
<li class="chapter" data-level="18.2.2" data-path="exposome.html"><a href="exposome.html#exposures-and-phenotypes-names"><i class="fa fa-check"></i><b>18.2.2</b> Exposures and phenotypes names</a></li>
<li class="chapter" data-level="18.2.3" data-path="exposome.html"><a href="exposome.html#summary-of-variables"><i class="fa fa-check"></i><b>18.2.3</b> Summary of variables</a></li>
<li class="chapter" data-level="18.2.4" data-path="exposome.html"><a href="exposome.html#missing-data"><i class="fa fa-check"></i><b>18.2.4</b> Missing data</a></li>
<li class="chapter" data-level="18.2.5" data-path="exposome.html"><a href="exposome.html#exposures-normality"><i class="fa fa-check"></i><b>18.2.5</b> Exposures Normality</a></li>
<li class="chapter" data-level="18.2.6" data-path="exposome.html"><a href="exposome.html#exposures-behaviour"><i class="fa fa-check"></i><b>18.2.6</b> Exposures Behaviour</a></li>
</ul></li>
<li class="chapter" data-level="18.3" data-path="exposome.html"><a href="exposome.html#exposures-imputation"><i class="fa fa-check"></i><b>18.3</b> Exposures Imputation</a></li>
<li class="chapter" data-level="18.4" data-path="exposome.html"><a href="exposome.html#performing-an-exwas"><i class="fa fa-check"></i><b>18.4</b> Performing an ExWAS</a></li>
<li class="chapter" data-level="18.5" data-path="exposome.html"><a href="exposome.html#exposures-pca"><i class="fa fa-check"></i><b>18.5</b> Exposures PCA</a></li>
<li class="chapter" data-level="18.6" data-path="exposome.html"><a href="exposome.html#exposures-correlation"><i class="fa fa-check"></i><b>18.6</b> Exposures Correlation</a></li>
</ul></li>
<li class="part"><span><b>V Developers</b></span></li>
<li class="chapter" data-level="19" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html"><i class="fa fa-check"></i><b>19</b> DSLite: DataSHIELD Implementation on Local Datasets</a>
<ul>
<li class="chapter" data-level="19.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#development-environment-setup"><i class="fa fa-check"></i><b>19.1</b> Development Environment Setup</a>
<ul>
<li class="chapter" data-level="19.1.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-packages"><i class="fa fa-check"></i><b>19.1.1</b> DataSHIELD Packages</a></li>
<li class="chapter" data-level="19.1.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#test-datasets"><i class="fa fa-check"></i><b>19.1.2</b> Test Datasets</a></li>
<li class="chapter" data-level="19.1.3" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#r-package-development-tools"><i class="fa fa-check"></i><b>19.1.3</b> R Package Development Tools</a></li>
</ul></li>
<li class="chapter" data-level="19.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-development-flow"><i class="fa fa-check"></i><b>19.2</b> DataSHIELD Development Flow</a>
<ul>
<li class="chapter" data-level="19.2.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#dsliteserver"><i class="fa fa-check"></i><b>19.2.1</b> DSLiteServer</a></li>
<li class="chapter" data-level="19.2.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-configuration"><i class="fa fa-check"></i><b>19.2.2</b> DataSHIELD Configuration</a></li>
</ul></li>
<li class="chapter" data-level="19.3" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-sessions"><i class="fa fa-check"></i><b>19.3</b> DataSHIELD Sessions</a></li>
<li class="chapter" data-level="19.4" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#debugging"><i class="fa fa-check"></i><b>19.4</b> Debugging</a></li>
<li class="chapter" data-level="19.5" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#limitations"><i class="fa fa-check"></i><b>19.5</b> Limitations</a>
<ul>
<li class="chapter" data-level="19.5.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#function-parameters-parser"><i class="fa fa-check"></i><b>19.5.1</b> Function Parameters Parser</a></li>
<li class="chapter" data-level="19.5.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#server-side-environments"><i class="fa fa-check"></i><b>19.5.2</b> Server Side Environments</a></li>
</ul></li>
<li class="chapter" data-level="19.6" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#environment-setup-for-datashield-package-development.-a-proposition"><i class="fa fa-check"></i><b>19.6</b> Environment setup for DataSHIELD package development. A proposition</a>
<ul>
<li class="chapter" data-level="19.6.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#fandfiles"><i class="fa fa-check"></i><b>19.6.1</b> Folders and files</a></li>
<li class="chapter" data-level="19.6.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#development-of-a-function"><i class="fa fa-check"></i><b>19.6.2</b> Development of a function</a></li>
<li class="chapter" data-level="19.6.3" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#testing-of-the-function"><i class="fa fa-check"></i><b>19.6.3</b> Testing of the function</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="20" data-path="package-info.html"><a href="package-info.html"><i class="fa fa-check"></i><b>20</b> Creating DataSHIELD packages</a></li>
<li class="chapter" data-level="21" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html"><i class="fa fa-check"></i><b>21</b> Tips and tricks</a>
<ul>
<li class="chapter" data-level="21.1" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#how-to-create-a-new-project-into-opal"><i class="fa fa-check"></i><b>21.1</b> How to create a new project into OPAL</a>
<ul>
<li class="chapter" data-level="21.1.1" data-path="extension-to-ga4gh-and-ega-bamvcf-files.html"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#manually"><i class="fa fa-check"></i><b>21.1.1</b> Manually</a></li>
<li class="chapter" data-level="21.1.2" data-path="extension-to-ga4gh-and-ega-bamvcf-files.html"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#using-r"><i class="fa fa-check"></i><b>21.1.2</b> Using R</a></li>
</ul></li>
<li class="chapter" data-level="21.2" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#how-to-upload-a-new-resource-into-opal"><i class="fa fa-check"></i><b>21.2</b> How to upload a new resource into OPAL</a>
<ul>
<li class="chapter" data-level="21.2.1" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#create_manually"><i class="fa fa-check"></i><b>21.2.1</b> Manually</a></li>
<li class="chapter" data-level="21.2.2" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#using-r-1"><i class="fa fa-check"></i><b>21.2.2</b> Using R</a></li>
</ul></li>
<li class="chapter" data-level="21.3" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#how-to-install-datashield-packages-into-opal-server"><i class="fa fa-check"></i><b>21.3</b> How to install DataSHIELD packages into OPAL server</a>
<ul>
<li class="chapter" data-level="21.3.1" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#manually-1"><i class="fa fa-check"></i><b>21.3.1</b> Manually</a></li>
<li class="chapter" data-level="21.3.2" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#using-r-2"><i class="fa fa-check"></i><b>21.3.2</b> Using R</a></li>
</ul></li>
<li class="chapter" data-level="21.4" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#how-to-install-r-packages-into-opal-server"><i class="fa fa-check"></i><b>21.4</b> How to install R packages into OPAL server</a>
<ul>
<li class="chapter" data-level="21.4.1" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#manually-2"><i class="fa fa-check"></i><b>21.4.1</b> Manually</a></li>
<li class="chapter" data-level="21.4.2" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#using-r-3"><i class="fa fa-check"></i><b>21.4.2</b> Using R</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="22" data-path="advancedtechnical-questions.html"><a href="advancedtechnical-questions.html"><i class="fa fa-check"></i><b>22</b> Advanced/technical questions</a></li>
<li class="part"><span><b>VI Appendix</b></span></li>
<li class="chapter" data-level="23" data-path="session-info.html"><a href="session-info.html"><i class="fa fa-check"></i><b>23</b> Session Info</a></li>
<li class="chapter" data-level="24" data-path="contributors.html"><a href="contributors.html"><i class="fa fa-check"></i><b>24</b> Contributors</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating privacy-protected non-disclosive big data analyses of data from different resources with R and DataSHIELD</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="differential-gene-expression-dge-analysis" class="section level1" number="14">
<h1><span class="header-section-number">14</span> Differential gene expression (DGE) analysis</h1>
<p>Let us illustrate how to perform transcriptomic data analysis using data from <a href="https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga">TCGA project</a>. We have uploaded to the opal server a resource called <code>tcga_liver</code> whose URL is <a href="http://duffel.rail.bio/recount/TCGA/rse_gene_liver.Rdata" class="uri">http://duffel.rail.bio/recount/TCGA/rse_gene_liver.Rdata</a> which is available through the <a href="https://jhubiostatistics.shinyapps.io/recount/">recount project</a>. This resource contains the <code>RangeSummarizedExperiment</code> with the RNAseq profiling of liver cancer data from TCGA. Next, we illustrate how a differential expression analysis to compare RNAseq profiling of women vs men (variable <code>gdc_cases.demographic.gender</code>). The DGE analysis is normally performed using <em><a href="https://bioconductor.org/packages/3.11/limma">limma</a></em> package. In that case, as we are analyzing RNA-seq data, <code>limma + voom</code> method will be required.</p>
<p>Let us start by creating the connection to the opal server:</p>
<div class="sourceCode" id="cb154"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb154-1"><a href="differential-gene-expression-dge-analysis.html#cb154-1"></a>builder &lt;-<span class="st"> </span><span class="kw">newDSLoginBuilder</span>()</span>
<span id="cb154-2"><a href="differential-gene-expression-dge-analysis.html#cb154-2"></a>builder<span class="op">$</span><span class="kw">append</span>(<span class="dt">server =</span> <span class="st">&quot;study1&quot;</span>, <span class="dt">url =</span> <span class="st">&quot;https://opal-demo.obiba.org&quot;</span>, </span>
<span id="cb154-3"><a href="differential-gene-expression-dge-analysis.html#cb154-3"></a>               <span class="dt">user =</span> <span class="st">&quot;dsuser&quot;</span>, <span class="dt">password =</span> <span class="st">&quot;password&quot;</span>, </span>
<span id="cb154-4"><a href="differential-gene-expression-dge-analysis.html#cb154-4"></a>               <span class="dt">resource =</span> <span class="st">&quot;RSRC.tcga_liver&quot;</span>, <span class="dt">driver =</span> <span class="st">&quot;OpalDriver&quot;</span>)</span>
<span id="cb154-5"><a href="differential-gene-expression-dge-analysis.html#cb154-5"></a></span>
<span id="cb154-6"><a href="differential-gene-expression-dge-analysis.html#cb154-6"></a>logindata &lt;-<span class="st"> </span>builder<span class="op">$</span><span class="kw">build</span>()</span>
<span id="cb154-7"><a href="differential-gene-expression-dge-analysis.html#cb154-7"></a></span>
<span id="cb154-8"><a href="differential-gene-expression-dge-analysis.html#cb154-8"></a>conns &lt;-<span class="st"> </span><span class="kw">datashield.login</span>(<span class="dt">logins =</span> logindata, <span class="dt">assign =</span> <span class="ot">TRUE</span>, </span>
<span id="cb154-9"><a href="differential-gene-expression-dge-analysis.html#cb154-9"></a>                          <span class="dt">symbol =</span> <span class="st">&quot;res&quot;</span>)</span></code></pre></div>
<p>Then, let us coerce the resource to a <code>RangedSummarizedExperiment</code> which is the type of object that is available in the <a href="https://jhubiostatistics.shinyapps.io/recount/">recount project</a>.</p>
<div class="sourceCode" id="cb155"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb155-1"><a href="differential-gene-expression-dge-analysis.html#cb155-1"></a><span class="kw">datashield.assign.expr</span>(conns, <span class="dt">symbol =</span> <span class="st">&quot;rse&quot;</span>, </span>
<span id="cb155-2"><a href="differential-gene-expression-dge-analysis.html#cb155-2"></a>                       <span class="dt">expr =</span> <span class="kw">quote</span>(<span class="kw">as.resource.object</span>(res)))</span>
<span id="cb155-3"><a href="differential-gene-expression-dge-analysis.html#cb155-3"></a><span class="kw">ds.class</span>(<span class="st">&quot;rse&quot;</span>)</span></code></pre></div>
<pre><code>$study1
[1] &quot;RangedSummarizedExperiment&quot;
attr(,&quot;package&quot;)
[1] &quot;SummarizedExperiment&quot;</code></pre>
<p>The number of features and samples can be inspected by</p>
<div class="sourceCode" id="cb157"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb157-1"><a href="differential-gene-expression-dge-analysis.html#cb157-1"></a><span class="kw">ds.dim</span>(<span class="st">&quot;rse&quot;</span>)</span></code></pre></div>
<pre><code>$`dimensions of rse in study1`
[1] 58037   424

$`dimensions of rse in combined studies`
[1] 58037   424</code></pre>
<p>And the names of the features using the same function used in the case of analyzing an <code>ExpressionSet</code></p>
<div class="sourceCode" id="cb159"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb159-1"><a href="differential-gene-expression-dge-analysis.html#cb159-1"></a>name.features &lt;-<span class="st"> </span><span class="kw">ds.featureNames</span>(<span class="st">&quot;rse&quot;</span>)</span>
<span id="cb159-2"><a href="differential-gene-expression-dge-analysis.html#cb159-2"></a><span class="kw">lapply</span>(name.features, head)</span></code></pre></div>
<pre><code>$study1
[1] &quot;ENSG00000000003.14&quot; &quot;ENSG00000000005.5&quot;  &quot;ENSG00000000419.12&quot;
[4] &quot;ENSG00000000457.13&quot; &quot;ENSG00000000460.16&quot; &quot;ENSG00000000938.12&quot;</code></pre>
<p>Also the covariate names can be inspected by</p>
<div class="sourceCode" id="cb161"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb161-1"><a href="differential-gene-expression-dge-analysis.html#cb161-1"></a>name.vars &lt;-<span class="st"> </span><span class="kw">ds.featureData</span>(<span class="st">&quot;rse&quot;</span>)</span>
<span id="cb161-2"><a href="differential-gene-expression-dge-analysis.html#cb161-2"></a><span class="kw">lapply</span>(name.vars, head, <span class="dt">n=</span><span class="dv">15</span>)</span></code></pre></div>
<pre><code>$study1
 [1] &quot;project&quot;                                       
 [2] &quot;sample&quot;                                        
 [3] &quot;experiment&quot;                                    
 [4] &quot;run&quot;                                           
 [5] &quot;read_count_as_reported_by_sra&quot;                 
 [6] &quot;reads_downloaded&quot;                              
 [7] &quot;proportion_of_reads_reported_by_sra_downloaded&quot;
 [8] &quot;paired_end&quot;                                    
 [9] &quot;sra_misreported_paired_end&quot;                    
[10] &quot;mapped_read_count&quot;                             
[11] &quot;auc&quot;                                           
[12] &quot;sharq_beta_tissue&quot;                             
[13] &quot;sharq_beta_cell_type&quot;                          
[14] &quot;biosample_submission_date&quot;                     
[15] &quot;biosample_publication_date&quot;                    </code></pre>
<p>We can visualize the levels of the variable having gender information</p>
<div class="sourceCode" id="cb163"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb163-1"><a href="differential-gene-expression-dge-analysis.html#cb163-1"></a><span class="kw">ds.table</span>(<span class="st">&quot;rse$gdc_cases.demographic.gender&quot;</span>)</span></code></pre></div>
<pre><code>
 Data in all studies were valid 

Study 1 :  No errors reported from this study</code></pre>
<pre><code>$output.list
$output.list$TABLE_rvar.by.study_row.props
                                study
rse$gdc_cases.demographic.gender 1
                          female 1
                          male   1

$output.list$TABLE_rvar.by.study_col.props
                                study
rse$gdc_cases.demographic.gender         1
                          female 0.3372642
                          male   0.6627358

$output.list$TABLE_rvar.by.study_counts
                                study
rse$gdc_cases.demographic.gender   1
                          female 143
                          male   281

$output.list$TABLES.COMBINED_all.sources_proportions
rse$gdc_cases.demographic.gender
female   male 
 0.337  0.663 

$output.list$TABLES.COMBINED_all.sources_counts
rse$gdc_cases.demographic.gender
female   male 
   143    281 


$validity.message
[1] &quot;Data in all studies were valid&quot;</code></pre>
<p>The differential expression analysis is then performed by:</p>
<div class="sourceCode" id="cb166"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb166-1"><a href="differential-gene-expression-dge-analysis.html#cb166-1"></a>ans.gender &lt;-<span class="st"> </span><span class="kw">ds.limma</span>(<span class="dt">model =</span>  <span class="op">~</span><span class="st"> </span>gdc_cases.demographic.gender, </span>
<span id="cb166-2"><a href="differential-gene-expression-dge-analysis.html#cb166-2"></a>                       <span class="dt">Set =</span> <span class="st">&quot;rse&quot;</span>, <span class="dt">type.data =</span> <span class="st">&quot;RNAseq&quot;</span>, </span>
<span id="cb166-3"><a href="differential-gene-expression-dge-analysis.html#cb166-3"></a>                       <span class="dt">sva =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<p>Notice that we have set <code>type.data='RNAseq'</code> to consider that our data are counts obtained from a RNA-seq experiment. By indicating so, the differential analysis is performed by using <code>voom</code> + <code>limma</code> as previously mention.</p>
<p>The top differentially expressed genes can be visualized by:</p>
<div class="sourceCode" id="cb167"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb167-1"><a href="differential-gene-expression-dge-analysis.html#cb167-1"></a>ans.gender</span></code></pre></div>
<pre><code>$study1
# A tibble: 58,037 x 10
   id     logFC   CI.L   CI.R AveExpr     t   P.Value adj.P.Val     B
   &lt;chr&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt; &lt;dbl&gt;
 1 ENSG~  10.9   10.5   11.3   -5.59   53.6 7.05e-191 4.09e-186  402.
 2 ENSG~  11.4   10.9   11.8   -4.35   49.4 5.93e-178 1.72e-173  376.
 3 ENSG~   9.63   9.24  10.0    0.856  47.9 5.05e-173 9.78e-169  366.
 4 ENSG~  10.2    9.76  10.6   -5.88   46.3 5.27e-168 7.65e-164  355.
 5 ENSG~  11.3   10.8   11.8   -0.226  44.7 1.29e-162 1.50e-158  344.
 6 ENSG~  11.4   10.9   11.9   -1.10   44.5 7.80e-162 7.55e-158  343.
 7 ENSG~  12.3   11.8   12.9   -0.882  44.4 1.62e-161 1.34e-157  342.
 8 ENSG~   8.66   8.27   9.05  -0.781  43.9 9.83e-160 7.13e-156  338.
 9 ENSG~  11.4   10.8   11.9    2.37   43.3 1.44e-157 9.29e-154  334.
10 ENSG~ -12.4  -12.9  -11.8   -7.71  -43.1 5.20e-157 3.02e-153  333.
# ... with 58,027 more rows, and 1 more variable: SE &lt;dbl&gt;

attr(,&quot;class&quot;)
[1] &quot;dsLimma&quot; &quot;list&quot;   </code></pre>
<p>We have also implemented two other functions <code>ds.DESeq2</code> and <code>ds.edgeR</code> that perform DGE analysis using <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">DESeq2</a> and <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf">edgeR</a> methods. This is the R code used to that purpose:</p>
<p><strong>To be supplied</strong></p>
<p>We close the DataSHIELD session by:</p>
<div class="sourceCode" id="cb169"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb169-1"><a href="differential-gene-expression-dge-analysis.html#cb169-1"></a><span class="kw">datashield.logout</span>(conns)</span></code></pre></div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="Omic.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="epigenome-wide-association-analysis-ewas.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
