<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>15 Epigenome-wide association analysis (EWAS) | Orchestrating non-disclosive big data analyses of data from different resources with R and DataSHIELD</title>
  <meta name="description" content="15 Epigenome-wide association analysis (EWAS) | Orchestrating non-disclosive big data analyses of data from different resources with R and DataSHIELD" />
  <meta name="generator" content="bookdown 0.20 and GitBook 2.6.7" />

  <meta property="og:title" content="15 Epigenome-wide association analysis (EWAS) | Orchestrating non-disclosive big data analyses of data from different resources with R and DataSHIELD" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="15 Epigenome-wide association analysis (EWAS) | Orchestrating non-disclosive big data analyses of data from different resources with R and DataSHIELD" />
  
  
  



<meta name="date" content="2020-11-14" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="differential-gene-expression-dge-analysis.html"/>
<link rel="next" href="genome-wide-association-analyses-gwas.html"/>
<script src="libs/header-attrs-2.3/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Preamble</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#what-you-will-learn"><i class="fa fa-check"></i><b>1.1</b> What you will learn</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#preliminaries"><i class="fa fa-check"></i><b>1.2</b> Preliminaries</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#acknowledgments"><i class="fa fa-check"></i><b>1.3</b> Acknowledgments</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="learnR.html"><a href="learnR.html"><i class="fa fa-check"></i><b>2</b> Learning R</a></li>
<li class="chapter" data-level="3" data-path="BioC.html"><a href="BioC.html"><i class="fa fa-check"></i><b>3</b> Bioconductor Data Structures</a>
<ul>
<li class="chapter" data-level="3.1" data-path="BioC.html"><a href="BioC.html#snp-array-data"><i class="fa fa-check"></i><b>3.1</b> SNP Array Data</a></li>
<li class="chapter" data-level="3.2" data-path="BioC.html"><a href="BioC.html#expression-sets"><i class="fa fa-check"></i><b>3.2</b> Expression Sets</a></li>
<li class="chapter" data-level="3.3" data-path="BioC.html"><a href="BioC.html#genomic-ranges"><i class="fa fa-check"></i><b>3.3</b> Genomic Ranges</a></li>
<li class="chapter" data-level="3.4" data-path="BioC.html"><a href="BioC.html#summarized-experiments"><i class="fa fa-check"></i><b>3.4</b> Summarized Experiments</a></li>
<li class="chapter" data-level="3.5" data-path="BioC.html"><a href="BioC.html#ranged-summarized-experiments"><i class="fa fa-check"></i><b>3.5</b> Ranged Summarized Experiments</a></li>
<li class="chapter" data-level="3.6" data-path="BioC.html"><a href="BioC.html#multi-data-set"><i class="fa fa-check"></i><b>3.6</b> Multi Data Set</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="opal.html"><a href="opal.html"><i class="fa fa-check"></i><b>4</b> Opal</a>
<ul>
<li class="chapter" data-level="4.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>4.1</b> Introduction</a></li>
<li class="chapter" data-level="4.2" data-path="opal.html"><a href="opal.html#data-management"><i class="fa fa-check"></i><b>4.2</b> Data Management</a></li>
<li class="chapter" data-level="4.3" data-path="opal.html"><a href="opal.html#security"><i class="fa fa-check"></i><b>4.3</b> Security</a></li>
<li class="chapter" data-level="4.4" data-path="opal.html"><a href="opal.html#r-integration"><i class="fa fa-check"></i><b>4.4</b> R Integration</a></li>
<li class="chapter" data-level="4.5" data-path="opal.html"><a href="opal.html#opal-demo-site"><i class="fa fa-check"></i><b>4.5</b> Opal demo site</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="datashield.html"><a href="datashield.html"><i class="fa fa-check"></i><b>5</b> DataSHIELD</a>
<ul>
<li class="chapter" data-level="5.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>5.1</b> Introduction</a></li>
<li class="chapter" data-level="5.2" data-path="datashield.html"><a href="datashield.html#datashield-r-interface-dsi"><i class="fa fa-check"></i><b>5.2</b> DataSHIELD R Interface (DSI)</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="datashield.html"><a href="datashield.html#class-structures"><i class="fa fa-check"></i><b>5.2.1</b> Class Structures</a></li>
<li class="chapter" data-level="5.2.2" data-path="datashield.html"><a href="datashield.html#higher-level-functions"><i class="fa fa-check"></i><b>5.2.2</b> Higher Level Functions</a></li>
<li class="chapter" data-level="5.2.3" data-path="datashield.html"><a href="datashield.html#options"><i class="fa fa-check"></i><b>5.2.3</b> Options</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="datashield.html"><a href="datashield.html#datashieldopal-implementation"><i class="fa fa-check"></i><b>5.3</b> DataSHIELD/Opal Implementation</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="datashield.html"><a href="datashield.html#client"><i class="fa fa-check"></i><b>5.3.1</b> Client</a></li>
<li class="chapter" data-level="5.3.2" data-path="datashield.html"><a href="datashield.html#server"><i class="fa fa-check"></i><b>5.3.2</b> Server</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="datashield.html"><a href="datashield.html#demo"><i class="fa fa-check"></i><b>5.4</b> Demo</a></li>
</ul></li>
<li class="part"><span><b>II Focus Topics</b></span></li>
<li class="chapter" data-level="6" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>6</b> The Resources</a>
<ul>
<li class="chapter" data-level="6.1" data-path="resources.html"><a href="resources.html#concept"><i class="fa fa-check"></i><b>6.1</b> Concept</a></li>
<li class="chapter" data-level="6.2" data-path="resources.html"><a href="resources.html#types"><i class="fa fa-check"></i><b>6.2</b> Types</a></li>
<li class="chapter" data-level="6.3" data-path="resources.html"><a href="resources.html#definition"><i class="fa fa-check"></i><b>6.3</b> Definition</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="resourcer.html"><a href="resourcer.html"><i class="fa fa-check"></i><b>7</b> The resourcer Package</a>
<ul>
<li class="chapter" data-level="7.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>7.1</b> Introduction</a></li>
<li class="chapter" data-level="7.2" data-path="resourcer.html"><a href="resourcer.html#file-resources"><i class="fa fa-check"></i><b>7.2</b> File Resources</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="resourcer.html"><a href="resourcer.html#file-getter"><i class="fa fa-check"></i><b>7.2.1</b> File Getter</a></li>
<li class="chapter" data-level="7.2.2" data-path="resourcer.html"><a href="resourcer.html#file-data-format"><i class="fa fa-check"></i><b>7.2.2</b> File Data Format</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="resourcer.html"><a href="resourcer.html#database-resources"><i class="fa fa-check"></i><b>7.3</b> Database Resources</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="resourcer.html"><a href="resourcer.html#dbi-connectors"><i class="fa fa-check"></i><b>7.3.1</b> DBI Connectors</a></li>
<li class="chapter" data-level="7.3.2" data-path="resourcer.html"><a href="resourcer.html#use-dplyr"><i class="fa fa-check"></i><b>7.3.2</b> Use dplyr</a></li>
<li class="chapter" data-level="7.3.3" data-path="resourcer.html"><a href="resourcer.html#document-databases"><i class="fa fa-check"></i><b>7.3.3</b> Document Databases</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="resourcer.html"><a href="resourcer.html#computation-resources"><i class="fa fa-check"></i><b>7.4</b> Computation Resources</a></li>
<li class="chapter" data-level="7.5" data-path="resourcer.html"><a href="resourcer.html#interacting-with-r-resources"><i class="fa fa-check"></i><b>7.5</b> Interacting with R Resources</a></li>
<li class="chapter" data-level="7.6" data-path="resourcer.html"><a href="resourcer.html#extending_resources"><i class="fa fa-check"></i><b>7.6</b> Extending Resources</a></li>
<li class="chapter" data-level="7.7" data-path="resourcer.html"><a href="resourcer.html#resource-forms"><i class="fa fa-check"></i><b>7.7</b> Resource Forms</a></li>
<li class="chapter" data-level="7.8" data-path="resourcer.html"><a href="resourcer.html#examples-of-resources"><i class="fa fa-check"></i><b>7.8</b> Examples of Resources</a></li>
<li class="chapter" data-level="7.9" data-path="resourcer.html"><a href="resourcer.html#using-resources-with-datashield"><i class="fa fa-check"></i><b>7.9</b> Using Resources with DataSHIELD</a>
<ul>
<li class="chapter" data-level="7.9.1" data-path="resourcer.html"><a href="resourcer.html#tsv-file-into-a-tibble-or-data.frame"><i class="fa fa-check"></i><b>7.9.1</b> TSV File into a tibble or data.frame</a></li>
<li class="chapter" data-level="7.9.2" data-path="resourcer.html"><a href="resourcer.html#r-data-file-into-a-r-object"><i class="fa fa-check"></i><b>7.9.2</b> R Data File into a R object</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>III Resources Extensions</b></span></li>
<li class="chapter" data-level="8" data-path="omic-extension.html"><a href="omic-extension.html"><i class="fa fa-check"></i><b>8</b> Extension to VCF files to peform GWAS with Bioconductor</a></li>
<li class="chapter" data-level="9" data-path="shell-extension.html"><a href="shell-extension.html"><i class="fa fa-check"></i><b>9</b> Extension to secure shell programs: GWAS with PLINK</a></li>
<li class="chapter" data-level="10" data-path="extension-to-ga4gh-and-ega-bamvcf-files.html"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html"><i class="fa fa-check"></i><b>10</b> Extension to GA4GH and EGA (BAM/VCF files)</a>
<ul>
<li class="chapter" data-level="10.1" data-path="extension-to-ga4gh-and-ega-bamvcf-files.html"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#upload-ga4gh-and-ega-resources-into-opal"><i class="fa fa-check"></i><b>10.1</b> Upload GA4GH and EGA resources into OPAL</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="extension-to-ga4gh-and-ega-bamvcf-files.html"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#manually"><i class="fa fa-check"></i><b>10.1.1</b> Manually</a></li>
<li class="chapter" data-level="10.1.2" data-path="extension-to-ga4gh-and-ega-bamvcf-files.html"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#using-r"><i class="fa fa-check"></i><b>10.1.2</b> Using R</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="extension-to-ga4gh-and-ega-bamvcf-files.html"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#adding-a-genomic-resource"><i class="fa fa-check"></i><b>10.2</b> Adding a genomic resource</a></li>
<li class="chapter" data-level="10.3" data-path="extension-to-ga4gh-and-ega-bamvcf-files.html"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#descriptive-omic-data-analysis"><i class="fa fa-check"></i><b>10.3</b> Descriptive omic data analysis</a></li>
</ul></li>
<li class="part"><span><b>IV Workflows</b></span></li>
<li class="chapter" data-level="11" data-path="setup.html"><a href="setup.html"><i class="fa fa-check"></i><b>11</b> Setup</a>
<ul>
<li class="chapter" data-level="11.1" data-path="setup.html"><a href="setup.html#providing-datashield-packages-in-the-opal-server"><i class="fa fa-check"></i><b>11.1</b> Providing DataSHIELD packages in the opal server</a></li>
<li class="chapter" data-level="11.2" data-path="setup.html"><a href="setup.html#required-r-packages-in-the-client-site-e.g.-local-machine"><i class="fa fa-check"></i><b>11.2</b> Required R Packages in the client site (e.g. local machine)</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="basic-statistical-analyses.html"><a href="basic-statistical-analyses.html"><i class="fa fa-check"></i><b>12</b> Basic statistical analyses</a>
<ul>
<li class="chapter" data-level="12.1" data-path="basic-statistical-analyses.html"><a href="basic-statistical-analyses.html#analysis-from-a-single-study"><i class="fa fa-check"></i><b>12.1</b> Analysis from a single study</a></li>
<li class="chapter" data-level="12.2" data-path="basic-statistical-analyses.html"><a href="basic-statistical-analyses.html#analysis-from-a-multiple-studies"><i class="fa fa-check"></i><b>12.2</b> Analysis from a multiple studies</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="Omic.html"><a href="Omic.html"><i class="fa fa-check"></i><b>13</b> Omic data analysis: types of implemented analyses</a></li>
<li class="chapter" data-level="14" data-path="differential-gene-expression-dge-analysis.html"><a href="differential-gene-expression-dge-analysis.html"><i class="fa fa-check"></i><b>14</b> Differential gene expression (DGE) analysis</a></li>
<li class="chapter" data-level="15" data-path="epigenome-wide-association-analysis-ewas.html"><a href="epigenome-wide-association-analysis-ewas.html"><i class="fa fa-check"></i><b>15</b> Epigenome-wide association analysis (EWAS)</a>
<ul>
<li class="chapter" data-level="15.1" data-path="epigenome-wide-association-analysis-ewas.html"><a href="epigenome-wide-association-analysis-ewas.html#single-cpg-analysis"><i class="fa fa-check"></i><b>15.1</b> Single CpG analysis</a></li>
<li class="chapter" data-level="15.2" data-path="epigenome-wide-association-analysis-ewas.html"><a href="epigenome-wide-association-analysis-ewas.html#multiple-cpg-analysis"><i class="fa fa-check"></i><b>15.2</b> Multiple CpG analysis</a></li>
<li class="chapter" data-level="15.3" data-path="epigenome-wide-association-analysis-ewas.html"><a href="epigenome-wide-association-analysis-ewas.html#adjusting-for-surrogate-variables"><i class="fa fa-check"></i><b>15.3</b> Adjusting for Surrogate Variables</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="genome-wide-association-analyses-gwas.html"><a href="genome-wide-association-analyses-gwas.html"><i class="fa fa-check"></i><b>16</b> Genome-wide association analyses (GWAS)</a>
<ul>
<li class="chapter" data-level="16.1" data-path="genome-wide-association-analyses-gwas.html"><a href="genome-wide-association-analyses-gwas.html#gwasBioc"><i class="fa fa-check"></i><b>16.1</b> GWAS with Bioconductor</a></li>
<li class="chapter" data-level="16.2" data-path="genome-wide-association-analyses-gwas.html"><a href="genome-wide-association-analyses-gwas.html#gwas-with-plink"><i class="fa fa-check"></i><b>16.2</b> GWAS with PLINK</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="GIS.html"><a href="GIS.html"><i class="fa fa-check"></i><b>17</b> Geospatial data analysis</a>
<ul>
<li class="chapter" data-level="17.1" data-path="GIS.html"><a href="GIS.html#introducing-the-analysis"><i class="fa fa-check"></i><b>17.1</b> Introducing the analysis</a></li>
<li class="chapter" data-level="17.2" data-path="GIS.html"><a href="GIS.html#setting-up-the-analysis"><i class="fa fa-check"></i><b>17.2</b> Setting up the analysis</a></li>
<li class="chapter" data-level="17.3" data-path="GIS.html"><a href="GIS.html#data-manipulation"><i class="fa fa-check"></i><b>17.3</b> Data manipulation</a></li>
<li class="chapter" data-level="17.4" data-path="GIS.html"><a href="GIS.html#generating-the-final-result"><i class="fa fa-check"></i><b>17.4</b> Generating the final result</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="exposome.html"><a href="exposome.html"><i class="fa fa-check"></i><b>18</b> Exposome data analysis</a>
<ul>
<li class="chapter" data-level="18.1" data-path="exposome.html"><a href="exposome.html#loading-the-exposome-set"><i class="fa fa-check"></i><b>18.1</b> Loading the Exposome Set</a>
<ul>
<li class="chapter" data-level="18.1.1" data-path="exposome.html"><a href="exposome.html#from-tables-to-exposomeset"><i class="fa fa-check"></i><b>18.1.1</b> From tables to <code>ExposomeSet</code></a></li>
<li class="chapter" data-level="18.1.2" data-path="exposome.html"><a href="exposome.html#from-resource-to-exposomeset"><i class="fa fa-check"></i><b>18.1.2</b> From resource to <code>ExposomeSet</code></a></li>
</ul></li>
<li class="chapter" data-level="18.2" data-path="exposome.html"><a href="exposome.html#exploring-the-loaded-dataset"><i class="fa fa-check"></i><b>18.2</b> Exploring the loaded dataset</a>
<ul>
<li class="chapter" data-level="18.2.1" data-path="exposome.html"><a href="exposome.html#family-names"><i class="fa fa-check"></i><b>18.2.1</b> Family names</a></li>
<li class="chapter" data-level="18.2.2" data-path="exposome.html"><a href="exposome.html#exposures-and-phenotypes-names"><i class="fa fa-check"></i><b>18.2.2</b> Exposures and phenotypes names</a></li>
<li class="chapter" data-level="18.2.3" data-path="exposome.html"><a href="exposome.html#summary-of-variables"><i class="fa fa-check"></i><b>18.2.3</b> Summary of variables</a></li>
<li class="chapter" data-level="18.2.4" data-path="exposome.html"><a href="exposome.html#missing-data"><i class="fa fa-check"></i><b>18.2.4</b> Missing data</a></li>
<li class="chapter" data-level="18.2.5" data-path="exposome.html"><a href="exposome.html#exposures-normality"><i class="fa fa-check"></i><b>18.2.5</b> Exposures Normality</a></li>
<li class="chapter" data-level="18.2.6" data-path="exposome.html"><a href="exposome.html#exposures-behaviour"><i class="fa fa-check"></i><b>18.2.6</b> Exposures Behaviour</a></li>
</ul></li>
<li class="chapter" data-level="18.3" data-path="exposome.html"><a href="exposome.html#exposures-imputation"><i class="fa fa-check"></i><b>18.3</b> Exposures Imputation</a></li>
<li class="chapter" data-level="18.4" data-path="exposome.html"><a href="exposome.html#performing-an-exwas"><i class="fa fa-check"></i><b>18.4</b> Performing an ExWAS</a></li>
<li class="chapter" data-level="18.5" data-path="exposome.html"><a href="exposome.html#exposures-pca"><i class="fa fa-check"></i><b>18.5</b> Exposures PCA</a></li>
<li class="chapter" data-level="18.6" data-path="exposome.html"><a href="exposome.html#exposures-correlation"><i class="fa fa-check"></i><b>18.6</b> Exposures Correlation</a></li>
</ul></li>
<li class="part"><span><b>V Developers</b></span></li>
<li class="chapter" data-level="19" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html"><i class="fa fa-check"></i><b>19</b> DSLite: DataSHIELD Implementation on Local Datasets</a>
<ul>
<li class="chapter" data-level="19.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#development-environment-setup"><i class="fa fa-check"></i><b>19.1</b> Development Environment Setup</a>
<ul>
<li class="chapter" data-level="19.1.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-packages"><i class="fa fa-check"></i><b>19.1.1</b> DataSHIELD Packages</a></li>
<li class="chapter" data-level="19.1.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#test-datasets"><i class="fa fa-check"></i><b>19.1.2</b> Test Datasets</a></li>
<li class="chapter" data-level="19.1.3" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#r-package-development-tools"><i class="fa fa-check"></i><b>19.1.3</b> R Package Development Tools</a></li>
</ul></li>
<li class="chapter" data-level="19.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-development-flow"><i class="fa fa-check"></i><b>19.2</b> DataSHIELD Development Flow</a>
<ul>
<li class="chapter" data-level="19.2.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#dsliteserver"><i class="fa fa-check"></i><b>19.2.1</b> DSLiteServer</a></li>
<li class="chapter" data-level="19.2.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-configuration"><i class="fa fa-check"></i><b>19.2.2</b> DataSHIELD Configuration</a></li>
</ul></li>
<li class="chapter" data-level="19.3" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-sessions"><i class="fa fa-check"></i><b>19.3</b> DataSHIELD Sessions</a></li>
<li class="chapter" data-level="19.4" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#debugging"><i class="fa fa-check"></i><b>19.4</b> Debugging</a></li>
<li class="chapter" data-level="19.5" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#limitations"><i class="fa fa-check"></i><b>19.5</b> Limitations</a>
<ul>
<li class="chapter" data-level="19.5.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#function-parameters-parser"><i class="fa fa-check"></i><b>19.5.1</b> Function Parameters Parser</a></li>
<li class="chapter" data-level="19.5.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#server-side-environments"><i class="fa fa-check"></i><b>19.5.2</b> Server Side Environments</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="20" data-path="creating-datashield-packages.html"><a href="creating-datashield-packages.html"><i class="fa fa-check"></i><b>20</b> Creating DataSHIELD packages</a></li>
<li class="chapter" data-level="21" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html"><i class="fa fa-check"></i><b>21</b> Tips and tricks</a>
<ul>
<li class="chapter" data-level="21.1" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#how-to-create-a-new-project-into-opal"><i class="fa fa-check"></i><b>21.1</b> How to create a new project into OPAL</a>
<ul>
<li class="chapter" data-level="21.1.1" data-path="extension-to-ga4gh-and-ega-bamvcf-files.html"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#manually"><i class="fa fa-check"></i><b>21.1.1</b> Manually</a></li>
<li class="chapter" data-level="21.1.2" data-path="extension-to-ga4gh-and-ega-bamvcf-files.html"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#using-r"><i class="fa fa-check"></i><b>21.1.2</b> Using R</a></li>
</ul></li>
<li class="chapter" data-level="21.2" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#how-to-upload-a-new-resource-into-opal"><i class="fa fa-check"></i><b>21.2</b> How to upload a new resource into OPAL</a>
<ul>
<li class="chapter" data-level="21.2.1" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#create_manually"><i class="fa fa-check"></i><b>21.2.1</b> Manually</a></li>
<li class="chapter" data-level="21.2.2" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#using-r-1"><i class="fa fa-check"></i><b>21.2.2</b> Using R</a></li>
</ul></li>
<li class="chapter" data-level="21.3" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#how-to-install-datashield-packages-into-opal-server"><i class="fa fa-check"></i><b>21.3</b> How to install DataSHIELD packages into OPAL server</a>
<ul>
<li class="chapter" data-level="21.3.1" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#manually-1"><i class="fa fa-check"></i><b>21.3.1</b> Manually</a></li>
<li class="chapter" data-level="21.3.2" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#using-r-2"><i class="fa fa-check"></i><b>21.3.2</b> Using R</a></li>
</ul></li>
<li class="chapter" data-level="21.4" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#how-to-install-r-packages-into-opal-server"><i class="fa fa-check"></i><b>21.4</b> How to install R packages into OPAL server</a>
<ul>
<li class="chapter" data-level="21.4.1" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#manually-2"><i class="fa fa-check"></i><b>21.4.1</b> Manually</a></li>
<li class="chapter" data-level="21.4.2" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#using-r-3"><i class="fa fa-check"></i><b>21.4.2</b> Using R</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="22" data-path="advancedtechnical-questions.html"><a href="advancedtechnical-questions.html"><i class="fa fa-check"></i><b>22</b> Advanced/technical questions</a></li>
<li class="part"><span><b>VI Appendix</b></span></li>
<li class="chapter" data-level="23" data-path="session-info.html"><a href="session-info.html"><i class="fa fa-check"></i><b>23</b> Session Info</a></li>
<li class="chapter" data-level="24" data-path="contributors.html"><a href="contributors.html"><i class="fa fa-check"></i><b>24</b> Contributors</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating non-disclosive big data analyses of data from different resources with R and DataSHIELD</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="epigenome-wide-association-analysis-ewas" class="section level1" number="15">
<h1><span class="header-section-number">15</span> Epigenome-wide association analysis (EWAS)</h1>
<p>EWAS requires basically the same statistical methods as those used in DGE. It should be notice that the <strong>pooled analysis</strong> we are going to illustrate here can also be performed with transcriptomic data since each study must have different range values. If so, gene expression harmonization should be performed, for instance, by standardizing the data at each study. For EWAS where methylation is measured using beta values (e.g CpG data are in the range 0-1) this is not a problem. In any case, adopting the <strong>meta-analysis</strong> approach could be a safe option.</p>
<p>We have downloaded data from <a href="https://www.ncbi.nlm.nih.gov/geo/">GEO</a> corresponding to the accesion number GSE66351 which includes DNA methylation profiling (Illumina 450K array) of 190 individuals. Data corresponds to CpGs beta values measured in the superior temporal gyrus and prefrontal cortex brain regions of patients with Alzheimer’s. Data have been downloaded using <em><a href="https://bioconductor.org/packages/3.11/GEOquery">GEOquery</a></em> package that gets GEO data as <code>ExpressionSet</code> objects. Researchers who are not familiar with <code>ExpressionSet</code>s can read <a href="BioC.html#BioC">this Section</a>. Notice that data are encoded as beta-values that ensure data harmonization across studies.</p>
<p>In order to illustrate how to perform data analyses using federated data, we have split the data into two <code>ExpressionSet</code>s having 100 and 90 samples as if they were two different studies. Figure <a href="resources.html#fig:testResources">6.2</a> shows the two resources defined for both studies (GSE66351_1 and GSE66351_2)</p>
<p>In order to perform omic data analyses, we need first to login and assign resources to DataSHIELD. This can be performed using the <code>as.resource.object()</code> function</p>
<div class="sourceCode" id="cb170"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb170-1"><a href="epigenome-wide-association-analysis-ewas.html#cb170-1"></a>builder &lt;-<span class="st"> </span>DSI<span class="op">::</span><span class="kw">newDSLoginBuilder</span>()</span>
<span id="cb170-2"><a href="epigenome-wide-association-analysis-ewas.html#cb170-2"></a>builder<span class="op">$</span><span class="kw">append</span>(<span class="dt">server =</span> <span class="st">&quot;study1&quot;</span>, <span class="dt">url =</span> <span class="st">&quot;https://opal-demo.obiba.org&quot;</span>, </span>
<span id="cb170-3"><a href="epigenome-wide-association-analysis-ewas.html#cb170-3"></a>               <span class="dt">user =</span> <span class="st">&quot;dsuser&quot;</span>, <span class="dt">password =</span> <span class="st">&quot;password&quot;</span>, </span>
<span id="cb170-4"><a href="epigenome-wide-association-analysis-ewas.html#cb170-4"></a>               <span class="dt">resource =</span> <span class="st">&quot;RSRC.GSE66351_1&quot;</span>, <span class="dt">driver =</span> <span class="st">&quot;OpalDriver&quot;</span>)</span>
<span id="cb170-5"><a href="epigenome-wide-association-analysis-ewas.html#cb170-5"></a>builder<span class="op">$</span><span class="kw">append</span>(<span class="dt">server =</span> <span class="st">&quot;study2&quot;</span>, <span class="dt">url =</span> <span class="st">&quot;https://opal-demo.obiba.org&quot;</span>, </span>
<span id="cb170-6"><a href="epigenome-wide-association-analysis-ewas.html#cb170-6"></a>               <span class="dt">user =</span> <span class="st">&quot;dsuser&quot;</span>, <span class="dt">password =</span> <span class="st">&quot;password&quot;</span>, </span>
<span id="cb170-7"><a href="epigenome-wide-association-analysis-ewas.html#cb170-7"></a>               <span class="dt">resource =</span> <span class="st">&quot;RSRC.GSE66351_2&quot;</span>, <span class="dt">driver =</span> <span class="st">&quot;OpalDriver&quot;</span>)</span>
<span id="cb170-8"><a href="epigenome-wide-association-analysis-ewas.html#cb170-8"></a></span>
<span id="cb170-9"><a href="epigenome-wide-association-analysis-ewas.html#cb170-9"></a>logindata &lt;-<span class="st"> </span>builder<span class="op">$</span><span class="kw">build</span>()</span>
<span id="cb170-10"><a href="epigenome-wide-association-analysis-ewas.html#cb170-10"></a></span>
<span id="cb170-11"><a href="epigenome-wide-association-analysis-ewas.html#cb170-11"></a>conns &lt;-<span class="st"> </span>DSI<span class="op">::</span><span class="kw">datashield.login</span>(<span class="dt">logins =</span> logindata, <span class="dt">assign =</span> <span class="ot">TRUE</span>, </span>
<span id="cb170-12"><a href="epigenome-wide-association-analysis-ewas.html#cb170-12"></a>                               <span class="dt">symbol =</span> <span class="st">&quot;res&quot;</span>)</span>
<span id="cb170-13"><a href="epigenome-wide-association-analysis-ewas.html#cb170-13"></a></span>
<span id="cb170-14"><a href="epigenome-wide-association-analysis-ewas.html#cb170-14"></a></span>
<span id="cb170-15"><a href="epigenome-wide-association-analysis-ewas.html#cb170-15"></a><span class="co"># Assign to the original R class (e.g ExpressionSet)</span></span>
<span id="cb170-16"><a href="epigenome-wide-association-analysis-ewas.html#cb170-16"></a><span class="kw">datashield.assign.expr</span>(conns, <span class="dt">symbol =</span> <span class="st">&quot;methy&quot;</span>, </span>
<span id="cb170-17"><a href="epigenome-wide-association-analysis-ewas.html#cb170-17"></a>                       <span class="dt">expr =</span> <span class="kw">quote</span>(<span class="kw">as.resource.object</span>(res)))</span></code></pre></div>
<p>Now, we can see that the resources are actually loaded into the R servers as their original class</p>
<div class="sourceCode" id="cb171"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb171-1"><a href="epigenome-wide-association-analysis-ewas.html#cb171-1"></a><span class="kw">ds.class</span>(<span class="st">&quot;methy&quot;</span>)</span></code></pre></div>
<pre><code>$study1
[1] &quot;ExpressionSet&quot;
attr(,&quot;package&quot;)
[1] &quot;Biobase&quot;

$study2
[1] &quot;ExpressionSet&quot;
attr(,&quot;package&quot;)
[1] &quot;Biobase&quot;</code></pre>
<p>Then, some Bioconductor-type functions can be use to return non-disclosive information of <code>ExpressionSet</code>s from each server to the client, using similar functions as those defined in the <code>dsBaseClient</code> package. For example, feature names can be returned by</p>
<div class="sourceCode" id="cb173"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb173-1"><a href="epigenome-wide-association-analysis-ewas.html#cb173-1"></a>fn &lt;-<span class="st"> </span><span class="kw">ds.featureNames</span>(<span class="st">&quot;methy&quot;</span>)</span>
<span id="cb173-2"><a href="epigenome-wide-association-analysis-ewas.html#cb173-2"></a><span class="kw">lapply</span>(fn, head)</span></code></pre></div>
<pre><code>$study1
[1] &quot;cg00000029&quot; &quot;cg00000108&quot; &quot;cg00000109&quot; &quot;cg00000165&quot; &quot;cg00000236&quot;
[6] &quot;cg00000289&quot;

$study2
[1] &quot;cg00000029&quot; &quot;cg00000108&quot; &quot;cg00000109&quot; &quot;cg00000165&quot; &quot;cg00000236&quot;
[6] &quot;cg00000289&quot;</code></pre>
<p>Experimental phenotypes variables can be obtained by</p>
<div class="sourceCode" id="cb175"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb175-1"><a href="epigenome-wide-association-analysis-ewas.html#cb175-1"></a><span class="kw">ds.varLabels</span>(<span class="st">&quot;methy&quot;</span>)</span></code></pre></div>
<pre><code>$study1
 [1] &quot;title&quot;                   &quot;geo_accession&quot;          
 [3] &quot;status&quot;                  &quot;submission_date&quot;        
 [5] &quot;last_update_date&quot;        &quot;type&quot;                   
 [7] &quot;channel_count&quot;           &quot;source_name_ch1&quot;        
 [9] &quot;organism_ch1&quot;            &quot;characteristics_ch1&quot;    
[11] &quot;characteristics_ch1.1&quot;   &quot;characteristics_ch1.2&quot;  
[13] &quot;characteristics_ch1.3&quot;   &quot;characteristics_ch1.4&quot;  
[15] &quot;characteristics_ch1.5&quot;   &quot;characteristics_ch1.6&quot;  
[17] &quot;characteristics_ch1.7&quot;   &quot;characteristics_ch1.8&quot;  
[19] &quot;molecule_ch1&quot;            &quot;extract_protocol_ch1&quot;   
[21] &quot;label_ch1&quot;               &quot;label_protocol_ch1&quot;     
[23] &quot;taxid_ch1&quot;               &quot;hyb_protocol&quot;           
[25] &quot;scan_protocol&quot;           &quot;description&quot;            
[27] &quot;data_processing&quot;         &quot;platform_id&quot;            
[29] &quot;contact_name&quot;            &quot;contact_email&quot;          
[31] &quot;contact_phone&quot;           &quot;contact_laboratory&quot;     
[33] &quot;contact_institute&quot;       &quot;contact_address&quot;        
[35] &quot;contact_city&quot;            &quot;contact_zip/postal_code&quot;
[37] &quot;contact_country&quot;         &quot;supplementary_file&quot;     
[39] &quot;supplementary_file.1&quot;    &quot;data_row_count&quot;         
[41] &quot;age&quot;                     &quot;braak_stage&quot;            
[43] &quot;brain_region&quot;            &quot;cell type&quot;              
[45] &quot;diagnosis&quot;               &quot;donor_id&quot;               
[47] &quot;sentrix_id&quot;              &quot;sentrix_position&quot;       
[49] &quot;Sex&quot;                    

$study2
 [1] &quot;title&quot;                   &quot;geo_accession&quot;          
 [3] &quot;status&quot;                  &quot;submission_date&quot;        
 [5] &quot;last_update_date&quot;        &quot;type&quot;                   
 [7] &quot;channel_count&quot;           &quot;source_name_ch1&quot;        
 [9] &quot;organism_ch1&quot;            &quot;characteristics_ch1&quot;    
[11] &quot;characteristics_ch1.1&quot;   &quot;characteristics_ch1.2&quot;  
[13] &quot;characteristics_ch1.3&quot;   &quot;characteristics_ch1.4&quot;  
[15] &quot;characteristics_ch1.5&quot;   &quot;characteristics_ch1.6&quot;  
[17] &quot;characteristics_ch1.7&quot;   &quot;characteristics_ch1.8&quot;  
[19] &quot;molecule_ch1&quot;            &quot;extract_protocol_ch1&quot;   
[21] &quot;label_ch1&quot;               &quot;label_protocol_ch1&quot;     
[23] &quot;taxid_ch1&quot;               &quot;hyb_protocol&quot;           
[25] &quot;scan_protocol&quot;           &quot;description&quot;            
[27] &quot;data_processing&quot;         &quot;platform_id&quot;            
[29] &quot;contact_name&quot;            &quot;contact_email&quot;          
[31] &quot;contact_phone&quot;           &quot;contact_laboratory&quot;     
[33] &quot;contact_institute&quot;       &quot;contact_address&quot;        
[35] &quot;contact_city&quot;            &quot;contact_zip/postal_code&quot;
[37] &quot;contact_country&quot;         &quot;supplementary_file&quot;     
[39] &quot;supplementary_file.1&quot;    &quot;data_row_count&quot;         
[41] &quot;age&quot;                     &quot;braak_stage&quot;            
[43] &quot;brain_region&quot;            &quot;cell type&quot;              
[45] &quot;diagnosis&quot;               &quot;donor_id&quot;               
[47] &quot;sentrix_id&quot;              &quot;sentrix_position&quot;       
[49] &quot;Sex&quot;                    

attr(,&quot;class&quot;)
[1] &quot;dsvarLabels&quot; &quot;list&quot;       </code></pre>
<div id="single-cpg-analysis" class="section level2" number="15.1">
<h2><span class="header-section-number">15.1</span> Single CpG analysis</h2>
<p>Once the methylation data have been loaded into the opal server, we can perform different type of analyses using functions from the <code>dsOmicsClient</code> package. Let us start by illustrating how to analyze a single CpG from two studies by using an approach that is mathematically equivalent to placing all individual-level.</p>
<div class="sourceCode" id="cb177"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb177-1"><a href="epigenome-wide-association-analysis-ewas.html#cb177-1"></a>ans &lt;-<span class="st"> </span><span class="kw">ds.lmFeature</span>(<span class="dt">feature =</span> <span class="st">&quot;cg07363416&quot;</span>, </span>
<span id="cb177-2"><a href="epigenome-wide-association-analysis-ewas.html#cb177-2"></a>                    <span class="dt">model =</span> <span class="op">~</span><span class="st"> </span>diagnosis <span class="op">+</span><span class="st"> </span>Sex, </span>
<span id="cb177-3"><a href="epigenome-wide-association-analysis-ewas.html#cb177-3"></a>                    <span class="dt">Set =</span> <span class="st">&quot;methy&quot;</span>,</span>
<span id="cb177-4"><a href="epigenome-wide-association-analysis-ewas.html#cb177-4"></a>                    <span class="dt">datasources =</span> conns)</span>
<span id="cb177-5"><a href="epigenome-wide-association-analysis-ewas.html#cb177-5"></a>ans</span></code></pre></div>
<pre><code>             Estimate Std. Error   p-value
cg07363416 0.03459886 0.02504291 0.1670998
attr(,&quot;class&quot;)
[1] &quot;dsLmFeature&quot; &quot;matrix&quot;      &quot;array&quot;      </code></pre>
</div>
<div id="multiple-cpg-analysis" class="section level2" number="15.2">
<h2><span class="header-section-number">15.2</span> Multiple CpG analysis</h2>
<p>The same analysis can be performed for all features (e.g. CpGs) just avoiding the <code>feature</code> argument. This process can be parallelized using <code>mclapply</code> function from the <code>multicore</code> package.</p>
<div class="sourceCode" id="cb179"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb179-1"><a href="epigenome-wide-association-analysis-ewas.html#cb179-1"></a>ans &lt;-<span class="st"> </span><span class="kw">ds.lmFeature</span>(<span class="dt">model =</span> <span class="op">~</span><span class="st"> </span>diagnosis <span class="op">+</span><span class="st"> </span>Sex, </span>
<span id="cb179-2"><a href="epigenome-wide-association-analysis-ewas.html#cb179-2"></a>                    <span class="dt">Set =</span> <span class="st">&quot;methy&quot;</span>,</span>
<span id="cb179-3"><a href="epigenome-wide-association-analysis-ewas.html#cb179-3"></a>                    <span class="dt">datasources =</span> conns,</span>
<span id="cb179-4"><a href="epigenome-wide-association-analysis-ewas.html#cb179-4"></a>                    <span class="dt">mc.cores =</span> <span class="dv">20</span>)</span></code></pre></div>
<p>This method corresponds to the <strong>pooled analysis</strong> approach and can be very time consiming since the function repeatedly calls the DataSHIELD function <code>ds.glm()</code>. We can adopt another strategy that is to run a glm of each feature independently at each study using <code>limma</code> package (which is really fast) and then combine the results (i.e. <strong>meta-analysis</strong> approach).</p>
<div class="sourceCode" id="cb180"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb180-1"><a href="epigenome-wide-association-analysis-ewas.html#cb180-1"></a>ans.limma &lt;-<span class="st"> </span><span class="kw">ds.limma</span>(<span class="dt">model =</span> <span class="op">~</span><span class="st"> </span>diagnosis <span class="op">+</span><span class="st"> </span>Sex,</span>
<span id="cb180-2"><a href="epigenome-wide-association-analysis-ewas.html#cb180-2"></a>                      <span class="dt">Set =</span> <span class="st">&quot;methy&quot;</span>, </span>
<span id="cb180-3"><a href="epigenome-wide-association-analysis-ewas.html#cb180-3"></a>                      <span class="dt">datasources =</span> conns)</span></code></pre></div>
<p>Then, we can visualize the top genes at each study (i.e server) by</p>
<div class="sourceCode" id="cb181"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb181-1"><a href="epigenome-wide-association-analysis-ewas.html#cb181-1"></a><span class="kw">lapply</span>(ans.limma, head)</span></code></pre></div>
<pre><code>$study1
# A tibble: 6 x 10
  id      logFC    CI.L    CI.R AveExpr     t P.Value adj.P.Val     B
  &lt;chr&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;     &lt;dbl&gt; &lt;dbl&gt;
1 cg13~ -0.147  -0.191  -0.103    0.380 -6.62 1.90e-9  0.000466 10.6 
2 cg23~ -0.0569 -0.0741 -0.0397   0.200 -6.58 2.32e-9  0.000466 10.4 
3 cg13~ -0.0820 -0.107  -0.0570   0.437 -6.50 3.27e-9  0.000466 10.0 
4 cg12~ -0.0519 -0.0678 -0.0359   0.145 -6.45 4.25e-9  0.000466  9.76
5 cg24~ -0.0452 -0.0593 -0.0312   0.139 -6.39 5.47e-9  0.000466  9.51
6 cg02~ -0.125  -0.165  -0.0860   0.247 -6.33 7.31e-9  0.000466  9.23
# ... with 1 more variable: SE &lt;dbl&gt;

$study2
# A tibble: 6 x 10
  id      logFC    CI.L    CI.R AveExpr     t P.Value adj.P.Val     B
  &lt;chr&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;     &lt;dbl&gt; &lt;dbl&gt;
1 cg04~ -0.101  -0.135  -0.0669  0.345  -5.91 6.21e-8    0.0172  7.18
2 cg07~ -0.0431 -0.0577 -0.0284  0.776  -5.85 8.22e-8    0.0172  6.90
3 cg27~ -0.0688 -0.0924 -0.0452  0.277  -5.79 1.07e-7    0.0172  6.64
4 cg08~ -0.0461 -0.0627 -0.0296  0.166  -5.55 2.98e-7    0.0360  5.64
5 cg18~ -0.0491 -0.0671 -0.0311  0.157  -5.42 5.07e-7    0.0489  5.12
6 cg02~ -0.0199 -0.0272 -0.0125  0.0947 -5.36 6.70e-7    0.0538  4.84
# ... with 1 more variable: SE &lt;dbl&gt;</code></pre>
<p>The annotation can be added by using the argument <code>annotCols</code>. It should be a vector with the columns of the annotation available in the <code>ExpressionSet</code> or <code>RangedSummarizedExperiment</code> that want to be showed. The columns of the annotation can be obtained by</p>
<div class="sourceCode" id="cb183"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb183-1"><a href="epigenome-wide-association-analysis-ewas.html#cb183-1"></a><span class="kw">ds.fvarLabels</span>(<span class="st">&quot;methy&quot;</span>)</span></code></pre></div>
<pre><code>$study1
 [1] &quot;ID&quot;                          &quot;Name&quot;                       
 [3] &quot;AddressA_ID&quot;                 &quot;AlleleA_ProbeSeq&quot;           
 [5] &quot;AddressB_ID&quot;                 &quot;AlleleB_ProbeSeq&quot;           
 [7] &quot;Infinium_Design_Type&quot;        &quot;Next_Base&quot;                  
 [9] &quot;Color_Channel&quot;               &quot;Forward_Sequence&quot;           
[11] &quot;Genome_Build&quot;                &quot;CHR&quot;                        
[13] &quot;MAPINFO&quot;                     &quot;SourceSeq&quot;                  
[15] &quot;Chromosome_36&quot;               &quot;Coordinate_36&quot;              
[17] &quot;Strand&quot;                      &quot;Probe_SNPs&quot;                 
[19] &quot;Probe_SNPs_10&quot;               &quot;Random_Loci&quot;                
[21] &quot;Methyl27_Loci&quot;               &quot;UCSC_RefGene_Name&quot;          
[23] &quot;UCSC_RefGene_Accession&quot;      &quot;UCSC_RefGene_Group&quot;         
[25] &quot;UCSC_CpG_Islands_Name&quot;       &quot;Relation_to_UCSC_CpG_Island&quot;
[27] &quot;Phantom&quot;                     &quot;DMR&quot;                        
[29] &quot;Enhancer&quot;                    &quot;HMM_Island&quot;                 
[31] &quot;Regulatory_Feature_Name&quot;     &quot;Regulatory_Feature_Group&quot;   
[33] &quot;DHS&quot;                         &quot;RANGE_START&quot;                
[35] &quot;RANGE_END&quot;                   &quot;RANGE_GB&quot;                   
[37] &quot;SPOT_ID&quot;                    

$study2
 [1] &quot;ID&quot;                          &quot;Name&quot;                       
 [3] &quot;AddressA_ID&quot;                 &quot;AlleleA_ProbeSeq&quot;           
 [5] &quot;AddressB_ID&quot;                 &quot;AlleleB_ProbeSeq&quot;           
 [7] &quot;Infinium_Design_Type&quot;        &quot;Next_Base&quot;                  
 [9] &quot;Color_Channel&quot;               &quot;Forward_Sequence&quot;           
[11] &quot;Genome_Build&quot;                &quot;CHR&quot;                        
[13] &quot;MAPINFO&quot;                     &quot;SourceSeq&quot;                  
[15] &quot;Chromosome_36&quot;               &quot;Coordinate_36&quot;              
[17] &quot;Strand&quot;                      &quot;Probe_SNPs&quot;                 
[19] &quot;Probe_SNPs_10&quot;               &quot;Random_Loci&quot;                
[21] &quot;Methyl27_Loci&quot;               &quot;UCSC_RefGene_Name&quot;          
[23] &quot;UCSC_RefGene_Accession&quot;      &quot;UCSC_RefGene_Group&quot;         
[25] &quot;UCSC_CpG_Islands_Name&quot;       &quot;Relation_to_UCSC_CpG_Island&quot;
[27] &quot;Phantom&quot;                     &quot;DMR&quot;                        
[29] &quot;Enhancer&quot;                    &quot;HMM_Island&quot;                 
[31] &quot;Regulatory_Feature_Name&quot;     &quot;Regulatory_Feature_Group&quot;   
[33] &quot;DHS&quot;                         &quot;RANGE_START&quot;                
[35] &quot;RANGE_END&quot;                   &quot;RANGE_GB&quot;                   
[37] &quot;SPOT_ID&quot;                    

attr(,&quot;class&quot;)
[1] &quot;dsfvarLabels&quot; &quot;list&quot;        </code></pre>
<p>Then we can run the analysis and obtain the output with the chromosome and gene symbol by:</p>
<div class="sourceCode" id="cb185"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb185-1"><a href="epigenome-wide-association-analysis-ewas.html#cb185-1"></a>ans.limma.annot &lt;-<span class="st"> </span><span class="kw">ds.limma</span>(<span class="dt">model =</span> <span class="op">~</span><span class="st"> </span>diagnosis <span class="op">+</span><span class="st"> </span>Sex,</span>
<span id="cb185-2"><a href="epigenome-wide-association-analysis-ewas.html#cb185-2"></a>                            <span class="dt">Set =</span> <span class="st">&quot;methy&quot;</span>, </span>
<span id="cb185-3"><a href="epigenome-wide-association-analysis-ewas.html#cb185-3"></a>                            <span class="dt">annotCols =</span> <span class="kw">c</span>(<span class="st">&quot;CHR&quot;</span>, <span class="st">&quot;UCSC_RefGene_Name&quot;</span>),</span>
<span id="cb185-4"><a href="epigenome-wide-association-analysis-ewas.html#cb185-4"></a>                            <span class="dt">datasources =</span> conns)</span></code></pre></div>
<div class="sourceCode" id="cb186"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb186-1"><a href="epigenome-wide-association-analysis-ewas.html#cb186-1"></a><span class="kw">lapply</span>(ans.limma.annot, head)</span></code></pre></div>
<pre><code>$study1
# A tibble: 6 x 12
  id    CHR   UCSC_RefGene_Na~   logFC    CI.L    CI.R AveExpr     t
  &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;              &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;
1 cg13~ 2     &quot;ECEL1P2&quot;        -0.147  -0.191  -0.103    0.380 -6.62
2 cg23~ 2     &quot;MTA3&quot;           -0.0569 -0.0741 -0.0397   0.200 -6.58
3 cg13~ 17    &quot;&quot;               -0.0820 -0.107  -0.0570   0.437 -6.50
4 cg12~ 19    &quot;MEX3D&quot;          -0.0519 -0.0678 -0.0359   0.145 -6.45
5 cg24~ 19    &quot;ISOC2;ISOC2;IS~ -0.0452 -0.0593 -0.0312   0.139 -6.39
6 cg02~ 2     &quot;ECEL1P2&quot;        -0.125  -0.165  -0.0860   0.247 -6.33
# ... with 4 more variables: P.Value &lt;dbl&gt;, adj.P.Val &lt;dbl&gt;, B &lt;dbl&gt;,
#   SE &lt;dbl&gt;

$study2
# A tibble: 6 x 12
  id    CHR   UCSC_RefGene_Na~   logFC    CI.L    CI.R AveExpr     t
  &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;              &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;
1 cg04~ 11    &quot;CD6&quot;            -0.101  -0.135  -0.0669  0.345  -5.91
2 cg07~ 6     &quot;MUC21&quot;          -0.0431 -0.0577 -0.0284  0.776  -5.85
3 cg27~ 11    &quot;CD6&quot;            -0.0688 -0.0924 -0.0452  0.277  -5.79
4 cg08~ 1     &quot;&quot;               -0.0461 -0.0627 -0.0296  0.166  -5.55
5 cg18~ 3     &quot;RARRES1;RARRES~ -0.0491 -0.0671 -0.0311  0.157  -5.42
6 cg02~ 8     &quot;&quot;               -0.0199 -0.0272 -0.0125  0.0947 -5.36
# ... with 4 more variables: P.Value &lt;dbl&gt;, adj.P.Val &lt;dbl&gt;, B &lt;dbl&gt;,
#   SE &lt;dbl&gt;</code></pre>
<p>Then, the last step is to meta-analyze the results. Different methods can be used to this end. We have implemented a method that meta-analyze the p-pvalues of each study as follows:</p>
<div class="sourceCode" id="cb188"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb188-1"><a href="epigenome-wide-association-analysis-ewas.html#cb188-1"></a>ans.meta &lt;-<span class="st"> </span><span class="kw">metaPvalues</span>(ans.limma)</span>
<span id="cb188-2"><a href="epigenome-wide-association-analysis-ewas.html#cb188-2"></a>ans.meta</span></code></pre></div>
<pre><code># A tibble: 481,868 x 4
   id                study1      study2   p.meta
   &lt;chr&gt;              &lt;dbl&gt;       &lt;dbl&gt;    &lt;dbl&gt;
 1 cg13138089 0.00000000190 0.00000763  4.78e-13
 2 cg25317941 0.0000000179  0.00000196  1.12e-12
 3 cg02812891 0.00000000731 0.00000707  1.63e-12
 4 cg12706938 0.00000000425 0.0000161   2.14e-12
 5 cg16026647 0.000000101   0.000000797 2.51e-12
 6 cg12695465 0.00000000985 0.0000144   4.33e-12
 7 cg21171625 0.000000146   0.00000225  9.78e-12
 8 cg13772815 0.00000000327 0.000122    1.18e-11
 9 cg00228891 0.000000166   0.00000283  1.38e-11
10 cg21488617 0.0000000186  0.0000299   1.62e-11
# ... with 481,858 more rows</code></pre>
<p>We can verify that the results are pretty similar to those obtained using pooled analyses. Here we compute the association for two of the top-CpGs:</p>
<div class="sourceCode" id="cb190"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb190-1"><a href="epigenome-wide-association-analysis-ewas.html#cb190-1"></a>res1 &lt;-<span class="st"> </span><span class="kw">ds.lmFeature</span>(<span class="dt">feature =</span> <span class="st">&quot;cg13138089&quot;</span>, </span>
<span id="cb190-2"><a href="epigenome-wide-association-analysis-ewas.html#cb190-2"></a>                     <span class="dt">model =</span> <span class="op">~</span><span class="st"> </span>diagnosis <span class="op">+</span><span class="st"> </span>Sex, </span>
<span id="cb190-3"><a href="epigenome-wide-association-analysis-ewas.html#cb190-3"></a>                     <span class="dt">Set =</span> <span class="st">&quot;methy&quot;</span>,</span>
<span id="cb190-4"><a href="epigenome-wide-association-analysis-ewas.html#cb190-4"></a>                     <span class="dt">datasources =</span> conns)</span>
<span id="cb190-5"><a href="epigenome-wide-association-analysis-ewas.html#cb190-5"></a>res1</span></code></pre></div>
<pre><code>             Estimate Std. Error      p-value
cg13138089 -0.1373348 0.01712405 1.057482e-15
attr(,&quot;class&quot;)
[1] &quot;dsLmFeature&quot; &quot;matrix&quot;      &quot;array&quot;      </code></pre>
<div class="sourceCode" id="cb192"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb192-1"><a href="epigenome-wide-association-analysis-ewas.html#cb192-1"></a>res2 &lt;-<span class="st"> </span><span class="kw">ds.lmFeature</span>(<span class="dt">feature =</span> <span class="st">&quot;cg13772815&quot;</span>, </span>
<span id="cb192-2"><a href="epigenome-wide-association-analysis-ewas.html#cb192-2"></a>                     <span class="dt">model =</span> <span class="op">~</span><span class="st"> </span>diagnosis <span class="op">+</span><span class="st"> </span>Sex, </span>
<span id="cb192-3"><a href="epigenome-wide-association-analysis-ewas.html#cb192-3"></a>                     <span class="dt">Set =</span> <span class="st">&quot;methy&quot;</span>,</span>
<span id="cb192-4"><a href="epigenome-wide-association-analysis-ewas.html#cb192-4"></a>                     <span class="dt">datasources =</span> conns)</span>
<span id="cb192-5"><a href="epigenome-wide-association-analysis-ewas.html#cb192-5"></a>res2</span></code></pre></div>
<pre><code>              Estimate  Std. Error      p-value
cg13772815 -0.06786137 0.009128915 1.056225e-13
attr(,&quot;class&quot;)
[1] &quot;dsLmFeature&quot; &quot;matrix&quot;      &quot;array&quot;      </code></pre>
<p>We can create a QQ-plot by using the function <code>qqplot</code> available in our package.</p>
<div class="sourceCode" id="cb194"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb194-1"><a href="epigenome-wide-association-analysis-ewas.html#cb194-1"></a><span class="kw">qqplot</span>(ans.meta<span class="op">$</span>p.meta)</span></code></pre></div>
<p><img src="fig/qqplot-1.png" width="672" /></p>
<p>Here In some cases inflation can be observed, so that, correction for cell-type or surrogate variables must be performed. We describe how we can do that in the next two sections.</p>
</div>
<div id="adjusting-for-surrogate-variables" class="section level2" number="15.3">
<h2><span class="header-section-number">15.3</span> Adjusting for Surrogate Variables</h2>
<p>The vast majority of omic studies require to control for unwanted variability. The surrogate variable analysis (SVA) can address this issue by estimating some hidden covariates that capture differences across individuals due to some artifacts such as batch effects or sample quality sam among others. The method is implemented in <em><a href="https://bioconductor.org/packages/3.11/SVA">SVA</a></em> package.</p>
<p>Performing this type of analysis using the <code>ds.lmFeature</code> function is not allowed since estimating SVA would require to implement a non-disclosive method that computes SVA from the different servers. This will be a future topic of the <code>dsOmicsClient</code>. NOTE that, estimating SVA separately at each server would not be a good idea since the aim of SVA is to capture differences mainly due to experimental issues among ALL individuals. What we can do instead is to use the <code>ds.limma</code> function to perform the analyses adjusted for SVA at each study.</p>
<div class="sourceCode" id="cb195"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb195-1"><a href="epigenome-wide-association-analysis-ewas.html#cb195-1"></a>ans.sva &lt;-<span class="st"> </span><span class="kw">ds.limma</span>(<span class="dt">model =</span> <span class="op">~</span><span class="st"> </span>diagnosis <span class="op">+</span><span class="st"> </span>Sex, </span>
<span id="cb195-2"><a href="epigenome-wide-association-analysis-ewas.html#cb195-2"></a>                    <span class="dt">Set =</span> <span class="st">&quot;methy&quot;</span>,</span>
<span id="cb195-3"><a href="epigenome-wide-association-analysis-ewas.html#cb195-3"></a>                    <span class="dt">sva =</span> <span class="ot">TRUE</span>, <span class="dt">annotCols =</span> <span class="kw">c</span>(<span class="st">&quot;CHR&quot;</span>, <span class="st">&quot;UCSC_RefGene_Name&quot;</span>))</span>
<span id="cb195-4"><a href="epigenome-wide-association-analysis-ewas.html#cb195-4"></a>ans.sva</span></code></pre></div>
<pre><code>$study1
# A tibble: 481,868 x 12
   id    CHR   UCSC_RefGene_Na~   logFC    CI.L    CI.R AveExpr     t
   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;              &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;
 1 cg10~ 19    &quot;GNG7&quot;           -0.0547 -0.0721 -0.0373   0.338 -6.25
 2 cg13~ 17    &quot;&quot;               -0.0569 -0.0757 -0.0381   0.437 -6.00
 3 cg11~ 19    &quot;PODNL1;PODNL1;~  0.0334  0.0223  0.0445   0.568  5.97
 4 cg27~ 17    &quot;SLC47A2;SLC47A~  0.0274  0.0182  0.0365   0.548  5.95
 5 cg21~ 6     &quot;&quot;               -0.0453 -0.0609 -0.0297   0.799 -5.78
 6 cg23~ 2     &quot;MTA3&quot;           -0.0327 -0.0440 -0.0215   0.200 -5.77
 7 cg10~ 16    &quot;SALL1;SALL1&quot;     0.0366  0.0240  0.0492   0.137  5.77
 8 cg24~ 1     &quot;&quot;                0.0317  0.0208  0.0427   0.821  5.76
 9 cg13~ 9     &quot;&quot;               -0.0326 -0.0440 -0.0213   0.367 -5.72
10 cg13~ 2     &quot;ECEL1P2&quot;        -0.106  -0.144  -0.0686   0.380 -5.61
# ... with 481,858 more rows, and 4 more variables: P.Value &lt;dbl&gt;,
#   adj.P.Val &lt;dbl&gt;, B &lt;dbl&gt;, SE &lt;dbl&gt;

$study2
# A tibble: 481,868 x 12
   id    CHR   UCSC_RefGene_Na~   logFC    CI.L    CI.R AveExpr     t
   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;              &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;
 1 cg16~ 12    &quot;LRP1&quot;           -0.0418 -0.0548 -0.0289   0.388 -6.42
 2 cg25~ 1     &quot;&quot;               -0.0664 -0.0874 -0.0453   0.569 -6.27
 3 cg12~ 11    &quot;NRXN2;NRXN2&quot;    -0.0273 -0.0367 -0.0179   0.728 -5.81
 4 cg07~ 3     &quot;&quot;               -0.0393 -0.0529 -0.0258   0.160 -5.78
 5 cg07~ 6     &quot;MUC21&quot;          -0.0427 -0.0579 -0.0276   0.776 -5.61
 6 cg00~ 1     &quot;CR1L&quot;           -0.0568 -0.0771 -0.0365   0.350 -5.56
 7 cg11~ 12    &quot;CNTN1;CNTN1&quot;    -0.0443 -0.0601 -0.0284   0.152 -5.56
 8 cg03~ 7     &quot;PGAM2;PGAM2&quot;    -0.0442 -0.0600 -0.0283   0.211 -5.56
 9 cg07~ 1     &quot;KCNAB2;KCNAB2&quot;   0.0611  0.0392  0.0831   0.659  5.54
10 cg25~ 17    &quot;WNK4&quot;           -0.0392 -0.0533 -0.0251   0.512 -5.52
# ... with 481,858 more rows, and 4 more variables: P.Value &lt;dbl&gt;,
#   adj.P.Val &lt;dbl&gt;, B &lt;dbl&gt;, SE &lt;dbl&gt;

attr(,&quot;class&quot;)
[1] &quot;dsLimma&quot; &quot;list&quot;   </code></pre>
<p>Then, data can be combined meta-anlyzed as follows:</p>
<div class="sourceCode" id="cb197"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb197-1"><a href="epigenome-wide-association-analysis-ewas.html#cb197-1"></a>ans.meta.sv &lt;-<span class="st"> </span><span class="kw">metaPvalues</span>(ans.sva)</span>
<span id="cb197-2"><a href="epigenome-wide-association-analysis-ewas.html#cb197-2"></a>ans.meta.sv</span></code></pre></div>
<pre><code># A tibble: 481,868 x 4
   id               study1      study2   p.meta
   &lt;chr&gt;             &lt;dbl&gt;       &lt;dbl&gt;    &lt;dbl&gt;
 1 cg00228891 0.00000397   0.000000318 3.58e-11
 2 cg01301319 0.00000609   0.00000123  2.00e-10
 3 cg22962123 0.00000106   0.00000767  2.17e-10
 4 cg24302412 0.0000139    0.000000762 2.77e-10
 5 cg02812891 0.000000408  0.0000327   3.47e-10
 6 cg23859635 0.000000109  0.000190    5.29e-10
 7 cg13138089 0.000000215  0.000105    5.74e-10
 8 cg24938077 0.0000125    0.00000254  8.02e-10
 9 cg13772815 0.0000000391 0.00132     1.27e- 9
10 cg21212881 0.000000340  0.000248    2.04e- 9
# ... with 481,858 more rows</code></pre>
<p>The DataSHIELD session must by closed by:</p>
<div class="sourceCode" id="cb199"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb199-1"><a href="epigenome-wide-association-analysis-ewas.html#cb199-1"></a><span class="kw">datashield.logout</span>(conns)</span></code></pre></div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="differential-gene-expression-dge-analysis.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="genome-wide-association-analyses-gwas.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
