<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>10 Extension to GA4GH and EGA (BAM/VCF files) | Orchestrating non-disclosive big data analyses of data from different resources with R and DataSHIELD</title>
  <meta name="description" content="10 Extension to GA4GH and EGA (BAM/VCF files) | Orchestrating non-disclosive big data analyses of data from different resources with R and DataSHIELD" />
  <meta name="generator" content="bookdown 0.20 and GitBook 2.6.7" />

  <meta property="og:title" content="10 Extension to GA4GH and EGA (BAM/VCF files) | Orchestrating non-disclosive big data analyses of data from different resources with R and DataSHIELD" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="10 Extension to GA4GH and EGA (BAM/VCF files) | Orchestrating non-disclosive big data analyses of data from different resources with R and DataSHIELD" />
  
  
  



<meta name="date" content="2020-11-14" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="shell-extension.html"/>
<link rel="next" href="setup.html"/>
<script src="libs/header-attrs-2.3/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="chapter" data-level="" data-path="citation.html"><a href="citation.html"><i class="fa fa-check"></i>Citation</a></li>
<li class="part"><span><b>I Preamble</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#what-you-will-learn"><i class="fa fa-check"></i><b>1.1</b> What you will learn</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#preliminaries"><i class="fa fa-check"></i><b>1.2</b> Preliminaries</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#acknowledgments"><i class="fa fa-check"></i><b>1.3</b> Acknowledgments</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="learnR.html"><a href="learnR.html"><i class="fa fa-check"></i><b>2</b> Learning R</a></li>
<li class="chapter" data-level="3" data-path="BioC.html"><a href="BioC.html"><i class="fa fa-check"></i><b>3</b> Bioconductor Data Structures</a>
<ul>
<li class="chapter" data-level="3.1" data-path="BioC.html"><a href="BioC.html#snp-array-data"><i class="fa fa-check"></i><b>3.1</b> SNP Array Data</a></li>
<li class="chapter" data-level="3.2" data-path="BioC.html"><a href="BioC.html#expression-sets"><i class="fa fa-check"></i><b>3.2</b> Expression Sets</a></li>
<li class="chapter" data-level="3.3" data-path="BioC.html"><a href="BioC.html#genomic-ranges"><i class="fa fa-check"></i><b>3.3</b> Genomic Ranges</a></li>
<li class="chapter" data-level="3.4" data-path="BioC.html"><a href="BioC.html#summarized-experiments"><i class="fa fa-check"></i><b>3.4</b> Summarized Experiments</a></li>
<li class="chapter" data-level="3.5" data-path="BioC.html"><a href="BioC.html#ranged-summarized-experiments"><i class="fa fa-check"></i><b>3.5</b> Ranged Summarized Experiments</a></li>
<li class="chapter" data-level="3.6" data-path="BioC.html"><a href="BioC.html#multi-data-set"><i class="fa fa-check"></i><b>3.6</b> Multi Data Set</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="opal.html"><a href="opal.html"><i class="fa fa-check"></i><b>4</b> Opal</a>
<ul>
<li class="chapter" data-level="4.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>4.1</b> Introduction</a></li>
<li class="chapter" data-level="4.2" data-path="opal.html"><a href="opal.html#data-management"><i class="fa fa-check"></i><b>4.2</b> Data Management</a></li>
<li class="chapter" data-level="4.3" data-path="opal.html"><a href="opal.html#security"><i class="fa fa-check"></i><b>4.3</b> Security</a></li>
<li class="chapter" data-level="4.4" data-path="opal.html"><a href="opal.html#r-integration"><i class="fa fa-check"></i><b>4.4</b> R Integration</a></li>
<li class="chapter" data-level="4.5" data-path="opal.html"><a href="opal.html#opal-demo-site"><i class="fa fa-check"></i><b>4.5</b> Opal demo site</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="datashield.html"><a href="datashield.html"><i class="fa fa-check"></i><b>5</b> DataSHIELD</a>
<ul>
<li class="chapter" data-level="5.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>5.1</b> Introduction</a></li>
<li class="chapter" data-level="5.2" data-path="datashield.html"><a href="datashield.html#datashield-r-interface-dsi"><i class="fa fa-check"></i><b>5.2</b> DataSHIELD R Interface (DSI)</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="datashield.html"><a href="datashield.html#class-structures"><i class="fa fa-check"></i><b>5.2.1</b> Class Structures</a></li>
<li class="chapter" data-level="5.2.2" data-path="datashield.html"><a href="datashield.html#higher-level-functions"><i class="fa fa-check"></i><b>5.2.2</b> Higher Level Functions</a></li>
<li class="chapter" data-level="5.2.3" data-path="datashield.html"><a href="datashield.html#options"><i class="fa fa-check"></i><b>5.2.3</b> Options</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="datashield.html"><a href="datashield.html#datashieldopal-implementation"><i class="fa fa-check"></i><b>5.3</b> DataSHIELD/Opal Implementation</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="datashield.html"><a href="datashield.html#client"><i class="fa fa-check"></i><b>5.3.1</b> Client</a></li>
<li class="chapter" data-level="5.3.2" data-path="datashield.html"><a href="datashield.html#server"><i class="fa fa-check"></i><b>5.3.2</b> Server</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="datashield.html"><a href="datashield.html#demo"><i class="fa fa-check"></i><b>5.4</b> Demo</a></li>
</ul></li>
<li class="part"><span><b>II Focus Topics</b></span></li>
<li class="chapter" data-level="6" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>6</b> The Resources</a>
<ul>
<li class="chapter" data-level="6.1" data-path="resources.html"><a href="resources.html#concept"><i class="fa fa-check"></i><b>6.1</b> Concept</a></li>
<li class="chapter" data-level="6.2" data-path="resources.html"><a href="resources.html#types"><i class="fa fa-check"></i><b>6.2</b> Types</a></li>
<li class="chapter" data-level="6.3" data-path="resources.html"><a href="resources.html#definition"><i class="fa fa-check"></i><b>6.3</b> Definition</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="resourcer.html"><a href="resourcer.html"><i class="fa fa-check"></i><b>7</b> The resourcer Package</a>
<ul>
<li class="chapter" data-level="7.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>7.1</b> Introduction</a></li>
<li class="chapter" data-level="7.2" data-path="resourcer.html"><a href="resourcer.html#file-resources"><i class="fa fa-check"></i><b>7.2</b> File Resources</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="resourcer.html"><a href="resourcer.html#file-getter"><i class="fa fa-check"></i><b>7.2.1</b> File Getter</a></li>
<li class="chapter" data-level="7.2.2" data-path="resourcer.html"><a href="resourcer.html#file-data-format"><i class="fa fa-check"></i><b>7.2.2</b> File Data Format</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="resourcer.html"><a href="resourcer.html#database-resources"><i class="fa fa-check"></i><b>7.3</b> Database Resources</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="resourcer.html"><a href="resourcer.html#dbi-connectors"><i class="fa fa-check"></i><b>7.3.1</b> DBI Connectors</a></li>
<li class="chapter" data-level="7.3.2" data-path="resourcer.html"><a href="resourcer.html#use-dplyr"><i class="fa fa-check"></i><b>7.3.2</b> Use dplyr</a></li>
<li class="chapter" data-level="7.3.3" data-path="resourcer.html"><a href="resourcer.html#document-databases"><i class="fa fa-check"></i><b>7.3.3</b> Document Databases</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="resourcer.html"><a href="resourcer.html#computation-resources"><i class="fa fa-check"></i><b>7.4</b> Computation Resources</a></li>
<li class="chapter" data-level="7.5" data-path="resourcer.html"><a href="resourcer.html#interacting-with-r-resources"><i class="fa fa-check"></i><b>7.5</b> Interacting with R Resources</a></li>
<li class="chapter" data-level="7.6" data-path="resourcer.html"><a href="resourcer.html#extending_resources"><i class="fa fa-check"></i><b>7.6</b> Extending Resources</a></li>
<li class="chapter" data-level="7.7" data-path="resourcer.html"><a href="resourcer.html#resource-forms"><i class="fa fa-check"></i><b>7.7</b> Resource Forms</a></li>
<li class="chapter" data-level="7.8" data-path="resourcer.html"><a href="resourcer.html#examples-of-resources"><i class="fa fa-check"></i><b>7.8</b> Examples of Resources</a></li>
<li class="chapter" data-level="7.9" data-path="resourcer.html"><a href="resourcer.html#using-resources-with-datashield"><i class="fa fa-check"></i><b>7.9</b> Using Resources with DataSHIELD</a>
<ul>
<li class="chapter" data-level="7.9.1" data-path="resourcer.html"><a href="resourcer.html#tsv-file-into-a-tibble-or-data.frame"><i class="fa fa-check"></i><b>7.9.1</b> TSV File into a tibble or data.frame</a></li>
<li class="chapter" data-level="7.9.2" data-path="resourcer.html"><a href="resourcer.html#r-data-file-into-a-r-object"><i class="fa fa-check"></i><b>7.9.2</b> R Data File into a R object</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>III Resources Extensions</b></span></li>
<li class="chapter" data-level="8" data-path="omic-extension.html"><a href="omic-extension.html"><i class="fa fa-check"></i><b>8</b> Extension to VCF files to peform GWAS with Bioconductor</a></li>
<li class="chapter" data-level="9" data-path="shell-extension.html"><a href="shell-extension.html"><i class="fa fa-check"></i><b>9</b> Extension to secure shell programs: GWAS with PLINK</a></li>
<li class="chapter" data-level="10" data-path="extension-to-ga4gh-and-ega-bamvcf-files.html"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html"><i class="fa fa-check"></i><b>10</b> Extension to GA4GH and EGA (BAM/VCF files)</a>
<ul>
<li class="chapter" data-level="10.1" data-path="extension-to-ga4gh-and-ega-bamvcf-files.html"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#upload-ga4gh-and-ega-resources-into-opal"><i class="fa fa-check"></i><b>10.1</b> Upload GA4GH and EGA resources into OPAL</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="extension-to-ga4gh-and-ega-bamvcf-files.html"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#manually"><i class="fa fa-check"></i><b>10.1.1</b> Manually</a></li>
<li class="chapter" data-level="10.1.2" data-path="extension-to-ga4gh-and-ega-bamvcf-files.html"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#using-r"><i class="fa fa-check"></i><b>10.1.2</b> Using R</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="extension-to-ga4gh-and-ega-bamvcf-files.html"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#adding-a-genomic-resource"><i class="fa fa-check"></i><b>10.2</b> Adding a genomic resource</a></li>
<li class="chapter" data-level="10.3" data-path="extension-to-ga4gh-and-ega-bamvcf-files.html"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#descriptive-omic-data-analysis"><i class="fa fa-check"></i><b>10.3</b> Descriptive omic data analysis</a></li>
</ul></li>
<li class="part"><span><b>IV Workflows</b></span></li>
<li class="chapter" data-level="11" data-path="setup.html"><a href="setup.html"><i class="fa fa-check"></i><b>11</b> Setup</a>
<ul>
<li class="chapter" data-level="11.1" data-path="setup.html"><a href="setup.html#providing-datashield-packages-in-the-opal-server"><i class="fa fa-check"></i><b>11.1</b> Providing DataSHIELD packages in the opal server</a></li>
<li class="chapter" data-level="11.2" data-path="setup.html"><a href="setup.html#required-r-packages-in-the-client-site-e.g.-local-machine"><i class="fa fa-check"></i><b>11.2</b> Required R Packages in the client site (e.g. local machine)</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="basic-statistical-analyses.html"><a href="basic-statistical-analyses.html"><i class="fa fa-check"></i><b>12</b> Basic statistical analyses</a>
<ul>
<li class="chapter" data-level="12.1" data-path="basic-statistical-analyses.html"><a href="basic-statistical-analyses.html#analysis-from-a-single-study"><i class="fa fa-check"></i><b>12.1</b> Analysis from a single study</a></li>
<li class="chapter" data-level="12.2" data-path="basic-statistical-analyses.html"><a href="basic-statistical-analyses.html#analysis-from-a-multiple-studies"><i class="fa fa-check"></i><b>12.2</b> Analysis from a multiple studies</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="Omic.html"><a href="Omic.html"><i class="fa fa-check"></i><b>13</b> Omic data analysis: types of implemented analyses</a></li>
<li class="chapter" data-level="14" data-path="differential-gene-expression-dge-analysis.html"><a href="differential-gene-expression-dge-analysis.html"><i class="fa fa-check"></i><b>14</b> Differential gene expression (DGE) analysis</a></li>
<li class="chapter" data-level="15" data-path="epigenome-wide-association-analysis-ewas.html"><a href="epigenome-wide-association-analysis-ewas.html"><i class="fa fa-check"></i><b>15</b> Epigenome-wide association analysis (EWAS)</a>
<ul>
<li class="chapter" data-level="15.1" data-path="epigenome-wide-association-analysis-ewas.html"><a href="epigenome-wide-association-analysis-ewas.html#single-cpg-analysis"><i class="fa fa-check"></i><b>15.1</b> Single CpG analysis</a></li>
<li class="chapter" data-level="15.2" data-path="epigenome-wide-association-analysis-ewas.html"><a href="epigenome-wide-association-analysis-ewas.html#multiple-cpg-analysis"><i class="fa fa-check"></i><b>15.2</b> Multiple CpG analysis</a></li>
<li class="chapter" data-level="15.3" data-path="epigenome-wide-association-analysis-ewas.html"><a href="epigenome-wide-association-analysis-ewas.html#adjusting-for-surrogate-variables"><i class="fa fa-check"></i><b>15.3</b> Adjusting for Surrogate Variables</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="genome-wide-association-analyses-gwas.html"><a href="genome-wide-association-analyses-gwas.html"><i class="fa fa-check"></i><b>16</b> Genome-wide association analyses (GWAS)</a>
<ul>
<li class="chapter" data-level="16.1" data-path="genome-wide-association-analyses-gwas.html"><a href="genome-wide-association-analyses-gwas.html#gwasBioc"><i class="fa fa-check"></i><b>16.1</b> GWAS with Bioconductor</a></li>
<li class="chapter" data-level="16.2" data-path="genome-wide-association-analyses-gwas.html"><a href="genome-wide-association-analyses-gwas.html#gwas-with-plink"><i class="fa fa-check"></i><b>16.2</b> GWAS with PLINK</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="GIS.html"><a href="GIS.html"><i class="fa fa-check"></i><b>17</b> Geospatial data analysis</a>
<ul>
<li class="chapter" data-level="17.1" data-path="GIS.html"><a href="GIS.html#introducing-the-analysis"><i class="fa fa-check"></i><b>17.1</b> Introducing the analysis</a></li>
<li class="chapter" data-level="17.2" data-path="GIS.html"><a href="GIS.html#setting-up-the-analysis"><i class="fa fa-check"></i><b>17.2</b> Setting up the analysis</a></li>
<li class="chapter" data-level="17.3" data-path="GIS.html"><a href="GIS.html#data-manipulation"><i class="fa fa-check"></i><b>17.3</b> Data manipulation</a></li>
<li class="chapter" data-level="17.4" data-path="GIS.html"><a href="GIS.html#generating-the-final-result"><i class="fa fa-check"></i><b>17.4</b> Generating the final result</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="exposome.html"><a href="exposome.html"><i class="fa fa-check"></i><b>18</b> Exposome data analysis</a>
<ul>
<li class="chapter" data-level="18.1" data-path="exposome.html"><a href="exposome.html#loading-the-exposome-set"><i class="fa fa-check"></i><b>18.1</b> Loading the Exposome Set</a>
<ul>
<li class="chapter" data-level="18.1.1" data-path="exposome.html"><a href="exposome.html#from-tables-to-exposomeset"><i class="fa fa-check"></i><b>18.1.1</b> From tables to <code>ExposomeSet</code></a></li>
<li class="chapter" data-level="18.1.2" data-path="exposome.html"><a href="exposome.html#from-resource-to-exposomeset"><i class="fa fa-check"></i><b>18.1.2</b> From resource to <code>ExposomeSet</code></a></li>
</ul></li>
<li class="chapter" data-level="18.2" data-path="exposome.html"><a href="exposome.html#exploring-the-loaded-dataset"><i class="fa fa-check"></i><b>18.2</b> Exploring the loaded dataset</a>
<ul>
<li class="chapter" data-level="18.2.1" data-path="exposome.html"><a href="exposome.html#family-names"><i class="fa fa-check"></i><b>18.2.1</b> Family names</a></li>
<li class="chapter" data-level="18.2.2" data-path="exposome.html"><a href="exposome.html#exposures-and-phenotypes-names"><i class="fa fa-check"></i><b>18.2.2</b> Exposures and phenotypes names</a></li>
<li class="chapter" data-level="18.2.3" data-path="exposome.html"><a href="exposome.html#summary-of-variables"><i class="fa fa-check"></i><b>18.2.3</b> Summary of variables</a></li>
<li class="chapter" data-level="18.2.4" data-path="exposome.html"><a href="exposome.html#missing-data"><i class="fa fa-check"></i><b>18.2.4</b> Missing data</a></li>
<li class="chapter" data-level="18.2.5" data-path="exposome.html"><a href="exposome.html#exposures-normality"><i class="fa fa-check"></i><b>18.2.5</b> Exposures Normality</a></li>
<li class="chapter" data-level="18.2.6" data-path="exposome.html"><a href="exposome.html#exposures-behaviour"><i class="fa fa-check"></i><b>18.2.6</b> Exposures Behaviour</a></li>
</ul></li>
<li class="chapter" data-level="18.3" data-path="exposome.html"><a href="exposome.html#exposures-imputation"><i class="fa fa-check"></i><b>18.3</b> Exposures Imputation</a></li>
<li class="chapter" data-level="18.4" data-path="exposome.html"><a href="exposome.html#performing-an-exwas"><i class="fa fa-check"></i><b>18.4</b> Performing an ExWAS</a></li>
<li class="chapter" data-level="18.5" data-path="exposome.html"><a href="exposome.html#exposures-pca"><i class="fa fa-check"></i><b>18.5</b> Exposures PCA</a></li>
<li class="chapter" data-level="18.6" data-path="exposome.html"><a href="exposome.html#exposures-correlation"><i class="fa fa-check"></i><b>18.6</b> Exposures Correlation</a></li>
</ul></li>
<li class="part"><span><b>V Developers</b></span></li>
<li class="chapter" data-level="19" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html"><i class="fa fa-check"></i><b>19</b> DSLite: DataSHIELD Implementation on Local Datasets</a>
<ul>
<li class="chapter" data-level="19.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#development-environment-setup"><i class="fa fa-check"></i><b>19.1</b> Development Environment Setup</a>
<ul>
<li class="chapter" data-level="19.1.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-packages"><i class="fa fa-check"></i><b>19.1.1</b> DataSHIELD Packages</a></li>
<li class="chapter" data-level="19.1.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#test-datasets"><i class="fa fa-check"></i><b>19.1.2</b> Test Datasets</a></li>
<li class="chapter" data-level="19.1.3" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#r-package-development-tools"><i class="fa fa-check"></i><b>19.1.3</b> R Package Development Tools</a></li>
</ul></li>
<li class="chapter" data-level="19.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-development-flow"><i class="fa fa-check"></i><b>19.2</b> DataSHIELD Development Flow</a>
<ul>
<li class="chapter" data-level="19.2.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#dsliteserver"><i class="fa fa-check"></i><b>19.2.1</b> DSLiteServer</a></li>
<li class="chapter" data-level="19.2.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-configuration"><i class="fa fa-check"></i><b>19.2.2</b> DataSHIELD Configuration</a></li>
</ul></li>
<li class="chapter" data-level="19.3" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-sessions"><i class="fa fa-check"></i><b>19.3</b> DataSHIELD Sessions</a></li>
<li class="chapter" data-level="19.4" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#debugging"><i class="fa fa-check"></i><b>19.4</b> Debugging</a></li>
<li class="chapter" data-level="19.5" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#limitations"><i class="fa fa-check"></i><b>19.5</b> Limitations</a>
<ul>
<li class="chapter" data-level="19.5.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#function-parameters-parser"><i class="fa fa-check"></i><b>19.5.1</b> Function Parameters Parser</a></li>
<li class="chapter" data-level="19.5.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#server-side-environments"><i class="fa fa-check"></i><b>19.5.2</b> Server Side Environments</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="20" data-path="creating-datashield-packages.html"><a href="creating-datashield-packages.html"><i class="fa fa-check"></i><b>20</b> Creating DataSHIELD packages</a></li>
<li class="chapter" data-level="21" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html"><i class="fa fa-check"></i><b>21</b> Tips and tricks</a>
<ul>
<li class="chapter" data-level="21.1" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#how-to-create-a-new-project-into-opal"><i class="fa fa-check"></i><b>21.1</b> How to create a new project into OPAL</a>
<ul>
<li class="chapter" data-level="21.1.1" data-path="extension-to-ga4gh-and-ega-bamvcf-files.html"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#manually"><i class="fa fa-check"></i><b>21.1.1</b> Manually</a></li>
<li class="chapter" data-level="21.1.2" data-path="extension-to-ga4gh-and-ega-bamvcf-files.html"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#using-r"><i class="fa fa-check"></i><b>21.1.2</b> Using R</a></li>
</ul></li>
<li class="chapter" data-level="21.2" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#how-to-upload-a-new-resource-into-opal"><i class="fa fa-check"></i><b>21.2</b> How to upload a new resource into OPAL</a>
<ul>
<li class="chapter" data-level="21.2.1" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#create_manually"><i class="fa fa-check"></i><b>21.2.1</b> Manually</a></li>
<li class="chapter" data-level="21.2.2" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#using-r-1"><i class="fa fa-check"></i><b>21.2.2</b> Using R</a></li>
</ul></li>
<li class="chapter" data-level="21.3" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#how-to-install-datashield-packages-into-opal-server"><i class="fa fa-check"></i><b>21.3</b> How to install DataSHIELD packages into OPAL server</a>
<ul>
<li class="chapter" data-level="21.3.1" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#manually-1"><i class="fa fa-check"></i><b>21.3.1</b> Manually</a></li>
<li class="chapter" data-level="21.3.2" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#using-r-2"><i class="fa fa-check"></i><b>21.3.2</b> Using R</a></li>
</ul></li>
<li class="chapter" data-level="21.4" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#how-to-install-r-packages-into-opal-server"><i class="fa fa-check"></i><b>21.4</b> How to install R packages into OPAL server</a>
<ul>
<li class="chapter" data-level="21.4.1" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#manually-2"><i class="fa fa-check"></i><b>21.4.1</b> Manually</a></li>
<li class="chapter" data-level="21.4.2" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#using-r-3"><i class="fa fa-check"></i><b>21.4.2</b> Using R</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="22" data-path="advancedtechnical-questions.html"><a href="advancedtechnical-questions.html"><i class="fa fa-check"></i><b>22</b> Advanced/technical questions</a></li>
<li class="part"><span><b>VI Appendix</b></span></li>
<li class="chapter" data-level="23" data-path="session-info.html"><a href="session-info.html"><i class="fa fa-check"></i><b>23</b> Session Info</a></li>
<li class="chapter" data-level="24" data-path="contributors.html"><a href="contributors.html"><i class="fa fa-check"></i><b>24</b> Contributors</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating non-disclosive big data analyses of data from different resources with R and DataSHIELD</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="extension-to-ga4gh-and-ega-bamvcf-files" class="section level1" number="10">
<h1><span class="header-section-number">10</span> Extension to GA4GH and EGA (BAM/VCF files)</h1>
<p>The Global Alliance for Genomics and Health (GA4GH) has a wide <a href="https://www.ga4gh.org/genomic-data-toolkit/">toolkit</a> focused on many aspects of the study of the genome. One of the aspects is the distribution of data, which ensures safe transfer protocols for data providers to grant access to their data.</p>
<p>The <a href="https://samtools.github.io/hts-specs/htsget.html">htsget API</a> is a standard developed by GA4GH which focuses on the streaming of BAM (Binary version of a SAM file) and VCF (Variant Call Format) files. In order to connect to open data servers that have the htsget API implemented, a <a href="resourcer.html#extending_resources">Client and a Resolver</a> have been located in the <em><a href="https://bioconductor.org/packages/3.11/dsOmics">dsOmics</a></em> package. In order to integrate this resource into the <em><a href="https://bioconductor.org/packages/3.11/dsOmics">dsOmics</a></em> analysis pipeline, the BAM and VCF files are treated using the functions <code>snpgdsVCF2GDS</code> and <code>BAM2VCF</code> to obtain a <a href="https://bioconductor.org/packages/release/bioc/vignettes/gdsfmt/inst/doc/gdsfmt.html#introduction">Genomic Data Storage</a> (GDS) object.</p>
<p>The generic resource implemented for the GA4GH htsget API allows the connection to open databases, federated databases that have the same protocol implemented rely on different authentication processes in order to allow the streaming of data files, for that reason they have to be implemented on a case to case basis.</p>
<p>The European Genome-Phenome Archive (EGA) is a federated database that has the htsget API implemented. In order to stream data, an authentication using <a href="https://oauth.net/2/">OAuth2.0</a> is required. Following the general htsget resource implementation, a <a href="resourcer.html#extending_resources">Client and a Resolver</a> have been located in the <em><a href="https://bioconductor.org/packages/3.11/dsOmics">dsOmics</a></em> package. The difference to use this resource is that it does not require an URL, as the EGA-htsget endpoint is static, and that a user and password has to be provided to perform the authentication to EGA.</p>
<div id="upload-ga4gh-and-ega-resources-into-opal" class="section level2" number="10.1">
<h2><span class="header-section-number">10.1</span> Upload GA4GH and EGA resources into OPAL</h2>
<div id="manually" class="section level3" number="10.1.1">
<h3><span class="header-section-number">10.1.1</span> Manually</h3>
<div id="Manual-GA4GH" class="section level4" number="10.1.1.1">
<h4><span class="header-section-number">10.1.1.1</span> GA4GH</h4>
<p>To upload a GA4GH htsget resource into OPAL, the first steps to follow are already described <a href="tips-and-tricks.html#create_manually">here</a>. The particularities of GA4GH htsget resources come to the last step, when inputing the resource parameters.</p>
<p>The resource category to select is <em>GA4GH htsget</em>. The resouce has to be named and an optional description can be provided if desired (Illustrated on the Figure <a href="extension-to-ga4gh-and-ega-bamvcf-files.html#fig:htsget1">10.1</a>).</p>
<div class="figure" style="text-align: center"><span id="fig:htsget1"></span>
<p class="caption">
Figure 10.1: Naming a GA4GH htsget resource
</p>
<img src="fig/opal_resource_htsget1.png" alt="Naming a GA4GH htsget resource" width="630" height="5%" />
</div>
<p>Once named, some additional paramters have to be inputed to define the resource (all of them are mandatory):</p>
<ul>
<li>Host: URL of the server to which stream data.</li>
<li>File ID: Name of the VCF/BAM file on the server.</li>
<li>Chromosome: Name of the chromosome to filter the selected file. The enconding has to match the selected file, if the file uses UCSC enconding use “chr1” if it uses NCBI use “1”. Contact the data provider if in doubt.</li>
<li>Start: The start position of the filter.</li>
<li>End: The end position of the filter.</li>
<li>Format: Format of the file, this field can take two values
<ul>
<li>VCF: To select a filtered VCF file</li>
<li>BAM: To select a filtered BAM file</li>
</ul></li>
</ul>
<p>This parameters are illustrated on the Figure <a href="extension-to-ga4gh-and-ega-bamvcf-files.html#fig:htsget2">10.2</a></p>
<div class="figure" style="text-align: center"><span id="fig:htsget2"></span>
<p class="caption">
Figure 10.2: Selecting the parameters on a GA4GH htsget resource
</p>
<img src="fig/opal_resource_htsget2.png" alt="Selecting the parameters on a GA4GH htsget resource" width="630" height="5%" />
</div>
</div>
<div id="Manual-EGA" class="section level4" number="10.1.1.2">
<h4><span class="header-section-number">10.1.1.2</span> EGA</h4>
<p>To upload a EGA htsget resource into OPAL, the first steps to follow are already described <a href="tips-and-tricks.html#create_manually">here</a>. The particularities of EGA htsget resources come to the last step, when inputing the resource parameters.</p>
<p>The resource category to select is <em>EGA database</em>. The resouce has to be named and an optional description can be provided if desired (Illustrated on the Figure <a href="extension-to-ga4gh-and-ega-bamvcf-files.html#fig:htsgetEGA1">10.3</a>).</p>
<div class="figure" style="text-align: center"><span id="fig:htsgetEGA1"></span>
<p class="caption">
Figure 10.3: Naming a EGA database resource
</p>
<img src="fig/opal_resource_htsgetEGA1.png" alt="Naming a EGA database resource" width="626" height="5%" />
</div>
<p>Once named, some additional paramters have to be inputed to define the resource (all of them are mandatory):</p>
<ul>
<li>File ID: Name of the VCF/BAM file on EGA.</li>
<li>Chromosome: Name of the chromosome to filter the selected file. The enconding has to match the selected file, if the file uses UCSC enconding use “chr1” if it uses NCBI use “1”. Contact the data provider if in doubt.</li>
<li>Start: The start position of the filter.</li>
<li>End: The end position of the filter.</li>
<li>Format: Format of the file, this field can take two values
<ul>
<li>VCF: To select a filtered VCF file</li>
<li>BAM: To select a filtered BAM file</li>
</ul></li>
</ul>
<p>This parameters are illustrated on the Figure <a href="extension-to-ga4gh-and-ega-bamvcf-files.html#fig:htsgetEGA2">10.4</a></p>
<div class="figure" style="text-align: center"><span id="fig:htsgetEGA2"></span>
<p class="caption">
Figure 10.4: Selecting the parameters on a EGA database resource
</p>
<img src="fig/opal_resource_htsgetEGA2.png" alt="Selecting the parameters on a EGA database resource" width="602" height="5%" />
</div>
<p>Access to the EGA database is controlled via a user and password combination that grants access to different files on the server. For that, they are mandatory to create the resource. They have to be inputed on the <em>Credentials</em> tab, illustrated on the Figure <a href="extension-to-ga4gh-and-ega-bamvcf-files.html#fig:htsgetEGA3">10.5</a>.</p>
<div class="figure" style="text-align: center"><span id="fig:htsgetEGA3"></span>
<p class="caption">
Figure 10.5: Credentials tab for EGA database resource
</p>
<img src="fig/opal_resource_htsgetEGA3.png" alt="Credentials tab for EGA database resource" width="627" height="5%" />
</div>
</div>
</div>
<div id="using-r" class="section level3" number="10.1.2">
<h3><span class="header-section-number">10.1.2</span> Using R</h3>
<div id="ga4gh" class="section level4" number="10.1.2.1">
<h4><span class="header-section-number">10.1.2.1</span> GA4GH</h4>
<p>A GA4GH htsget can be added to a given OPAL project by a simple function call. This process requires to exactly write the the URL where the genomic data file is located that in our case is:</p>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb112-1"></a><span class="co"># VCF FILES</span></span>
<span id="cb112-2"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb112-2"></a>hostURL<span class="op">/</span>reads<span class="op">/</span>fileID?format=BAM<span class="op">&amp;</span>referenceName=regName<span class="op">&amp;</span>start=start<span class="op">&amp;</span>end=end</span>
<span id="cb112-3"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb112-3"></a></span>
<span id="cb112-4"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb112-4"></a>Example<span class="op">:</span><span class="st"> </span>https<span class="op">:</span><span class="er">//</span>htsget.ga4gh.org<span class="op">/</span>reads<span class="op">/</span>giab.NA12878.NIST7035<span class="fl">.2</span>?format=BAM<span class="op">&amp;</span>referenceName=chr21<span class="op">&amp;</span>start=<span class="dv">1</span><span class="op">&amp;</span>end=<span class="dv">20000000</span></span>
<span id="cb112-5"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb112-5"></a></span>
<span id="cb112-6"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb112-6"></a><span class="co"># BAM FILES</span></span>
<span id="cb112-7"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb112-7"></a>hostURL<span class="op">/</span>variants<span class="op">/</span>fileID?format=VCF<span class="op">&amp;</span>referenceName=regName<span class="op">&amp;</span>start=start<span class="op">&amp;</span>end=end</span>
<span id="cb112-8"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb112-8"></a></span>
<span id="cb112-9"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb112-9"></a>Example<span class="op">:</span><span class="st"> </span>https<span class="op">:</span><span class="er">//</span>htsget.ga4gh.org<span class="op">/</span>variants<span class="op">/</span>1000genomes.phase1.chr1?format=VCF<span class="op">&amp;</span>referenceName=<span class="dv">1</span><span class="op">&amp;</span>start=<span class="dv">15000</span><span class="op">&amp;</span>end=<span class="dv">16000</span></span></code></pre></div>
<p>This can be a little bit tricky. If so, this can be done using the <a href="extension-to-ga4gh-and-ega-bamvcf-files.html#Manual-GA4GH">API as previously described</a>.</p>
<p>Then, the R code to create this resource from the VCF file in OPAL is:</p>
<div class="sourceCode" id="cb113"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb113-1"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb113-1"></a><span class="kw">library</span>(opalr)</span>
<span id="cb113-2"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb113-2"></a>o &lt;-<span class="st"> </span><span class="kw">opal.login</span>(<span class="st">&quot;administrator&quot;</span>,<span class="st">&quot;password&quot;</span>, </span>
<span id="cb113-3"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb113-3"></a>                <span class="dt">url=</span><span class="st">&quot;https://opal-demo.obiba.org&quot;</span>)</span>
<span id="cb113-4"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb113-4"></a></span>
<span id="cb113-5"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb113-5"></a><span class="kw">opal.resource_create</span>(o, <span class="st">&quot;RSRC&quot;</span>, <span class="st">&quot; ga4gh_1000g&quot;</span>, </span>
<span id="cb113-6"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb113-6"></a>  <span class="dt">url =</span> <span class="st">&quot;https://htsget.ga4gh.org/variants/1000genomes.phase1.chr1?format=VCF&amp;referenceName=1&amp;start=15000&amp;end=16000&quot;</span>, </span>
<span id="cb113-7"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb113-7"></a>  <span class="dt">format =</span> <span class="st">&quot;GA4GHVCF&quot;</span>)</span>
<span id="cb113-8"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb113-8"></a></span>
<span id="cb113-9"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb113-9"></a><span class="co"># to test the resource assignment</span></span>
<span id="cb113-10"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb113-10"></a><span class="kw">opal.assign.resource</span>(o, <span class="st">&quot;client&quot;</span>, <span class="st">&quot;RSRC. ga4gh_1000g&quot;</span>)</span>
<span id="cb113-11"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb113-11"></a><span class="kw">opal.execute</span>(o, <span class="st">&quot;class(client)&quot;</span>)</span>
<span id="cb113-12"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb113-12"></a></span>
<span id="cb113-13"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb113-13"></a><span class="kw">opal.logout</span>(o)</span></code></pre></div>
</div>
<div id="ega" class="section level4" number="10.1.2.2">
<h4><span class="header-section-number">10.1.2.2</span> EGA</h4>
<p>A EGA database can be added to a given OPAL project by a simple function call. This process requires to exactly write the the URL where the genomic data file is located that in our case is:</p>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb114-1"></a><span class="co"># Same URL style for BAM and VCF</span></span>
<span id="cb114-2"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb114-2"></a>https<span class="op">:</span><span class="er">//</span>ega.ebi.ac.uk<span class="op">:</span><span class="dv">8052</span><span class="op">/</span>elixir<span class="op">/</span>tickets<span class="op">/</span>tickets<span class="op">/</span>fileID?referenceName=refName<span class="op">&amp;</span>start=start<span class="op">&amp;</span>end=end</span>
<span id="cb114-3"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb114-3"></a></span>
<span id="cb114-4"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb114-4"></a>Example<span class="op">:</span><span class="st"> </span>https<span class="op">:</span><span class="er">//</span>ega.ebi.ac.uk<span class="op">:</span><span class="dv">8052</span><span class="op">/</span>elixir<span class="op">/</span>tickets<span class="op">/</span>tickets<span class="op">/</span>EGAF00001753756?referenceName=chr21<span class="op">&amp;</span>start=<span class="dv">1</span><span class="op">&amp;</span>end=<span class="dv">100</span></span></code></pre></div>
<p>This can be a little bit tricky. If so, this can be done using the <a href="extension-to-ga4gh-and-ega-bamvcf-files.html#Manual-EGA">API as previously described</a>.</p>
<p>Then, the R code to create this resource in OPAL is:</p>
<div class="sourceCode" id="cb115"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb115-1"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb115-1"></a><span class="kw">library</span>(opalr)</span>
<span id="cb115-2"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb115-2"></a>o &lt;-<span class="st"> </span><span class="kw">opal.login</span>(<span class="st">&quot;administrator&quot;</span>,<span class="st">&quot;password&quot;</span>, </span>
<span id="cb115-3"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb115-3"></a>                <span class="dt">url=</span><span class="st">&quot;https://opal-demo.obiba.org&quot;</span>)</span>
<span id="cb115-4"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb115-4"></a></span>
<span id="cb115-5"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb115-5"></a><span class="kw">opal.resource_create</span>(o, <span class="st">&quot;RSRC&quot;</span>, <span class="st">&quot;EGA&quot;</span>, </span>
<span id="cb115-6"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb115-6"></a>  <span class="dt">url =</span> <span class="st">&quot;https://ega.ebi.ac.uk:8052/elixir/tickets/tickets/EGAF00001753756?referenceName=chr21&amp;start=1&amp;end=100&quot;</span>, </span>
<span id="cb115-7"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb115-7"></a>  <span class="dt">format =</span> <span class="st">&quot;EGAhtsgetBAM&quot;</span>, <span class="co"># format = &quot;EGAhtsgetVCF&quot; for VCF files</span></span>
<span id="cb115-8"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb115-8"></a>  <span class="dt">identity =</span> <span class="st">&quot;ega-test-data@ebi.ac.uk&quot;</span>, <span class="co"># This is the test user</span></span>
<span id="cb115-9"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb115-9"></a>  <span class="dt">secret =</span> <span class="st">&quot;egarocks&quot;</span> <span class="co"># This is the test password</span></span>
<span id="cb115-10"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb115-10"></a>  ) </span>
<span id="cb115-11"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb115-11"></a></span>
<span id="cb115-12"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb115-12"></a><span class="co"># to test the resource assignment</span></span>
<span id="cb115-13"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb115-13"></a><span class="kw">opal.assign.resource</span>(o, <span class="st">&quot;client&quot;</span>, <span class="st">&quot;RSRC.EGA&quot;</span>)</span>
<span id="cb115-14"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb115-14"></a><span class="kw">opal.execute</span>(o, <span class="st">&quot;class(client)&quot;</span>)</span>
<span id="cb115-15"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb115-15"></a></span>
<span id="cb115-16"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb115-16"></a><span class="kw">opal.logout</span>(o)</span></code></pre></div>
</div>
</div>
</div>
<div id="adding-a-genomic-resource" class="section level2" number="10.2">
<h2><span class="header-section-number">10.2</span> Adding a genomic resource</h2>
<p>To load the presented resources into the study servers, the standard procedure already presented on this Bookdown has to be followed (see <a href="">here</a>).</p>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb116-1"></a><span class="kw">library</span>(DSI)</span>
<span id="cb116-2"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb116-2"></a><span class="kw">library</span>(DSOpal)</span>
<span id="cb116-3"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb116-3"></a><span class="kw">library</span>(dsBaseClient)</span>
<span id="cb116-4"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb116-4"></a><span class="kw">library</span>(dsOmicsClient)</span>
<span id="cb116-5"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb116-5"></a></span>
<span id="cb116-6"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb116-6"></a>builder &lt;-<span class="st"> </span><span class="kw">newDSLoginBuilder</span>()</span>
<span id="cb116-7"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb116-7"></a>builder<span class="op">$</span><span class="kw">append</span>(<span class="dt">server =</span> <span class="st">&quot;study1&quot;</span>, <span class="dt">url =</span> <span class="st">&quot;https://opal-demo.obiba.org&quot;</span>,</span>
<span id="cb116-8"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb116-8"></a>               <span class="dt">user =</span> <span class="st">&quot;dsuser&quot;</span>, <span class="dt">password =</span> <span class="st">&quot;password&quot;</span>,</span>
<span id="cb116-9"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb116-9"></a>               <span class="dt">resource =</span> <span class="st">&quot;RSRC.ga4gh_1000g&quot;</span>, <span class="dt">driver =</span> <span class="st">&quot;OpalDriver&quot;</span>)</span>
<span id="cb116-10"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb116-10"></a>logindata &lt;-<span class="st"> </span>builder<span class="op">$</span><span class="kw">build</span>()</span>
<span id="cb116-11"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb116-11"></a></span>
<span id="cb116-12"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb116-12"></a>conns &lt;-<span class="st"> </span><span class="kw">datashield.login</span>(<span class="dt">logins =</span> logindata, <span class="dt">assign =</span> <span class="ot">TRUE</span>,</span>
<span id="cb116-13"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb116-13"></a>                          <span class="dt">symbol =</span> <span class="st">&quot;ga4gh&quot;</span>)</span>
<span id="cb116-14"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb116-14"></a><span class="kw">datashield.assign.expr</span>(conns, <span class="dt">symbol =</span> <span class="st">&quot;ga4gh_gds&quot;</span>, </span>
<span id="cb116-15"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb116-15"></a>                       <span class="dt">expr =</span> <span class="kw">quote</span>(<span class="kw">as.resource.object</span>(ga4gh)))</span></code></pre></div>
<p>Now we can see that the resource is loaded in the study server as a GDS</p>
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb117-1"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb117-1"></a><span class="kw">ds.class</span>(<span class="st">&quot;ga4gh_gds&quot;</span>)</span></code></pre></div>
<pre><code>$study1
[1] &quot;GdsGenotypeReader&quot;
attr(,&quot;package&quot;)
[1] &quot;GWASTools&quot;</code></pre>
</div>
<div id="descriptive-omic-data-analysis" class="section level2" number="10.3">
<h2><span class="header-section-number">10.3</span> Descriptive omic data analysis</h2>
<p>Since the presented resources load GDS objects into the study servers, the same procedures introduced on <a href="genome-wide-association-analyses-gwas.html#gwasBioc">this same bookdown</a> can be applied. As an example, a PCA can be performed on the loaded data from a GA4GH htsget server that holds the 1000 Genome dataset. This example resource contains SNPs of the first human chromosome ranging from the start of it to the position 100000.</p>
<div class="sourceCode" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb119-1"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb119-1"></a><span class="kw">ds.PCASNPS</span>(<span class="st">&quot;ga4gh_gds&quot;</span>)</span></code></pre></div>
<div class="figure"><span id="fig:pcaGA4GH"></span>
<p class="caption">
Figure 10.6: PCA over the 1000 genomes dataset available at GA4GH
</p>
<img src="fig/pcaGA4GH-1.png" alt="PCA over the 1000 genomes dataset available at GA4GH " width="672" />
</div>
<p>For further descriptive analysis such as Hardy-Weinberg Equilibrium testing and Allelic frequency, a phenotypes (or covariates) table has to be provided and bundled through the <code>ds.GenotypeData</code> function, this table has to have at least information about the gender of the individuals.</p>
<p>This closes the access to the resources</p>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="extension-to-ga4gh-and-ega-bamvcf-files.html#cb120-1"></a><span class="kw">datashield.logout</span>(conns)</span></code></pre></div>

</div>
</div>



            </section>

          </div>
        </div>
      </div>
<a href="shell-extension.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="setup.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
