<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>9 Secure shell programs: GWAS with PLINK | Orchestrating non-disclosive big data analyses of data from different resources with R and DataSHIELD</title>
  <meta name="description" content="9 Secure shell programs: GWAS with PLINK | Orchestrating non-disclosive big data analyses of data from different resources with R and DataSHIELD" />
  <meta name="generator" content="bookdown 0.19 and GitBook 2.6.7" />

  <meta property="og:title" content="9 Secure shell programs: GWAS with PLINK | Orchestrating non-disclosive big data analyses of data from different resources with R and DataSHIELD" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="9 Secure shell programs: GWAS with PLINK | Orchestrating non-disclosive big data analyses of data from different resources with R and DataSHIELD" />
  
  
  



<meta name="date" content="2020-05-25" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="omic-extension.html"/>
<link rel="next" href="setup.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Preamble</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#what-you-will-learn"><i class="fa fa-check"></i><b>1.1</b> What you will learn</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#preliminaries"><i class="fa fa-check"></i><b>1.2</b> Preliminaries</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#acknowledgments"><i class="fa fa-check"></i><b>1.3</b> Acknowledgments</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="learnR.html"><a href="learnR.html"><i class="fa fa-check"></i><b>2</b> Learning R</a></li>
<li class="chapter" data-level="3" data-path="BioC.html"><a href="BioC.html"><i class="fa fa-check"></i><b>3</b> Bioconductor Data Structures</a><ul>
<li class="chapter" data-level="3.1" data-path="BioC.html"><a href="BioC.html#snp-array-data"><i class="fa fa-check"></i><b>3.1</b> SNP Array Data</a></li>
<li class="chapter" data-level="3.2" data-path="BioC.html"><a href="BioC.html#expression-sets"><i class="fa fa-check"></i><b>3.2</b> Expression Sets</a></li>
<li class="chapter" data-level="3.3" data-path="BioC.html"><a href="BioC.html#genomic-ranges"><i class="fa fa-check"></i><b>3.3</b> Genomic Ranges</a></li>
<li class="chapter" data-level="3.4" data-path="BioC.html"><a href="BioC.html#summarized-experiments"><i class="fa fa-check"></i><b>3.4</b> Summarized Experiments</a></li>
<li class="chapter" data-level="3.5" data-path="BioC.html"><a href="BioC.html#ranged-summarized-experiments"><i class="fa fa-check"></i><b>3.5</b> Ranged Summarized Experiments</a></li>
<li class="chapter" data-level="3.6" data-path="BioC.html"><a href="BioC.html#multi-data-set"><i class="fa fa-check"></i><b>3.6</b> Multi Data Set</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="opal.html"><a href="opal.html"><i class="fa fa-check"></i><b>4</b> Opal</a><ul>
<li class="chapter" data-level="4.1" data-path="opal.html"><a href="opal.html#introduction-1"><i class="fa fa-check"></i><b>4.1</b> Introduction</a></li>
<li class="chapter" data-level="4.2" data-path="opal.html"><a href="opal.html#data-management"><i class="fa fa-check"></i><b>4.2</b> Data Management</a></li>
<li class="chapter" data-level="4.3" data-path="opal.html"><a href="opal.html#security"><i class="fa fa-check"></i><b>4.3</b> Security</a></li>
<li class="chapter" data-level="4.4" data-path="opal.html"><a href="opal.html#r-integration"><i class="fa fa-check"></i><b>4.4</b> R Integration</a></li>
<li class="chapter" data-level="4.5" data-path="opal.html"><a href="opal.html#demo"><i class="fa fa-check"></i><b>4.5</b> Demo</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="datashield.html"><a href="datashield.html"><i class="fa fa-check"></i><b>5</b> DataSHIELD</a><ul>
<li class="chapter" data-level="5.1" data-path="datashield.html"><a href="datashield.html#introduction-2"><i class="fa fa-check"></i><b>5.1</b> Introduction</a></li>
<li class="chapter" data-level="5.2" data-path="datashield.html"><a href="datashield.html#datashield-r-interface-dsi"><i class="fa fa-check"></i><b>5.2</b> DataSHIELD R Interface (DSI)</a><ul>
<li class="chapter" data-level="5.2.1" data-path="datashield.html"><a href="datashield.html#classes-structure"><i class="fa fa-check"></i><b>5.2.1</b> Classes Structure</a></li>
<li class="chapter" data-level="5.2.2" data-path="datashield.html"><a href="datashield.html#higher-level-functions"><i class="fa fa-check"></i><b>5.2.2</b> Higher Level Functions</a></li>
<li class="chapter" data-level="5.2.3" data-path="datashield.html"><a href="datashield.html#options"><i class="fa fa-check"></i><b>5.2.3</b> Options</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="datashield.html"><a href="datashield.html#datashieldopal-implementation"><i class="fa fa-check"></i><b>5.3</b> DataSHIELD/Opal Implementation</a><ul>
<li class="chapter" data-level="5.3.1" data-path="datashield.html"><a href="datashield.html#client"><i class="fa fa-check"></i><b>5.3.1</b> Client</a></li>
<li class="chapter" data-level="5.3.2" data-path="datashield.html"><a href="datashield.html#server"><i class="fa fa-check"></i><b>5.3.2</b> Server</a></li>
<li class="chapter" data-level="5.3.3" data-path="datashield.html"><a href="datashield.html#demo-1"><i class="fa fa-check"></i><b>5.3.3</b> Demo</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>II Focus Topics</b></span></li>
<li class="chapter" data-level="6" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>6</b> The Resources</a><ul>
<li class="chapter" data-level="6.1" data-path="resources.html"><a href="resources.html#concept"><i class="fa fa-check"></i><b>6.1</b> Concept</a></li>
<li class="chapter" data-level="6.2" data-path="resources.html"><a href="resources.html#types"><i class="fa fa-check"></i><b>6.2</b> Types</a></li>
<li class="chapter" data-level="6.3" data-path="resources.html"><a href="resources.html#definition"><i class="fa fa-check"></i><b>6.3</b> Definition</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="resourcer.html"><a href="resourcer.html"><i class="fa fa-check"></i><b>7</b> The resourcer Package</a><ul>
<li class="chapter" data-level="7.1" data-path="resourcer.html"><a href="resourcer.html#introduction-3"><i class="fa fa-check"></i><b>7.1</b> Introduction</a></li>
<li class="chapter" data-level="7.2" data-path="resourcer.html"><a href="resourcer.html#file-resources"><i class="fa fa-check"></i><b>7.2</b> File Resources</a><ul>
<li class="chapter" data-level="7.2.1" data-path="resourcer.html"><a href="resourcer.html#file-getter"><i class="fa fa-check"></i><b>7.2.1</b> File Getter</a></li>
<li class="chapter" data-level="7.2.2" data-path="resourcer.html"><a href="resourcer.html#file-data-format"><i class="fa fa-check"></i><b>7.2.2</b> File Data Format</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="resourcer.html"><a href="resourcer.html#database-resources"><i class="fa fa-check"></i><b>7.3</b> Database Resources</a><ul>
<li class="chapter" data-level="7.3.1" data-path="resourcer.html"><a href="resourcer.html#dbi-connectors"><i class="fa fa-check"></i><b>7.3.1</b> DBI Connectors</a></li>
<li class="chapter" data-level="7.3.2" data-path="resourcer.html"><a href="resourcer.html#use-dplyr"><i class="fa fa-check"></i><b>7.3.2</b> Use dplyr</a></li>
<li class="chapter" data-level="7.3.3" data-path="resourcer.html"><a href="resourcer.html#document-databases"><i class="fa fa-check"></i><b>7.3.3</b> Document Databases</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="resourcer.html"><a href="resourcer.html#computation-resources"><i class="fa fa-check"></i><b>7.4</b> Computation Resources</a></li>
<li class="chapter" data-level="7.5" data-path="resourcer.html"><a href="resourcer.html#interacting-with-r-resources"><i class="fa fa-check"></i><b>7.5</b> Interacting with R Resources</a></li>
<li class="chapter" data-level="7.6" data-path="resourcer.html"><a href="resourcer.html#extending_resources"><i class="fa fa-check"></i><b>7.6</b> Extending Resources</a></li>
<li class="chapter" data-level="7.7" data-path="resourcer.html"><a href="resourcer.html#resource-forms"><i class="fa fa-check"></i><b>7.7</b> Resource Forms</a></li>
<li class="chapter" data-level="7.8" data-path="resourcer.html"><a href="resourcer.html#examples-of-resources"><i class="fa fa-check"></i><b>7.8</b> Examples of Resources</a><ul>
<li class="chapter" data-level="7.8.1" data-path="resourcer.html"><a href="resourcer.html#tsv-file-into-a-tibble-or-data.frame"><i class="fa fa-check"></i><b>7.8.1</b> TSV File into a tibble or data.frame</a></li>
<li class="chapter" data-level="7.8.2" data-path="resourcer.html"><a href="resourcer.html#r-data-file-into-a-r-object"><i class="fa fa-check"></i><b>7.8.2</b> R Data File into a R object</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>III Resources Extensions</b></span></li>
<li class="chapter" data-level="8" data-path="omic-extension.html"><a href="omic-extension.html"><i class="fa fa-check"></i><b>8</b> VCF files to GDS to peform GWAS with Bioconductor</a></li>
<li class="chapter" data-level="9" data-path="shell-extension.html"><a href="shell-extension.html"><i class="fa fa-check"></i><b>9</b> Secure shell programs: GWAS with PLINK</a></li>
<li class="part"><span><b>IV Workflows</b></span></li>
<li class="chapter" data-level="10" data-path="setup.html"><a href="setup.html"><i class="fa fa-check"></i><b>10</b> Setup</a><ul>
<li class="chapter" data-level="10.1" data-path="setup.html"><a href="setup.html#providing-datashield-packages-in-the-opal-server"><i class="fa fa-check"></i><b>10.1</b> Providing DataSHIELD packages in the opal server</a></li>
<li class="chapter" data-level="10.2" data-path="setup.html"><a href="setup.html#required-r-packages-in-the-client-site-e.g.local-machine"><i class="fa fa-check"></i><b>10.2</b> Required R Packages in the client site (e.g. local machine)</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="basic-statistical-analyses.html"><a href="basic-statistical-analyses.html"><i class="fa fa-check"></i><b>11</b> Basic statistical analyses</a><ul>
<li class="chapter" data-level="11.1" data-path="basic-statistical-analyses.html"><a href="basic-statistical-analyses.html#analysis-from-a-single-study"><i class="fa fa-check"></i><b>11.1</b> Analysis from a single study</a></li>
<li class="chapter" data-level="11.2" data-path="basic-statistical-analyses.html"><a href="basic-statistical-analyses.html#analysis-from-a-multiple-studies"><i class="fa fa-check"></i><b>11.2</b> Analysis from a multiple studies</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="Omic.html"><a href="Omic.html"><i class="fa fa-check"></i><b>12</b> Omic data analysis</a><ul>
<li class="chapter" data-level="12.1" data-path="Omic.html"><a href="Omic.html#types-of-analyses-implemented"><i class="fa fa-check"></i><b>12.1</b> Types of analyses implemented</a></li>
<li class="chapter" data-level="12.2" data-path="Omic.html"><a href="Omic.html#differential-gene-expression-dge-analysis"><i class="fa fa-check"></i><b>12.2</b> Differential gene expression (DGE) analysis</a></li>
<li class="chapter" data-level="12.3" data-path="Omic.html"><a href="Omic.html#epigenome-wide-association-analysis-ewas"><i class="fa fa-check"></i><b>12.3</b> Epigenome-wide association analysis (EWAS)</a><ul>
<li class="chapter" data-level="12.3.1" data-path="Omic.html"><a href="Omic.html#single-cpg-analysis"><i class="fa fa-check"></i><b>12.3.1</b> Single CpG analysis</a></li>
<li class="chapter" data-level="12.3.2" data-path="Omic.html"><a href="Omic.html#multiple-cpg-analysis"><i class="fa fa-check"></i><b>12.3.2</b> Multiple CpG analysis</a></li>
<li class="chapter" data-level="12.3.3" data-path="Omic.html"><a href="Omic.html#adjusting-for-surrogate-variables"><i class="fa fa-check"></i><b>12.3.3</b> Adjusting for Surrogate Variables</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="Omic.html"><a href="Omic.html#gwas-with-bioconductor"><i class="fa fa-check"></i><b>12.4</b> GWAS with Bioconductor</a></li>
<li class="chapter" data-level="12.5" data-path="Omic.html"><a href="Omic.html#gwas-with-plink"><i class="fa fa-check"></i><b>12.5</b> GWAS with PLINK</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="GIS.html"><a href="GIS.html"><i class="fa fa-check"></i><b>13</b> Geospatial data analysis</a><ul>
<li class="chapter" data-level="13.1" data-path="GIS.html"><a href="GIS.html#introducing-the-analysis"><i class="fa fa-check"></i><b>13.1</b> Introducing the analysis</a></li>
<li class="chapter" data-level="13.2" data-path="GIS.html"><a href="GIS.html#setting-up-the-analysis"><i class="fa fa-check"></i><b>13.2</b> Setting up the analysis</a></li>
<li class="chapter" data-level="13.3" data-path="GIS.html"><a href="GIS.html#data-manipulation"><i class="fa fa-check"></i><b>13.3</b> Data manipulation</a></li>
<li class="chapter" data-level="13.4" data-path="GIS.html"><a href="GIS.html#generating-the-final-result"><i class="fa fa-check"></i><b>13.4</b> Generating the final result</a></li>
</ul></li>
<li class="part"><span><b>V Developers</b></span></li>
<li class="chapter" data-level="14" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html"><i class="fa fa-check"></i><b>14</b> DSLite: DataSHIELD Implementation on Local Datasets</a><ul>
<li class="chapter" data-level="14.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#development-environment-setup"><i class="fa fa-check"></i><b>14.1</b> Development Environment Setup</a><ul>
<li class="chapter" data-level="14.1.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-packages"><i class="fa fa-check"></i><b>14.1.1</b> DataSHIELD Packages</a></li>
<li class="chapter" data-level="14.1.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#test-datasets"><i class="fa fa-check"></i><b>14.1.2</b> Test Datasets</a></li>
<li class="chapter" data-level="14.1.3" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#r-package-development-tools"><i class="fa fa-check"></i><b>14.1.3</b> R Package Development Tools</a></li>
</ul></li>
<li class="chapter" data-level="14.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-development-flow"><i class="fa fa-check"></i><b>14.2</b> DataSHIELD Development Flow</a><ul>
<li class="chapter" data-level="14.2.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#dsliteserver"><i class="fa fa-check"></i><b>14.2.1</b> DSLiteServer</a></li>
<li class="chapter" data-level="14.2.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-configuration"><i class="fa fa-check"></i><b>14.2.2</b> DataSHIELD Configuration</a></li>
</ul></li>
<li class="chapter" data-level="14.3" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-sessions"><i class="fa fa-check"></i><b>14.3</b> DataSHIELD Sessions</a></li>
<li class="chapter" data-level="14.4" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#debugging"><i class="fa fa-check"></i><b>14.4</b> Debugging</a></li>
<li class="chapter" data-level="14.5" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#limitations"><i class="fa fa-check"></i><b>14.5</b> Limitations</a><ul>
<li class="chapter" data-level="14.5.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#function-parameters-parser"><i class="fa fa-check"></i><b>14.5.1</b> Function Parameters Parser</a></li>
<li class="chapter" data-level="14.5.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#server-side-environments"><i class="fa fa-check"></i><b>14.5.2</b> Server Side Environments</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="15" data-path="creating-datashield-packages.html"><a href="creating-datashield-packages.html"><i class="fa fa-check"></i><b>15</b> Creating DataSHIELD packages</a></li>
<li class="chapter" data-level="16" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html"><i class="fa fa-check"></i><b>16</b> Tips and tricks</a><ul>
<li class="chapter" data-level="16.1" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#how-to-install-r-packages-into-opal-server-from-r"><i class="fa fa-check"></i><b>16.1</b> How to install R packages into OPAL server from R</a></li>
<li class="chapter" data-level="16.2" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#how-to-check-whether-there-are-open-r-sesions-in-the-opal-server"><i class="fa fa-check"></i><b>16.2</b> How to check whether there are open R sesions in the OPAL server</a></li>
<li class="chapter" data-level="16.3" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#how"><i class="fa fa-check"></i><b>16.3</b> How</a></li>
</ul></li>
<li class="part"><span><b>VI Appendix</b></span></li>
<li class="chapter" data-level="17" data-path="session-info.html"><a href="session-info.html"><i class="fa fa-check"></i><b>17</b> Session Info</a></li>
<li class="chapter" data-level="18" data-path="contributors.html"><a href="contributors.html"><i class="fa fa-check"></i><b>18</b> Contributors</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating non-disclosive big data analyses of data from different resources with R and DataSHIELD</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="shell_extension" class="section level1">
<h1><span class="header-section-number">9</span> Secure shell programs: GWAS with PLINK</h1>
<p>GWAS can also be performed using program that are executed using shell commands. This is the case of <a href="http://zzz.bwh.harvard.edu/plink/">PLINK</a>, one of the state-of-the-art programs to run GWAS and other genomic data analyses such gene-enviroment interactions or polygenic risc score analyses that requires efficient and scalable pipelines.</p>
<p>The resources also allow the use of secure SSH service to run programs on a remote server accessible through ssh containing data and analysis tools where R is just used for launching the analyses and aggregating results. This feature allow us to create functions to analyze data using specific shell programs.</p>
<p>Here we describe how <a href="http://zzz.bwh.harvard.edu/plink/">PLINK</a> program can be use to perform GWAS [<strong>Yannick</strong> some overview about how this is created should be described]</p>
<p>We use this following code to illustrate how analyses should be performed using the <em><a href="https://CRAN.R-project.org/package=resourcer">resourcer</a></em> package. This code could be considered as the base code for creating a DataSHIELD package for the OPAL server as performed in <code>plinkDS()</code> function implemented in the <em><a href="https://github.com/isglobal-brge">dsOmics</a></em> package</p>
<p>We access the ssh resource called <code>brge_plink</code> (Figure <a href="resources.html#fig:testResources">6.2</a>) using the <em><a href="https://CRAN.R-project.org/package=resourcer">resourcer</a></em> package as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(resourcer)
brge_plink &lt;-<span class="st"> </span>resourcer<span class="op">::</span><span class="kw">newResource</span>(<span class="dt">url=</span><span class="st">&quot;ssh://plink-test.obiba.org:2222/home/master/brge?exec=ls,plink,plink1&quot;</span>, 
                                     <span class="dt">identity =</span> <span class="st">&quot;master&quot;</span>, <span class="dt">secret =</span> <span class="st">&quot;master&quot;</span>)
client &lt;-<span class="st"> </span>resourcer<span class="op">::</span><span class="kw">newResourceClient</span>(brge_plink)</code></pre></div>
<p>This creates an object of this class:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">class</span>(client)</code></pre></div>
<pre><code>[1] &quot;SshResourceClient&quot;     &quot;CommandResourceClient&quot; &quot;ResourceClient&quot;       
[4] &quot;R6&quot;                   </code></pre>
<p>These are the actions we can do with an <code>SshResourceClient</code> object</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(client)</code></pre></div>
<pre><code> [1] &quot;.__enclos_env__&quot;    &quot;clone&quot;              &quot;close&quot;             
 [4] &quot;exec&quot;               &quot;removeTempDir&quot;      &quot;tempDir&quot;           
 [7] &quot;uploadFile&quot;         &quot;downloadFile&quot;       &quot;getConnection&quot;     
[10] &quot;getAllowedCommands&quot; &quot;initialize&quot;         &quot;asTbl&quot;             
[13] &quot;asDataFrame&quot;        &quot;getResource&quot;       </code></pre>
<p>For this specific resource (e.g. PLINK) we can execute these shell commands</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">client<span class="op">$</span><span class="kw">getAllowedCommands</span>()</code></pre></div>
<pre><code>[1] &quot;ls&quot;     &quot;plink&quot;  &quot;plink1&quot;</code></pre>
<p>For instance</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">client<span class="op">$</span><span class="kw">exec</span>(<span class="st">&quot;ls&quot;</span>, <span class="st">&quot;-la&quot;</span>)</code></pre></div>
<pre><code>$status
[1] 0

$output
[1] &quot;total 92992&quot;                                              
[2] &quot;drwxr-xr-x 2 master master     4096 Apr 29 09:50 .&quot;       
[3] &quot;drwxr-xr-x 6 master master     4096 May 12 19:25 ..&quot;      
[4] &quot;-rw-r--r-- 1 master master 57800003 Apr 29 09:50 brge.bed&quot;
[5] &quot;-rw-r--r-- 1 master master  2781294 Apr 29 09:50 brge.bim&quot;
[6] &quot;-rw-r--r-- 1 master master    45771 Apr 29 09:50 brge.fam&quot;
[7] &quot;-rw-rw-r-- 1 master master 34442346 Apr 29 09:50 brge.gds&quot;
[8] &quot;-rw-r--r-- 1 master master    59802 Apr 29 09:50 brge.phe&quot;
[9] &quot;-rw-r--r-- 1 master master    72106 Apr 29 09:50 brge.txt&quot;

$error
character(0)

$command
[1] &quot;cd /home/master/brge &amp;&amp; ls -la&quot;

attr(,&quot;class&quot;)
[1] &quot;resource.exec&quot;</code></pre>
<p>Then, to avoid multiple accesses to the resource, it is recommended to create a temporal directory to save our results</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tempDir &lt;-<span class="st"> </span>client<span class="op">$</span><span class="kw">tempDir</span>()
tempDir</code></pre></div>
<pre><code>[1] &quot;/tmp/ssh-6184&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">client<span class="op">$</span><span class="kw">exec</span>(<span class="st">&quot;ls&quot;</span>, tempDir)</code></pre></div>
<pre><code>$status
[1] 0

$output
character(0)

$error
character(0)

$command
[1] &quot;cd /home/master/brge &amp;&amp; ls /tmp/ssh-6184&quot;

attr(,&quot;class&quot;)
[1] &quot;resource.exec&quot;</code></pre>
<p>Then, we can use R to launch the shell commands</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">client<span class="op">$</span><span class="kw">exec</span>(<span class="st">&#39;plink1&#39;</span>, <span class="kw">c</span>(<span class="st">&#39;--bfile&#39;</span>, <span class="st">&#39;brge&#39;</span>, <span class="st">&#39;--freq&#39;</span>, <span class="st">&#39;--out&#39;</span>, <span class="kw">paste0</span>(tempDir, <span class="st">&#39;/out&#39;</span>), <span class="st">&#39;--noweb&#39;</span>))</code></pre></div>
<pre><code>$status
[1] 0

$output
 [1] &quot;&quot;                                                                          
 [2] &quot;@----------------------------------------------------------@&quot;              
 [3] &quot;|        PLINK!       |     v1.07      |   10/Aug/2009     |&quot;              
 [4] &quot;|----------------------------------------------------------|&quot;              
 [5] &quot;|  (C) 2009 Shaun Purcell, GNU General Public License, v2  |&quot;              
 [6] &quot;|----------------------------------------------------------|&quot;              
 [7] &quot;|  For documentation, citation &amp; bug-report instructions:  |&quot;              
 [8] &quot;|        http://pngu.mgh.harvard.edu/purcell/plink/        |&quot;              
 [9] &quot;@----------------------------------------------------------@&quot;              
[10] &quot;&quot;                                                                          
[11] &quot;Skipping web check... [ --noweb ] &quot;                                        
[12] &quot;Writing this text to log file [ /tmp/ssh-6184/out.log ]&quot;                   
[13] &quot;Analysis started: Mon May 25 08:19:21 2020&quot;                                
[14] &quot;&quot;                                                                          
[15] &quot;Options in effect:&quot;                                                        
[16] &quot;\t--bfile brge&quot;                                                            
[17] &quot;\t--freq&quot;                                                                  
[18] &quot;\t--out /tmp/ssh-6184/out&quot;                                                 
[19] &quot;\t--noweb&quot;                                                                 
[20] &quot;&quot;                                                                          
[21] &quot;Reading map (extended format) from [ brge.bim ] &quot;                          
[22] &quot;100000 markers to be included from [ brge.bim ]&quot;                           
[23] &quot;Reading pedigree information from [ brge.fam ] &quot;                           
[24] &quot;2312 individuals read from [ brge.fam ] &quot;                                  
[25] &quot;2312 individuals with nonmissing phenotypes&quot;                               
[26] &quot;Assuming a disease phenotype (1=unaff, 2=aff, 0=miss)&quot;                     
[27] &quot;Missing phenotype value is also -9&quot;                                        
[28] &quot;725 cases, 1587 controls and 0 missing&quot;                                    
[29] &quot;1097 males, 1215 females, and 0 of unspecified sex&quot;                        
[30] &quot;Reading genotype bitfile from [ brge.bed ] &quot;                               
[31] &quot;Detected that binary PED file is v1.00 SNP-major mode&quot;                     
[32] &quot;Before frequency and genotyping pruning, there are 100000 SNPs&quot;            
[33] &quot;2312 founders and 0 non-founders found&quot;                                    
[34] &quot;6009 heterozygous haploid genotypes; set to missing&quot;                       
[35] &quot;Writing list of heterozygous haploid genotypes to [ /tmp/ssh-6184/out.hh ]&quot;
[36] &quot;7 SNPs with no founder genotypes observed&quot;                                 
[37] &quot;Warning, MAF set to 0 for these SNPs (see --nonfounders)&quot;                  
[38] &quot;Writing list of these SNPs to [ /tmp/ssh-6184/out.nof ]&quot;                   
[39] &quot;Writing allele frequencies (founders-only) to [ /tmp/ssh-6184/out.frq ] &quot;  
[40] &quot;&quot;                                                                          
[41] &quot;Analysis finished: Mon May 25 08:19:26 2020&quot;                               
[42] &quot;&quot;                                                                          

$error
character(0)

$command
[1] &quot;cd /home/master/brge &amp;&amp; plink1 --bfile brge --freq --out /tmp/ssh-6184/out --noweb&quot;

attr(,&quot;class&quot;)
[1] &quot;resource.exec&quot;</code></pre>
<p>The results can be retrieve as an R object</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">outs &lt;-<span class="st"> </span>client<span class="op">$</span><span class="kw">exec</span>(<span class="st">&#39;ls&#39;</span>, tempDir)<span class="op">$</span>output
outs</code></pre></div>
<pre><code>[1] &quot;out.frq&quot; &quot;out.hh&quot;  &quot;out.log&quot; &quot;out.nof&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">client<span class="op">$</span><span class="kw">downloadFile</span>(<span class="kw">paste0</span>(tempDir, <span class="st">&#39;/out.frq&#39;</span>))  </code></pre></div>
<pre><code>NULL</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ans &lt;-<span class="st"> </span>readr<span class="op">::</span><span class="kw">read_table</span>(<span class="st">&quot;out.frq&quot;</span>)
ans</code></pre></div>
<pre><code># A tibble: 100,000 x 6
     CHR SNP         A1    A2         MAF NCHROBS
   &lt;dbl&gt; &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt;    &lt;dbl&gt;   &lt;dbl&gt;
 1     0 MitoC3993T  T     C     0.0149      4572
 2     0 MitoG4821A  A     G     0.00175     4564
 3     0 MitoG6027A  A     G     0.00434     4614
 4     0 MitoT6153C  C     T     0.0106      4616
 5     0 MitoC7275T  T     C     0.000866    4618
 6     0 MitoT9699C  C     T     0.0732      4604
 7     0 MitoA10045G G     A     0.00780     4616
 8     0 MitoG10311A A     G     0.00261     4604
 9     0 MitoA11252G G     A     0.182       4518
10     0 MitoT11900C C     T     0.000868    4608
# … with 99,990 more rows</code></pre>
<p>Finally temporal directories are removed and session closed</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">client<span class="op">$</span><span class="kw">removeTempDir</span>()
client<span class="op">$</span><span class="kw">close</span>()</code></pre></div>

</div>



            </section>

          </div>
        </div>
      </div>
<a href="omic-extension.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="setup.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
